{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2018-05-29T12:49:01.309377Z",
     "start_time": "2018-05-29T12:48:54.668438Z"
    },
    "code_folding": [
     0
    ]
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/mnt/home3/jj374/anaconda36/lib/python3.6/site-packages/statsmodels/compat/pandas.py:56: FutureWarning: The pandas.core.datetools module is deprecated and will be removed in a future version. Please use the pandas.tseries module instead.\n",
      "  from pandas.core import datetools\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "os.getcwd(): /mnt/beegfs/scratch_copy/ahringer/jj374/lab/relmapping\n"
     ]
    }
   ],
   "source": [
    "# Set annotation to annot_ce10\n",
    "%run ~/relmapping/annot/notebooks/__init__.ipynb\n",
    "annot_ = 'annot_ce10'\n",
    "def mp(fp, annot_=annot_): return os.path.join(annot_, 'metrics_type', fp)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2018-05-29T12:49:01.491135Z",
     "start_time": "2018-05-29T12:49:01.311257Z"
    },
    "code_folding": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>atac_wt_emb_height</th>\n",
       "      <th>atac_wt_l1_height</th>\n",
       "      <th>atac_wt_l2_height</th>\n",
       "      <th>atac_wt_l3_height</th>\n",
       "      <th>atac_wt_l4_height</th>\n",
       "      <th>atac_wt_ya_height</th>\n",
       "      <th>atac_glp1_d1_height</th>\n",
       "      <th>atac_glp1_d2_height</th>\n",
       "      <th>atac_glp1_d6_height</th>\n",
       "      <th>atac_glp1_d9_height</th>\n",
       "      <th>atac_glp1_d13_height</th>\n",
       "      <th>atac_source</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chrI</td>\n",
       "      <td>1900</td>\n",
       "      <td>2051</td>\n",
       "      <td>1.487219</td>\n",
       "      <td>1.436852</td>\n",
       "      <td>1.234831</td>\n",
       "      <td>1.476625</td>\n",
       "      <td>1.180041</td>\n",
       "      <td>3.028015</td>\n",
       "      <td>0.984149</td>\n",
       "      <td>0.912876</td>\n",
       "      <td>1.155934</td>\n",
       "      <td>1.187989</td>\n",
       "      <td>1.631096</td>\n",
       "      <td>atac_wt_se</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chrI</td>\n",
       "      <td>3826</td>\n",
       "      <td>3977</td>\n",
       "      <td>67.202148</td>\n",
       "      <td>58.451061</td>\n",
       "      <td>74.416000</td>\n",
       "      <td>133.738358</td>\n",
       "      <td>179.794342</td>\n",
       "      <td>211.372787</td>\n",
       "      <td>88.025909</td>\n",
       "      <td>69.674324</td>\n",
       "      <td>68.506355</td>\n",
       "      <td>64.593590</td>\n",
       "      <td>55.761173</td>\n",
       "      <td>atac_wt_pe</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chrI</td>\n",
       "      <td>4276</td>\n",
       "      <td>4427</td>\n",
       "      <td>1.581892</td>\n",
       "      <td>2.656172</td>\n",
       "      <td>2.186598</td>\n",
       "      <td>3.502825</td>\n",
       "      <td>5.871352</td>\n",
       "      <td>7.771934</td>\n",
       "      <td>1.802739</td>\n",
       "      <td>1.327403</td>\n",
       "      <td>1.547550</td>\n",
       "      <td>2.069787</td>\n",
       "      <td>1.812409</td>\n",
       "      <td>atac_wt_se</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chrI</td>\n",
       "      <td>11272</td>\n",
       "      <td>11423</td>\n",
       "      <td>2.606371</td>\n",
       "      <td>4.443353</td>\n",
       "      <td>9.830266</td>\n",
       "      <td>10.700469</td>\n",
       "      <td>11.683004</td>\n",
       "      <td>13.928377</td>\n",
       "      <td>20.864365</td>\n",
       "      <td>16.739265</td>\n",
       "      <td>17.585819</td>\n",
       "      <td>15.858321</td>\n",
       "      <td>17.166847</td>\n",
       "      <td>atac_wt_pe</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chrI</td>\n",
       "      <td>13070</td>\n",
       "      <td>13221</td>\n",
       "      <td>5.557951</td>\n",
       "      <td>3.608833</td>\n",
       "      <td>3.977978</td>\n",
       "      <td>3.398183</td>\n",
       "      <td>2.689872</td>\n",
       "      <td>2.953438</td>\n",
       "      <td>2.047788</td>\n",
       "      <td>2.084701</td>\n",
       "      <td>1.994170</td>\n",
       "      <td>2.586042</td>\n",
       "      <td>3.195713</td>\n",
       "      <td>atac_wt_pe</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  chrom  start    end  atac_wt_emb_height  atac_wt_l1_height  \\\n",
       "0  chrI   1900   2051            1.487219           1.436852   \n",
       "1  chrI   3826   3977           67.202148          58.451061   \n",
       "2  chrI   4276   4427            1.581892           2.656172   \n",
       "3  chrI  11272  11423            2.606371           4.443353   \n",
       "4  chrI  13070  13221            5.557951           3.608833   \n",
       "\n",
       "   atac_wt_l2_height  atac_wt_l3_height  atac_wt_l4_height  atac_wt_ya_height  \\\n",
       "0           1.234831           1.476625           1.180041           3.028015   \n",
       "1          74.416000         133.738358         179.794342         211.372787   \n",
       "2           2.186598           3.502825           5.871352           7.771934   \n",
       "3           9.830266          10.700469          11.683004          13.928377   \n",
       "4           3.977978           3.398183           2.689872           2.953438   \n",
       "\n",
       "   atac_glp1_d1_height  atac_glp1_d2_height  atac_glp1_d6_height  \\\n",
       "0             0.984149             0.912876             1.155934   \n",
       "1            88.025909            69.674324            68.506355   \n",
       "2             1.802739             1.327403             1.547550   \n",
       "3            20.864365            16.739265            17.585819   \n",
       "4             2.047788             2.084701             1.994170   \n",
       "\n",
       "   atac_glp1_d9_height  atac_glp1_d13_height atac_source  \n",
       "0             1.187989              1.631096  atac_wt_se  \n",
       "1            64.593590             55.761173  atac_wt_pe  \n",
       "2             2.069787              1.812409  atac_wt_se  \n",
       "3            15.858321             17.166847  atac_wt_pe  \n",
       "4             2.586042              3.195713  atac_wt_pe  "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load long cap jump & exon data\n",
    "df_atac = pd.read_csv(os.path.join(annot_, 'accessible_sites.tsv'), sep='\\t')\n",
    "l_atac_peak_pos = df_atac[['start', 'end']].mean(axis=1).map(int)\n",
    "df_atac.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2018-05-29T12:49:03.234770Z",
     "start_time": "2018-05-29T12:49:01.492987Z"
    },
    "code_folding": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>lcap_wt_emb_fwd_ucount_rep1</th>\n",
       "      <th>lcap_wt_emb_fwd_ucount_rep2</th>\n",
       "      <th>lcap_wt_emb_fwd_dcount_rep1</th>\n",
       "      <th>lcap_wt_emb_fwd_dcount_rep2</th>\n",
       "      <th>lcap_wt_emb_fwd_baseMean</th>\n",
       "      <th>lcap_wt_emb_fwd_log2FoldChange</th>\n",
       "      <th>lcap_wt_emb_fwd_lfcSE</th>\n",
       "      <th>lcap_wt_emb_fwd_stat</th>\n",
       "      <th>lcap_wt_emb_fwd_pvalue</th>\n",
       "      <th>lcap_wt_emb_fwd_padj</th>\n",
       "      <th>...</th>\n",
       "      <th>lcap_glp1_d13_rev_baseMean</th>\n",
       "      <th>lcap_glp1_d13_rev_log2FoldChange</th>\n",
       "      <th>lcap_glp1_d13_rev_lfcSE</th>\n",
       "      <th>lcap_glp1_d13_rev_stat</th>\n",
       "      <th>lcap_glp1_d13_rev_pvalue</th>\n",
       "      <th>lcap_glp1_d13_rev_padj</th>\n",
       "      <th>lcap_glp1_d13_rev_passed_jump</th>\n",
       "      <th>lcap_glp1_d13_rev_passed_incr</th>\n",
       "      <th>lcap_glp1_d13_rev_passed</th>\n",
       "      <th>lcap_glp1_d13_rev_summary</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>3.178684</td>\n",
       "      <td>-0.744335</td>\n",
       "      <td>1.818434</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>6.588504e-01</td>\n",
       "      <td>0.960937</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False / -0.74 / 0.961</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.897446</td>\n",
       "      <td>-1.225588</td>\n",
       "      <td>2.121204</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.718293</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>79.875590</td>\n",
       "      <td>9.320055</td>\n",
       "      <td>1.876127</td>\n",
       "      <td>4.967710</td>\n",
       "      <td>3.387408e-07</td>\n",
       "      <td>0.000008</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True / 9.32 / 7.63e-06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.481964</td>\n",
       "      <td>-2.557159</td>\n",
       "      <td>2.857203</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.814603</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>0.935771</td>\n",
       "      <td>0.961622</td>\n",
       "      <td>2.531971</td>\n",
       "      <td>0.379792</td>\n",
       "      <td>3.520500e-01</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False / 0.96 / nan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>10</td>\n",
       "      <td>3</td>\n",
       "      <td>4.368121</td>\n",
       "      <td>1.331422</td>\n",
       "      <td>1.613477</td>\n",
       "      <td>0.825188</td>\n",
       "      <td>0.204632</td>\n",
       "      <td>0.515832</td>\n",
       "      <td>...</td>\n",
       "      <td>1.818028</td>\n",
       "      <td>4.155157</td>\n",
       "      <td>2.536240</td>\n",
       "      <td>1.638314</td>\n",
       "      <td>5.067809e-02</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True / 4.16 / nan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>10</td>\n",
       "      <td>8</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>7.353624</td>\n",
       "      <td>-0.488411</td>\n",
       "      <td>1.176110</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.661030</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.939379</td>\n",
       "      <td>0.966986</td>\n",
       "      <td>2.530644</td>\n",
       "      <td>0.382111</td>\n",
       "      <td>3.511897e-01</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False / 0.97 / nan</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 308 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   lcap_wt_emb_fwd_ucount_rep1  lcap_wt_emb_fwd_ucount_rep2  \\\n",
       "0                            0                            0   \n",
       "1                            6                            0   \n",
       "2                            1                            1   \n",
       "3                            5                            0   \n",
       "4                           10                            8   \n",
       "\n",
       "   lcap_wt_emb_fwd_dcount_rep1  lcap_wt_emb_fwd_dcount_rep2  \\\n",
       "0                            0                            0   \n",
       "1                            1                            1   \n",
       "2                            0                            0   \n",
       "3                           10                            3   \n",
       "4                            6                            6   \n",
       "\n",
       "   lcap_wt_emb_fwd_baseMean  lcap_wt_emb_fwd_log2FoldChange  \\\n",
       "0                  0.000000                             NaN   \n",
       "1                  1.897446                       -1.225588   \n",
       "2                  0.481964                       -2.557159   \n",
       "3                  4.368121                        1.331422   \n",
       "4                  7.353624                       -0.488411   \n",
       "\n",
       "   lcap_wt_emb_fwd_lfcSE  lcap_wt_emb_fwd_stat  lcap_wt_emb_fwd_pvalue  \\\n",
       "0                    NaN                   NaN                     NaN   \n",
       "1               2.121204              0.000000                0.718293   \n",
       "2               2.857203              0.000000                0.814603   \n",
       "3               1.613477              0.825188                0.204632   \n",
       "4               1.176110              0.000000                0.661030   \n",
       "\n",
       "   lcap_wt_emb_fwd_padj            ...              \\\n",
       "0                   NaN            ...               \n",
       "1              1.000000            ...               \n",
       "2                   NaN            ...               \n",
       "3              0.515832            ...               \n",
       "4              1.000000            ...               \n",
       "\n",
       "   lcap_glp1_d13_rev_baseMean  lcap_glp1_d13_rev_log2FoldChange  \\\n",
       "0                    3.178684                         -0.744335   \n",
       "1                   79.875590                          9.320055   \n",
       "2                    0.935771                          0.961622   \n",
       "3                    1.818028                          4.155157   \n",
       "4                    0.939379                          0.966986   \n",
       "\n",
       "   lcap_glp1_d13_rev_lfcSE lcap_glp1_d13_rev_stat  lcap_glp1_d13_rev_pvalue  \\\n",
       "0                 1.818434               0.000000              6.588504e-01   \n",
       "1                 1.876127               4.967710              3.387408e-07   \n",
       "2                 2.531971               0.379792              3.520500e-01   \n",
       "3                 2.536240               1.638314              5.067809e-02   \n",
       "4                 2.530644               0.382111              3.511897e-01   \n",
       "\n",
       "   lcap_glp1_d13_rev_padj  lcap_glp1_d13_rev_passed_jump  \\\n",
       "0                0.960937                          False   \n",
       "1                0.000008                           True   \n",
       "2                     NaN                          False   \n",
       "3                     NaN                          False   \n",
       "4                     NaN                          False   \n",
       "\n",
       "   lcap_glp1_d13_rev_passed_incr  lcap_glp1_d13_rev_passed  \\\n",
       "0                          False                     False   \n",
       "1                           True                      True   \n",
       "2                          False                     False   \n",
       "3                           True                      True   \n",
       "4                          False                     False   \n",
       "\n",
       "   lcap_glp1_d13_rev_summary  \n",
       "0      False / -0.74 / 0.961  \n",
       "1     True / 9.32 / 7.63e-06  \n",
       "2         False / 0.96 / nan  \n",
       "3          True / 4.16 / nan  \n",
       "4         False / 0.97 / nan  \n",
       "\n",
       "[5 rows x 308 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#df_lcap = pd.read_csv('annot/S2_regulatory_annotation/_lcap_jump/lcap_summary.tsv' % locals(), sep='\\t')\n",
    "df_lcap_fwd = pd.read_csv(os.path.join(annot_, 'metrics_lcap', 'lcap_all_fwd.tsv'), sep='\\t')\n",
    "df_lcap_rev = pd.read_csv(os.path.join(annot_, 'metrics_lcap', 'lcap_all_rev.tsv'), sep='\\t')\n",
    "df_lcap = pd.concat([df_lcap_fwd, df_lcap_rev], axis=1)\n",
    "df_lcap.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2018-05-29T12:49:03.279758Z",
     "start_time": "2018-05-29T12:49:03.236523Z"
    },
    "code_folding": [
     0
    ],
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# (Setup stranded structs)\n",
    "df_prom_fwd = df_atac[['chrom', 'start', 'end']].copy()\n",
    "df_prom_rev = df_atac[['chrom', 'start', 'end']].copy()\n",
    "df_prom_fwd['strand'] = '+'\n",
    "df_prom_rev['strand'] = '-'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2018-05-29T12:49:14.950832Z",
     "start_time": "2018-05-29T12:49:03.281273Z"
    },
    "code_folding": [
     0
    ],
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/mnt/home1/ahringer/jj374/relmapping/scripts/yarp/yarp.py:411: FutureWarning: convert_objects is deprecated.  Use the data-type specific converters pd.to_datetime, pd.to_timedelta and pd.to_numeric.\n",
      "  df_name = df_name.convert_objects(convert_numeric=True)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "17922 17074 105494 99143 808 753\n"
     ]
    }
   ],
   "source": [
    "# df_exon1 / df_exon2: first/other exons, coordinates fixed to represent 5' ends\n",
    "#NAMES_EXON = ('chrom', 'start', 'end', 'transcript_id', 'score', 'strand', 'gene_id', 'gene_biotype', 'display_id')\n",
    "l_adj_ = [\n",
    "    'chrom', 'start', 'end', 'score', 'strand', 'thickStart', 'thickEnd', 'itemRgb', 'Name', \n",
    "    'exon_start', 'exon_end', 'utr_start', 'utr_end', 'aoe_end', 'aoe_start', 'gene_biotype', \n",
    "    'exon_id', 'gene_id', 'locus_id', 'init_cutoff_pos', \n",
    "]\n",
    "\n",
    "#df_pass1_exon1_fwd = read_gffbed('WS260_ce10/WS260_ce10.transcripts.annot_pass1_exon1_fwd.bed')\n",
    "#df_pass1_exon1_rev = read_gffbed('WS260_ce10/WS260_ce10.transcripts.annot_pass1_exon1_rev.bed')\n",
    "df_pass1_exon1_fwd = read_gffbed('WS260_ce10/WS260_ce10.transcripts.annot_pass1_exon1_fwd_utradj.bed')[l_adj_]\n",
    "df_pass1_exon1_rev = read_gffbed('WS260_ce10/WS260_ce10.transcripts.annot_pass1_exon1_rev_utradj.bed')[l_adj_]\n",
    "df_pass1_exon2_fwd = read_gffbed('WS260_ce10/WS260_ce10.transcripts.annot_pass1_exon2_fwd.bed')\n",
    "df_pass1_exon2_rev = read_gffbed('WS260_ce10/WS260_ce10.transcripts.annot_pass1_exon2_rev.bed')\n",
    "df_pass2_exon1_fwd = read_gffbed('WS260_ce10/WS260_ce10.transcripts.annot_pass2_exon1_fwd.bed')\n",
    "df_pass2_exon1_rev = read_gffbed('WS260_ce10/WS260_ce10.transcripts.annot_pass2_exon1_rev.bed')\n",
    "\n",
    "# Set 1st exon matching area to TSS + 250bp\n",
    "df_pass1_exon1_fwd['end'] = df_pass1_exon1_fwd['start'] + 1 + 250\n",
    "df_pass1_exon1_rev['start'] = df_pass1_exon1_rev['end'] - 1 - 250\n",
    "\n",
    "print(len(df_pass1_exon1_fwd), len(df_pass1_exon1_rev),\n",
    "      len(df_pass1_exon2_fwd), len(df_pass1_exon2_rev),\n",
    "      len(df_pass2_exon1_fwd), len(df_pass2_exon1_rev),\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2018-05-29T12:49:58.821548Z",
     "start_time": "2018-05-29T12:49:14.954236Z"
    },
    "code_folding": [
     0
    ],
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Match hypersensitive sites to closest first/other exon\n",
    "def df_closest(df_a, df_b, b_prefix, **kwargs):\n",
    "    df_a_sort = df_a\n",
    "    df_b_sort = df_b.sort_values(list(df_b.columns[:3]))\n",
    "    fn_ = BedTool.from_dataframe(df_a).closest(BedTool.from_dataframe(df_b_sort).fn, **kwargs).fn\n",
    "    names_ = list(itertools.chain(df_a.columns.values,\n",
    "        ['%s_%s' % (b_prefix, col) for col in df_b.columns.values],\n",
    "        ['%s_distance' % (b_prefix)]\n",
    "    ))\n",
    "    df_ = pd.read_csv(fn_, names=names_, sep='\\t')\n",
    "    return df_[names_[len(df_a.columns):]]\n",
    "\n",
    "# Closest exon1 -- protein_coding & pseudogene\n",
    "df_ = pd.concat([df_atac['chrom'], l_atac_peak_pos, l_atac_peak_pos + 1], axis=1).copy()\n",
    "df_pass1_exon1_fwd_out_ = df_closest(df_, df_pass1_exon1_fwd, 'pass1_exon1', D='ref', t='first', iu=True)\n",
    "df_pass1_exon1_rev_out_ = df_closest(df_, df_pass1_exon1_rev, 'pass1_exon1', D='ref', t='last', id=True)\n",
    "\n",
    "# Closest exon2 can be both upstream or downstream\n",
    "df_pass1_exon2_fwd_out_ = df_closest(df_, df_pass1_exon2_fwd, 'pass1_exon2', D='ref', t='first')\n",
    "df_pass1_exon2_rev_out_ = df_closest(df_, df_pass1_exon2_rev, 'pass1_exon2', D='ref', t='last')\n",
    "\n",
    "# Closest exon2 can be both upstream or downstream (downstream only)\n",
    "df_pass1_exon2_downstream_fwd_out_ = df_closest(df_, df_pass1_exon2_fwd, 'pass1_exon2_downstream', D='ref', t='first', iu=True)\n",
    "df_pass1_exon2_downstream_rev_out_ = df_closest(df_, df_pass1_exon2_rev, 'pass1_exon2_downstream', D='ref', t='last', id=True)\n",
    "\n",
    "# Closes pass2/exon1 -- tRNA, snoRNA, rRNA, miRNA, etc\n",
    "df_pass2_exon1_fwd_out_ = df_closest(df_, df_pass2_exon1_fwd, 'pass2_exon1', D='ref', t='first', iu=True)\n",
    "df_pass2_exon1_rev_out_ = df_closest(df_, df_pass2_exon1_rev, 'pass2_exon1', D='ref', t='last', id=True)\n",
    "\n",
    "df_exon_fwd = pd.concat([df_pass1_exon1_fwd_out_, df_pass1_exon2_fwd_out_, df_pass1_exon2_downstream_fwd_out_, df_pass2_exon1_fwd_out_], axis=1)\n",
    "df_exon_rev = pd.concat([df_pass1_exon1_rev_out_, df_pass1_exon2_rev_out_, df_pass1_exon2_downstream_rev_out_, df_pass2_exon1_rev_out_], axis=1)\n",
    "\n",
    "df_exon_fwd.to_csv(mp('closest_exon_fwd.tsv'), header=True, index=False, sep='\\t')\n",
    "df_exon_rev.to_csv(mp('closest_exon_rev.tsv'), header=True, index=False, sep='\\t')\n",
    "\n",
    "l_ = ['pass1_exon1_gene_id', 'pass1_exon1_locus_id', 'pass1_exon1_distance']\n",
    "write_gffbed(mp('closest_pass1_exon1_fwd.bed'),\n",
    "    chrom = df_atac['chrom'],\n",
    "    start = df_atac['start'],\n",
    "    end = df_atac['end'],\n",
    "    name = df_exon_fwd['pass1_exon1_Name'],\n",
    "    strand = '+',\n",
    "    attr = df_exon_fwd[l_], \n",
    ")\n",
    "write_gffbed(mp('closest_pass1_exon1_rev.bed'),\n",
    "    chrom = df_atac['chrom'],\n",
    "    start = df_atac['start'],\n",
    "    end = df_atac['end'],\n",
    "    name = df_exon_rev['pass1_exon1_Name'],\n",
    "    strand = '-',\n",
    "    attr = df_exon_rev[l_], \n",
    ")\n",
    "\n",
    "l_ = ['pass1_exon2_gene_id', 'pass1_exon2_locus_id', 'pass1_exon2_distance']\n",
    "write_gffbed(mp('closest_pass1_exon2_fwd.bed'),\n",
    "    chrom = df_atac['chrom'],\n",
    "    start = df_atac['start'],\n",
    "    end = df_atac['end'],\n",
    "    name = df_exon_fwd['pass1_exon2_Name'],\n",
    "    strand = '+',\n",
    "    attr = df_exon_fwd[l_], \n",
    ")\n",
    "write_gffbed(mp('closest_pass1_exon2_rev.bed'),\n",
    "    chrom = df_atac['chrom'],\n",
    "    start = df_atac['start'],\n",
    "    end = df_atac['end'],\n",
    "    name = df_exon_rev['pass1_exon2_Name'],\n",
    "    strand = '-',\n",
    "    attr = df_exon_rev[l_], \n",
    ")\n",
    "\n",
    "l_ = ['pass1_exon2_downstream_gene_id', 'pass1_exon2_downstream_locus_id', 'pass1_exon2_downstream_distance']\n",
    "write_gffbed(mp('closest_pass1_exon2_downstream_fwd.bed'),\n",
    "    chrom = df_atac['chrom'],\n",
    "    start = df_atac['start'],\n",
    "    end = df_atac['end'],\n",
    "    name = df_exon_fwd['pass1_exon2_downstream_Name'],\n",
    "    strand = '+',\n",
    "    attr = df_exon_fwd[l_], \n",
    ")\n",
    "write_gffbed(mp('closest_pass1_exon2_downstream_rev.bed'),\n",
    "    chrom = df_atac['chrom'],\n",
    "    start = df_atac['start'],\n",
    "    end = df_atac['end'],\n",
    "    name = df_exon_rev['pass1_exon2_downstream_Name'],\n",
    "    strand = '-',\n",
    "    attr = df_exon_rev[l_], \n",
    ")\n",
    "\n",
    "l_ = ['pass2_exon1_gene_id', 'pass2_exon1_locus_id', 'pass2_exon1_distance']\n",
    "write_gffbed(mp('closest_pass2_exon1_fwd.bed'),\n",
    "    chrom = df_atac['chrom'],\n",
    "    start = df_atac['start'],\n",
    "    end = df_atac['end'],\n",
    "    name = df_exon_fwd['pass2_exon1_Name'],\n",
    "    strand = '+',\n",
    "    attr = df_exon_fwd[l_], \n",
    ")\n",
    "write_gffbed(mp('closest_pass2_exon1_rev.bed'),\n",
    "    chrom = df_atac['chrom'],\n",
    "    start = df_atac['start'],\n",
    "    end = df_atac['end'],\n",
    "    name = df_exon_rev['pass2_exon1_Name'],\n",
    "    strand = '-',\n",
    "    attr = df_exon_rev[l_], \n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2018-05-29T12:52:44.665917Z",
     "start_time": "2018-05-29T12:49:58.826206Z"
    },
    "code_folding": [
     0
    ],
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[Parallel(n_jobs=15)]: Done  11 out of  11 | elapsed:  1.3min finished\n",
      "[Parallel(n_jobs=15)]: Done  11 out of  11 | elapsed:  1.3min finished\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "42246 annot_ce10/metrics_type/maxgap_fwd.bed\n",
      "42246 annot_ce10/metrics_type/maxgap_rev.bed\n"
     ]
    }
   ],
   "source": [
    "# maximum gap between hypersensitive site and downstream first exons\n",
    "def maxgap(a): return max(len(list(g)) if (k == 0) or (k != k) else 0 for k,g in itertools.groupby(a))\n",
    "\n",
    "def nanmaxgap(a): return maxgap([e if e == e else 0 for e in a])\n",
    "\n",
    "def maxgap_fwd_(fp_inp, chroms, atac_modes, exon1_starts, flank_len=250):\n",
    "    l = []\n",
    "    fh = pyBigWig.open(fp_inp)\n",
    "    for (chrom, atac_mode, exon1_start) in itertools.islice(zip(chroms, atac_modes, exon1_starts), None):\n",
    "        # Interval between atac-mode (+flank), and 5'-end of closest downstream exonl clipped by chromosome ends\n",
    "        start_ = max(atac_mode + flank_len, 0)\n",
    "        end_ = min(exon1_start + 1, fh.chroms(chrom))\n",
    "        if start_ < end_:\n",
    "            l.append(nanmaxgap(np.array(fh.values(chrom, int(start_), int(end_)))))\n",
    "        else:\n",
    "            l.append(-1)\n",
    "    fh.close()\n",
    "    return l\n",
    "\n",
    "def maxgap_rev_(fp_inp, chroms, atac_modes, exon1_ends, flank_len=250):\n",
    "    l = []\n",
    "    fh = pyBigWig.open(fp_inp)\n",
    "    for (chrom, atac_mode, exon1_end) in itertools.islice(zip(chroms, atac_modes, exon1_ends), None):\n",
    "        # Interval between atac-mode (+flank), and 5'-end of closest downstream exonl clipped by chromosome ends\n",
    "        start_ = max(exon1_end - 1, 0)\n",
    "        end_ = min(atac_mode - flank_len + 1, fh.chroms(chrom))\n",
    "        if start_ < end_:\n",
    "            l.append(nanmaxgap(np.array(fh.values(chrom, int(start_), int(end_)))))\n",
    "        else:\n",
    "            l.append(-1)\n",
    "    fh.close()\n",
    "    return l\n",
    "\n",
    "def maxgap_fwd(stage):\n",
    "    step = 'trim20.bwa_pe.rm_unmapped_pe.rm_chrM.rm_rRNA_broad.rm_blacklist.rm_q10.filled_fwd.mean_by_stage'\n",
    "    fp_fwd = pf('lcap808_%s' % (stage,), step, '.bw', 'lcap808')\n",
    "    l_maxgap = maxgap_fwd_(fp_fwd, df_atac['chrom'], l_atac_peak_pos, df_exon_fwd['pass1_exon1_start'])\n",
    "    return pd.DataFrame({'maxgap_%s_fwd' % (stage,): l_maxgap})\n",
    "\n",
    "def maxgap_rev(stage):\n",
    "    step = 'trim20.bwa_pe.rm_unmapped_pe.rm_chrM.rm_rRNA_broad.rm_blacklist.rm_q10.filled_rev.mean_by_stage'\n",
    "    fp_rev = pf('lcap808_%s' % (stage,), step, '.bw', 'lcap808')\n",
    "    l_maxgap = maxgap_rev_(fp_rev, df_atac['chrom'], l_atac_peak_pos, df_exon_rev['pass1_exon1_end'])\n",
    "    return pd.DataFrame({'maxgap_%s_rev' % (stage,): l_maxgap})\n",
    "\n",
    "df_maxgap_fwd = pd.concat(pmap(maxgap_fwd, config['stages'], n_jobs=15), axis=1)\n",
    "df_maxgap_rev = pd.concat(pmap(maxgap_rev, config['stages'], n_jobs=15), axis=1)\n",
    "\n",
    "# Visualise maxgap as .bed-files\n",
    "def itemRgb_(maxgap_):\n",
    "    if maxgap_ == 0:\n",
    "        return RED\n",
    "    else:\n",
    "        return BLUE\n",
    "\n",
    "write_gffbed(mp('maxgap_fwd.bed'),\n",
    "    chrom = df_atac['chrom'],\n",
    "    start = df_atac['start'],\n",
    "    end = df_atac['end'],\n",
    "    name = df_maxgap_fwd.min(axis=1),\n",
    "    strand = '+',\n",
    "    attr = df_maxgap_fwd,\n",
    "    itemRgb = map(itemRgb_, df_maxgap_fwd.min(axis=1)),\n",
    ")\n",
    "\n",
    "write_gffbed(mp('maxgap_rev.bed'),\n",
    "    chrom = df_atac['chrom'],\n",
    "    start = df_atac['start'],\n",
    "    end = df_atac['end'],\n",
    "    name = df_maxgap_rev.min(axis=1),\n",
    "    strand = '-',\n",
    "    attr = df_maxgap_rev,\n",
    "    itemRgb = map(itemRgb_, df_maxgap_rev.min(axis=1)),\n",
    ")\n",
    "\n",
    "!wc -l {mp('maxgap_fwd.bed')}\n",
    "!wc -l {mp('maxgap_rev.bed')}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2018-05-29T12:52:45.895337Z",
     "start_time": "2018-05-29T12:52:44.670595Z"
    },
    "code_folding": [
     0
    ]
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/mnt/home1/ahringer/jj374/relmapping/scripts/yarp/yarp.py:411: FutureWarning: convert_objects is deprecated.  Use the data-type specific converters pd.to_datetime, pd.to_timedelta and pd.to_numeric.\n",
      "  df_name = df_name.convert_objects(convert_numeric=True)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "11073 10928 fwd/rev genes\n",
      "10829 10668 fwd/rev genes longer than 250bp kept\n"
     ]
    }
   ],
   "source": [
    "# df_pass1_gene_%strand <- WS260_ce10 gene coordinates, excluding 250bp downstream of TSS\n",
    "df_ = pd.concat([\n",
    "    yp.read_wbgtf('WS260_ce10/WS260_ce10.genes.pseudogene.gtf.gz', parse_attr=True),\n",
    "    yp.read_wbgtf('WS260_ce10/WS260_ce10.genes.protein_coding.gtf.gz', parse_attr=True),\n",
    "], axis=0)\n",
    "df_['start'] = df_['start'] - 1 # coordinates gtf-to-bed\n",
    "\n",
    "df_pass1_gene_fwd = df_.query('strand == \"+\"').reset_index(drop=True)[['chrom', 'start', 'end', 'gene_id']]\n",
    "df_pass1_gene_rev = df_.query('strand == \"-\"').reset_index(drop=True)[['chrom', 'start', 'end', 'gene_id']]\n",
    "\n",
    "print(len(df_pass1_gene_fwd), len(df_pass1_gene_rev), 'fwd/rev genes')\n",
    "\n",
    "# do not mask 250bp downstream of TSS\n",
    "df_pass1_gene_fwd['start'] = df_pass1_gene_fwd['start'] + 250\n",
    "df_pass1_gene_rev['end'] = df_pass1_gene_rev['end'] - 250\n",
    "\n",
    "df_pass1_gene_fwd = df_pass1_gene_fwd.query('start < end')\n",
    "df_pass1_gene_rev = df_pass1_gene_rev.query('start < end')\n",
    "print(len(df_pass1_gene_fwd), len(df_pass1_gene_rev), 'fwd/rev genes longer than 250bp kept')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2018-05-29T12:52:46.816455Z",
     "start_time": "2018-05-29T12:52:45.897226Z"
    },
    "code_folding": [
     0
    ]
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "42265 42309 fwd/rev gene body hits\n"
     ]
    }
   ],
   "source": [
    "# df_pass1_gpos_%strand[gene_chrom/start/end] <- closest gene; if overlaps multiple genes, choose one with the more distal TSS\n",
    "def df_pos():\n",
    "    l_atac_peak_pos = df_atac[['start', 'end']].mean(axis=1).map(int)\n",
    "    df_ = pd.concat([df_atac['chrom'], l_atac_peak_pos, l_atac_peak_pos + 1], axis=1).copy()\n",
    "    df_.columns = ['chrom', 'start', 'end']\n",
    "    return df_\n",
    "\n",
    "def df_closest_(df_a, df_b, b_prefix, **kwargs):\n",
    "    df_a_sort = df_a\n",
    "    df_b_sort = df_b.sort_values(list(df_b.columns[:3]))\n",
    "    fn_ = BedTool.from_dataframe(df_a).closest(BedTool.from_dataframe(df_b_sort).fn, **kwargs).fn\n",
    "    names_ = list(itertools.chain(df_a.columns.values,\n",
    "        ['%s_%s' % (b_prefix, col) for col in df_b.columns.values],\n",
    "    ))\n",
    "    df_ = pd.read_csv(fn_, sep='\\t', names=names_)\n",
    "    return df_#[names_[len(df_a.columns):]]\n",
    "\n",
    "df_pass1_gpos_fwd_ = df_closest_(df_pos(), df_pass1_gene_fwd, 'pass1_gpos', t='all')\n",
    "df_pass1_gpos_rev_ = df_closest_(df_pos(), df_pass1_gene_rev, 'pass1_gpos', t='all')\n",
    "print(len(df_pass1_gpos_fwd_), len(df_pass1_gpos_rev_), 'fwd/rev gene body hits')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2018-05-29T12:52:55.177413Z",
     "start_time": "2018-05-29T12:52:46.819208Z"
    },
    "code_folding": [
     0
    ]
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "42245 42245\n",
      "33548\n",
      "33808\n"
     ]
    }
   ],
   "source": [
    "# df_pass1_gpos_%strand[is_nearly_intergenic] <- site is outside of gene bodies, or up to 250bp from the TSS\n",
    "df_pass1_gpos_fwd_['is_nearly_intergenic'] = ~((df_pass1_gpos_fwd_['pass1_gpos_start'] <= df_pass1_gpos_fwd_['start'])\\\n",
    "                                           & (df_pass1_gpos_fwd_['end'] <= df_pass1_gpos_fwd_['pass1_gpos_end']))\n",
    "df_pass1_gpos_rev_['is_nearly_intergenic'] = ~((df_pass1_gpos_rev_['pass1_gpos_start'] <= df_pass1_gpos_rev_['start'])\\\n",
    "                                           & (df_pass1_gpos_rev_['end'] <= df_pass1_gpos_rev_['pass1_gpos_end']))\n",
    "\n",
    "l_ = 'is_nearly_intergenic'\n",
    "d_ = {'is_nearly_intergenic': np.any}\n",
    "df_exon_fwd[l_] = df_pass1_gpos_fwd_.groupby(['chrom', 'start', 'end']).agg(d_).reset_index()[l_]\n",
    "df_exon_rev[l_] = df_pass1_gpos_rev_.groupby(['chrom', 'start', 'end']).agg(d_).reset_index()[l_]\n",
    "print(len(df_exon_fwd), len(df_exon_rev))\n",
    "print(len(df_exon_fwd.query('is_nearly_intergenic')))\n",
    "print(len(df_exon_rev.query('is_nearly_intergenic')))\n",
    "\n",
    "# Visualise maxgap as .bed-files\n",
    "def itemRgb_(is_nearly_intergenic_):\n",
    "    if is_nearly_intergenic_:\n",
    "        return yp.ORANGE\n",
    "    else:\n",
    "        return yp.BLUE\n",
    "\n",
    "write_gffbed(mp('is_nearly_intergenic_fwd.bed'),\n",
    "    chrom = df_atac['chrom'],\n",
    "    start = df_atac['start'],\n",
    "    end = df_atac['end'],\n",
    "    #name = df_maxgap_fwd.min(axis=1),\n",
    "    strand = '+',\n",
    "    #attr = df_pass1_gpos_fwd_,\n",
    "    itemRgb = map(itemRgb_, df_exon_fwd['is_nearly_intergenic']),\n",
    ")\n",
    "\n",
    "write_gffbed(mp('is_nearly_intergenic_rev.bed'),\n",
    "    chrom = df_atac['chrom'],\n",
    "    start = df_atac['start'],\n",
    "    end = df_atac['end'],\n",
    "    #name = df_maxgap_rev.min(axis=1),\n",
    "    strand = '-',\n",
    "    #attr = df_maxgap_rev,\n",
    "    itemRgb = map(itemRgb_, df_exon_rev['is_nearly_intergenic']),\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2018-05-29T12:52:55.327536Z",
     "start_time": "2018-05-29T12:52:55.181409Z"
    },
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "df_scap_fwd = pd.read_csv(os.path.join(annot_, 'metrics_scap', 'scap_fwd.tsv'), sep='\\t')\n",
    "df_scap_rev = pd.read_csv(os.path.join(annot_, 'metrics_scap', 'scap_rev.tsv'), sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2018-05-29T12:53:01.467221Z",
     "start_time": "2018-05-29T12:52:55.330090Z"
    },
    "code_folding": [
     0
    ]
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "wt_emb fwd\n",
      "wt_emb rev\n",
      "wt_l1 fwd\n",
      "wt_l1 rev\n",
      "wt_l2 fwd\n",
      "wt_l2 rev\n",
      "wt_l3 fwd\n",
      "wt_l3 rev\n",
      "wt_l4 fwd\n",
      "wt_l4 rev\n",
      "wt_ya fwd\n",
      "wt_ya rev\n",
      "glp1_d1 fwd\n",
      "glp1_d1 rev\n",
      "glp1_d2 fwd\n",
      "glp1_d2 rev\n",
      "glp1_d6 fwd\n",
      "glp1_d6 rev\n",
      "glp1_d9 fwd\n",
      "glp1_d9 rev\n",
      "glp1_d13 fwd\n",
      "glp1_d13 rev\n"
     ]
    }
   ],
   "source": [
    "# df_prom_%strand[annot_detailed_%stage] <- stage-specific detailed annotation, excluding lowconf\n",
    "d_detailed_annot = collections.OrderedDict([\n",
    "    ('coding_promoter_jump_scap', 'coding_promoter'),\n",
    "    ('coding_promoter_incr_scap', 'coding_promoter'),\n",
    "    ('coding_promoter_jump_incr', 'coding_promoter'),\n",
    "    ('coding_promoter_lowconf', 'coding_promoter'),\n",
    "    ('pseudogene_promoter_jump_scap', 'pseudogene_promoter'),\n",
    "    ('pseudogene_promoter_incr_scap', 'pseudogene_promoter'),\n",
    "    ('pseudogene_promoter_jump_incr', 'pseudogene_promoter'),\n",
    "    ('pseudogene_promoter_lowconf', 'pseudogene_promoter'),\n",
    "    ('non-coding_RNA', 'non-coding_RNA'),\n",
    "    ('unknown_promoter_jump_scap', 'unknown_promoter'),\n",
    "    #('unknown_promoter_jump_incr', 'unknown_promoter'),\n",
    "    ('transcription_initiation', 'transcription_initiation'),\n",
    "    ('no_transcription', 'no_transcription'),\n",
    "])\n",
    "\n",
    "d_annot_strand_legend = collections.OrderedDict([\n",
    "    ('coding_promoter', yp.RED),\n",
    "    ('pseudogene_promoter', yp.ORANGE),\n",
    "    ('non-coding_RNA', yp.BLACK),\n",
    "    ('unknown_promoter', yp.YELLOW),\n",
    "    ('transcription_initiation', yp.GREEN),\n",
    "    ('no_transcription', yp.BLUE),\n",
    "])\n",
    "\n",
    "d_annot_legend = collections.OrderedDict([\n",
    "    ('coding_promoter', yp.RED),\n",
    "    ('pseudogene_promoter', yp.ORANGE),\n",
    "    ('non-coding_RNA', yp.BLACK),\n",
    "    ('unknown_promoter', yp.YELLOW),\n",
    "    ('putative_enhancer', yp.GREEN),\n",
    "    ('other_element', yp.BLUE),\n",
    "])\n",
    "\n",
    "def prom_summary_stage(l_chrom, l_start, l_end, l_jump, l_incr, l_maxgap, l_exon1_dist, l_exon2_dist, l_pass2_exon2_dist, l_has_scap, l_scap_diff, l_gene_biotype, l_is_exon1_first, l_is_nearly_intergenic):\n",
    "    for (chrom, start, end, jump, incr, maxgap, exon1_dist, exon2_dist, pass2_exon1_dist, scap_mode_count, scap_diff, gene_biotype, is_exon1_first, is_nearly_intergenic) in \\\n",
    "            zip(l_chrom, l_start, l_end, l_jump, l_incr, l_maxgap, l_exon1_dist, l_exon2_dist, l_pass2_exon2_dist, l_has_scap, l_scap_diff, l_gene_biotype, l_is_exon1_first, l_is_nearly_intergenic):\n",
    "\n",
    "        near_exon1 = abs(exon1_dist) < 250\n",
    "        near_exon2 = abs(exon2_dist) < 250\n",
    "        near_pass2_exon1 = abs(pass2_exon1_dist) < 250\n",
    "        nearly_upstream_of_tss = exon1_dist > -250\n",
    "        has_scap = scap_mode_count >= 2\n",
    "        n_txn_evidence = int(incr) + int(jump) + int(has_scap)\n",
    "        txn_2of3 = n_txn_evidence >= 2\n",
    "\n",
    "        #is_gene_biotype = (gene_biotype == 'protein_coding') | (gene_biotype == 'pseudogene')\n",
    "        #v = False\n",
    "        #if chrom == \"chrI\" and start == 183421 and end == 183568: v = True\n",
    "        #if (chrom == \"chrX\") and (start == 1125552) and (end == 1125703):\n",
    "        #    v = True\n",
    "        #    print(jump, maxgap <= 0, scap_diff >= 0, has_scap)\n",
    "        #if start == 11829601-1: print(incr, jump, has_scap, n_txn_evidence)\n",
    "\n",
    "        if (near_exon1 or not(near_exon2)) and nearly_upstream_of_tss and is_exon1_first \\\n",
    "           and (maxgap <= 0) and (scap_diff >= 0) and txn_2of3:\n",
    "            #if v: print('coding_promoter')\n",
    "            if jump and has_scap:\n",
    "                yield 'coding_promoter_jump_scap' if (gene_biotype == 'protein_coding') else 'pseudogene_promoter_jump_scap'\n",
    "            elif incr and has_scap:\n",
    "                yield 'coding_promoter_incr_scap' if (gene_biotype == 'protein_coding') else 'pseudogene_promoter_incr_scap'\n",
    "            elif jump and incr:\n",
    "                yield 'coding_promoter_jump_incr' if (gene_biotype == 'protein_coding') else 'pseudogene_promoter_jump_incr'\n",
    "            else:\n",
    "                assert False\n",
    "        elif near_pass2_exon1:\n",
    "            #if v: print('non-coding_RNA')\n",
    "            yield 'non-coding_RNA'\n",
    "        elif ((jump and has_scap)) and is_nearly_intergenic and not(near_exon1 or near_exon2):\n",
    "            # or (jump and incr)\n",
    "            #if v: print('unknown_promoter')\n",
    "            if jump and has_scap:\n",
    "                yield 'unknown_promoter_jump_scap'\n",
    "            #elif jump and incr:\n",
    "            #    yield 'unknown_promoter_jump_incr'\n",
    "            else:\n",
    "                assert False\n",
    "        elif has_scap:\n",
    "            #if v: print('transcription_initiation')\n",
    "            yield 'transcription_initiation'\n",
    "        else:\n",
    "            #if v: print('no_transcription')\n",
    "            yield 'no_transcription'\n",
    "\n",
    "for stage in itertools.islice(config['stages'], None):\n",
    "    print(stage, 'fwd')\n",
    "    df_prom_fwd['annot_detailed_%s' % (stage,)] = list(prom_summary_stage(\n",
    "        l_chrom=df_atac['chrom'], l_start=df_atac['start'], l_end=df_atac['end'],\n",
    "        l_jump=df_lcap['lcap_%s_fwd_passed_jump' % (stage,)],\n",
    "        l_incr=df_lcap['lcap_%s_fwd_passed_incr' % (stage,)],\n",
    "        l_maxgap=df_maxgap_fwd['maxgap_%s_fwd' % (stage,)],\n",
    "        #l_exon1_dist=df_exon_fwd['pass1_exon1_distance'],\n",
    "        l_exon1_dist=df_exon_fwd['pass1_exon1_start'] - l_atac_peak_pos,\n",
    "        l_exon2_dist=df_exon_fwd['pass1_exon2_start'] - l_atac_peak_pos,\n",
    "        l_pass2_exon2_dist=df_exon_fwd['pass2_exon1_start'] - l_atac_peak_pos,\n",
    "        l_has_scap=df_scap_fwd['scap_mode_count'],\n",
    "        l_scap_diff=pd.to_numeric(df_exon_fwd['pass1_exon1_init_cutoff_pos'], errors='coerce') - df_scap_fwd['scap_mode'],\n",
    "        l_gene_biotype=df_exon_fwd['pass1_exon1_gene_biotype'],\n",
    "        l_is_exon1_first=df_exon_fwd['pass1_exon1_distance'] <= df_exon_fwd['pass1_exon2_downstream_distance'],\n",
    "        l_is_nearly_intergenic=df_exon_fwd['is_nearly_intergenic'],\n",
    "    ))\n",
    "    print(stage, 'rev')\n",
    "    df_prom_rev['annot_detailed_%s' % (stage,)] = list(prom_summary_stage(\n",
    "        l_chrom=df_atac['chrom'], l_start=df_atac['start'], l_end=df_atac['end'],\n",
    "        l_jump=df_lcap['lcap_%s_rev_passed_jump' % (stage,)],\n",
    "        l_incr=df_lcap['lcap_%s_rev_passed_incr' % (stage,)],\n",
    "        l_maxgap=df_maxgap_rev['maxgap_%s_rev' % (stage,)],\n",
    "        #l_exon1_dist=-df_exon_rev['pass1_exon1_distance'], \n",
    "        l_exon1_dist=l_atac_peak_pos - (df_exon_rev['pass1_exon1_end'] - 1),\n",
    "        l_exon2_dist=l_atac_peak_pos - (df_exon_rev['pass1_exon2_end'] - 1),\n",
    "        l_pass2_exon2_dist=l_atac_peak_pos - (df_exon_rev['pass2_exon1_end'] - 1),\n",
    "        l_has_scap=df_scap_rev['scap_mode_count'],\n",
    "        l_scap_diff=df_scap_rev['scap_mode'] - pd.to_numeric(df_exon_rev['pass1_exon1_init_cutoff_pos'], errors='coerce'),\n",
    "        l_gene_biotype=df_exon_rev['pass1_exon1_gene_biotype'],\n",
    "        l_is_exon1_first=df_exon_rev['pass1_exon2_downstream_distance'] <= df_exon_rev['pass1_exon1_distance'],\n",
    "        l_is_nearly_intergenic=df_exon_rev['is_nearly_intergenic'],\n",
    "    ))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2018-05-29T12:53:07.584699Z",
     "start_time": "2018-05-29T12:53:01.470749Z"
    },
    "code_folding": [
     0
    ],
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# df_prom_%strand[annot_detailed_all] <- stage-aggregated detailed annotation, excluding lowconf\n",
    "def annot_detailed_summary_(l_annot):\n",
    "    l_ranking = list(d_detailed_annot.keys())\n",
    "    l_annot_i = [l_ranking.index(annot) for annot in l_annot]\n",
    "    return l_ranking[min(l_annot_i)]\n",
    "\n",
    "df_prom_fwd['annot_detailed_summary'] = list(map(lambda ir: annot_detailed_summary_(ir[1].tolist()),\n",
    "    df_prom_fwd[['annot_detailed_%(stage)s' % locals() for stage in config['stages']]].iterrows()))\n",
    "\n",
    "df_prom_rev['annot_detailed_summary'] = list(map(lambda ir: annot_detailed_summary_(ir[1].tolist()),\n",
    "    df_prom_rev[['annot_detailed_%(stage)s' % locals() for stage in config['stages']]].iterrows()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2018-05-29T12:53:08.858689Z",
     "start_time": "2018-05-29T12:53:07.589221Z"
    },
    "code_folding": [
     0
    ]
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>wt_emb</th>\n",
       "      <th>wt_l1</th>\n",
       "      <th>wt_l2</th>\n",
       "      <th>wt_l3</th>\n",
       "      <th>wt_l4</th>\n",
       "      <th>wt_ya</th>\n",
       "      <th>glp1_d1</th>\n",
       "      <th>glp1_d2</th>\n",
       "      <th>glp1_d6</th>\n",
       "      <th>glp1_d9</th>\n",
       "      <th>glp1_d13</th>\n",
       "      <th>summary</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>coding_promoter_jump_scap</th>\n",
       "      <td>3228</td>\n",
       "      <td>3612</td>\n",
       "      <td>4238</td>\n",
       "      <td>3502</td>\n",
       "      <td>3598</td>\n",
       "      <td>3163</td>\n",
       "      <td>2472</td>\n",
       "      <td>2109</td>\n",
       "      <td>2459</td>\n",
       "      <td>2413</td>\n",
       "      <td>2444</td>\n",
       "      <td>6042</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>coding_promoter_incr_scap</th>\n",
       "      <td>389</td>\n",
       "      <td>281</td>\n",
       "      <td>292</td>\n",
       "      <td>438</td>\n",
       "      <td>336</td>\n",
       "      <td>450</td>\n",
       "      <td>402</td>\n",
       "      <td>318</td>\n",
       "      <td>482</td>\n",
       "      <td>307</td>\n",
       "      <td>518</td>\n",
       "      <td>784</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>coding_promoter_jump_incr</th>\n",
       "      <td>52</td>\n",
       "      <td>73</td>\n",
       "      <td>94</td>\n",
       "      <td>73</td>\n",
       "      <td>101</td>\n",
       "      <td>104</td>\n",
       "      <td>74</td>\n",
       "      <td>97</td>\n",
       "      <td>71</td>\n",
       "      <td>94</td>\n",
       "      <td>78</td>\n",
       "      <td>368</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>coding_promoter_lowconf</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pseudogene_promoter_jump_scap</th>\n",
       "      <td>49</td>\n",
       "      <td>43</td>\n",
       "      <td>43</td>\n",
       "      <td>32</td>\n",
       "      <td>41</td>\n",
       "      <td>49</td>\n",
       "      <td>24</td>\n",
       "      <td>13</td>\n",
       "      <td>25</td>\n",
       "      <td>19</td>\n",
       "      <td>21</td>\n",
       "      <td>90</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pseudogene_promoter_incr_scap</th>\n",
       "      <td>11</td>\n",
       "      <td>8</td>\n",
       "      <td>10</td>\n",
       "      <td>13</td>\n",
       "      <td>9</td>\n",
       "      <td>3</td>\n",
       "      <td>9</td>\n",
       "      <td>4</td>\n",
       "      <td>9</td>\n",
       "      <td>10</td>\n",
       "      <td>11</td>\n",
       "      <td>20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pseudogene_promoter_jump_incr</th>\n",
       "      <td>8</td>\n",
       "      <td>8</td>\n",
       "      <td>11</td>\n",
       "      <td>7</td>\n",
       "      <td>11</td>\n",
       "      <td>8</td>\n",
       "      <td>6</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pseudogene_promoter_lowconf</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>non-coding_RNA</th>\n",
       "      <td>498</td>\n",
       "      <td>492</td>\n",
       "      <td>487</td>\n",
       "      <td>493</td>\n",
       "      <td>496</td>\n",
       "      <td>493</td>\n",
       "      <td>499</td>\n",
       "      <td>501</td>\n",
       "      <td>496</td>\n",
       "      <td>494</td>\n",
       "      <td>492</td>\n",
       "      <td>469</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>unknown_promoter_jump_scap</th>\n",
       "      <td>468</td>\n",
       "      <td>600</td>\n",
       "      <td>798</td>\n",
       "      <td>406</td>\n",
       "      <td>328</td>\n",
       "      <td>263</td>\n",
       "      <td>189</td>\n",
       "      <td>150</td>\n",
       "      <td>214</td>\n",
       "      <td>285</td>\n",
       "      <td>362</td>\n",
       "      <td>1521</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>transcription_initiation</th>\n",
       "      <td>24432</td>\n",
       "      <td>24038</td>\n",
       "      <td>23203</td>\n",
       "      <td>24190</td>\n",
       "      <td>24264</td>\n",
       "      <td>24650</td>\n",
       "      <td>25477</td>\n",
       "      <td>25978</td>\n",
       "      <td>25387</td>\n",
       "      <td>25544</td>\n",
       "      <td>25225</td>\n",
       "      <td>20139</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>no_transcription</th>\n",
       "      <td>13110</td>\n",
       "      <td>13090</td>\n",
       "      <td>13069</td>\n",
       "      <td>13091</td>\n",
       "      <td>13061</td>\n",
       "      <td>13062</td>\n",
       "      <td>13093</td>\n",
       "      <td>13074</td>\n",
       "      <td>13097</td>\n",
       "      <td>13075</td>\n",
       "      <td>13092</td>\n",
       "      <td>12789</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                               wt_emb  wt_l1  wt_l2  wt_l3  wt_l4  wt_ya  \\\n",
       "coding_promoter_jump_scap        3228   3612   4238   3502   3598   3163   \n",
       "coding_promoter_incr_scap         389    281    292    438    336    450   \n",
       "coding_promoter_jump_incr          52     73     94     73    101    104   \n",
       "coding_promoter_lowconf             0      0      0      0      0      0   \n",
       "pseudogene_promoter_jump_scap      49     43     43     32     41     49   \n",
       "pseudogene_promoter_incr_scap      11      8     10     13      9      3   \n",
       "pseudogene_promoter_jump_incr       8      8     11      7     11      8   \n",
       "pseudogene_promoter_lowconf         0      0      0      0      0      0   \n",
       "non-coding_RNA                    498    492    487    493    496    493   \n",
       "unknown_promoter_jump_scap        468    600    798    406    328    263   \n",
       "transcription_initiation        24432  24038  23203  24190  24264  24650   \n",
       "no_transcription                13110  13090  13069  13091  13061  13062   \n",
       "\n",
       "                               glp1_d1  glp1_d2  glp1_d6  glp1_d9  glp1_d13  \\\n",
       "coding_promoter_jump_scap         2472     2109     2459     2413      2444   \n",
       "coding_promoter_incr_scap          402      318      482      307       518   \n",
       "coding_promoter_jump_incr           74       97       71       94        78   \n",
       "coding_promoter_lowconf              0        0        0        0         0   \n",
       "pseudogene_promoter_jump_scap       24       13       25       19        21   \n",
       "pseudogene_promoter_incr_scap        9        4        9       10        11   \n",
       "pseudogene_promoter_jump_incr        6        1        5        4         2   \n",
       "pseudogene_promoter_lowconf          0        0        0        0         0   \n",
       "non-coding_RNA                     499      501      496      494       492   \n",
       "unknown_promoter_jump_scap         189      150      214      285       362   \n",
       "transcription_initiation         25477    25978    25387    25544     25225   \n",
       "no_transcription                 13093    13074    13097    13075     13092   \n",
       "\n",
       "                               summary  \n",
       "coding_promoter_jump_scap         6042  \n",
       "coding_promoter_incr_scap          784  \n",
       "coding_promoter_jump_incr          368  \n",
       "coding_promoter_lowconf              0  \n",
       "pseudogene_promoter_jump_scap       90  \n",
       "pseudogene_promoter_incr_scap       20  \n",
       "pseudogene_promoter_jump_incr       23  \n",
       "pseudogene_promoter_lowconf          0  \n",
       "non-coding_RNA                     469  \n",
       "unknown_promoter_jump_scap        1521  \n",
       "transcription_initiation         20139  \n",
       "no_transcription                 12789  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Forward strand annotation summary table (excl. lowconf)\n",
    "df_fwd_ = pd.concat([df_prom_fwd['annot_detailed_%(stage)s' % locals()].value_counts() for stage in config['stages'] + ['summary']], axis=1).loc[d_detailed_annot.keys()]\n",
    "df_fwd_.columns = config['stages'] + ['summary']\n",
    "#df_fwd_.to_latex('%(dm_step)s_fig/prom_summary_fwd.latex' % locals())\n",
    "df_fwd_.fillna(0, downcast='infer')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2018-05-29T12:53:09.108681Z",
     "start_time": "2018-05-29T12:53:08.861535Z"
    },
    "code_folding": [
     0
    ]
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>wt_emb</th>\n",
       "      <th>wt_l1</th>\n",
       "      <th>wt_l2</th>\n",
       "      <th>wt_l3</th>\n",
       "      <th>wt_l4</th>\n",
       "      <th>wt_ya</th>\n",
       "      <th>glp1_d1</th>\n",
       "      <th>glp1_d2</th>\n",
       "      <th>glp1_d6</th>\n",
       "      <th>glp1_d9</th>\n",
       "      <th>glp1_d13</th>\n",
       "      <th>summary</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>coding_promoter_jump_scap</th>\n",
       "      <td>3292</td>\n",
       "      <td>3615</td>\n",
       "      <td>4240</td>\n",
       "      <td>3716</td>\n",
       "      <td>3592</td>\n",
       "      <td>3186</td>\n",
       "      <td>2469</td>\n",
       "      <td>2155</td>\n",
       "      <td>2457</td>\n",
       "      <td>2366</td>\n",
       "      <td>2553</td>\n",
       "      <td>6027</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>coding_promoter_incr_scap</th>\n",
       "      <td>339</td>\n",
       "      <td>284</td>\n",
       "      <td>293</td>\n",
       "      <td>354</td>\n",
       "      <td>351</td>\n",
       "      <td>434</td>\n",
       "      <td>379</td>\n",
       "      <td>342</td>\n",
       "      <td>471</td>\n",
       "      <td>365</td>\n",
       "      <td>455</td>\n",
       "      <td>701</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>coding_promoter_jump_incr</th>\n",
       "      <td>72</td>\n",
       "      <td>75</td>\n",
       "      <td>88</td>\n",
       "      <td>97</td>\n",
       "      <td>114</td>\n",
       "      <td>112</td>\n",
       "      <td>81</td>\n",
       "      <td>89</td>\n",
       "      <td>77</td>\n",
       "      <td>92</td>\n",
       "      <td>95</td>\n",
       "      <td>404</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>coding_promoter_lowconf</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pseudogene_promoter_jump_scap</th>\n",
       "      <td>55</td>\n",
       "      <td>54</td>\n",
       "      <td>62</td>\n",
       "      <td>52</td>\n",
       "      <td>48</td>\n",
       "      <td>46</td>\n",
       "      <td>22</td>\n",
       "      <td>11</td>\n",
       "      <td>27</td>\n",
       "      <td>23</td>\n",
       "      <td>26</td>\n",
       "      <td>107</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pseudogene_promoter_incr_scap</th>\n",
       "      <td>13</td>\n",
       "      <td>11</td>\n",
       "      <td>6</td>\n",
       "      <td>11</td>\n",
       "      <td>10</td>\n",
       "      <td>15</td>\n",
       "      <td>6</td>\n",
       "      <td>9</td>\n",
       "      <td>10</td>\n",
       "      <td>6</td>\n",
       "      <td>7</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pseudogene_promoter_jump_incr</th>\n",
       "      <td>7</td>\n",
       "      <td>8</td>\n",
       "      <td>11</td>\n",
       "      <td>8</td>\n",
       "      <td>11</td>\n",
       "      <td>11</td>\n",
       "      <td>5</td>\n",
       "      <td>3</td>\n",
       "      <td>5</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pseudogene_promoter_lowconf</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>non-coding_RNA</th>\n",
       "      <td>498</td>\n",
       "      <td>498</td>\n",
       "      <td>494</td>\n",
       "      <td>498</td>\n",
       "      <td>500</td>\n",
       "      <td>495</td>\n",
       "      <td>505</td>\n",
       "      <td>507</td>\n",
       "      <td>505</td>\n",
       "      <td>499</td>\n",
       "      <td>502</td>\n",
       "      <td>482</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>unknown_promoter_jump_scap</th>\n",
       "      <td>443</td>\n",
       "      <td>617</td>\n",
       "      <td>862</td>\n",
       "      <td>493</td>\n",
       "      <td>366</td>\n",
       "      <td>245</td>\n",
       "      <td>179</td>\n",
       "      <td>172</td>\n",
       "      <td>208</td>\n",
       "      <td>270</td>\n",
       "      <td>378</td>\n",
       "      <td>1611</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>transcription_initiation</th>\n",
       "      <td>24337</td>\n",
       "      <td>23899</td>\n",
       "      <td>23019</td>\n",
       "      <td>23853</td>\n",
       "      <td>24111</td>\n",
       "      <td>24554</td>\n",
       "      <td>25420</td>\n",
       "      <td>25784</td>\n",
       "      <td>25302</td>\n",
       "      <td>25447</td>\n",
       "      <td>25058</td>\n",
       "      <td>20024</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>no_transcription</th>\n",
       "      <td>13189</td>\n",
       "      <td>13184</td>\n",
       "      <td>13170</td>\n",
       "      <td>13163</td>\n",
       "      <td>13142</td>\n",
       "      <td>13147</td>\n",
       "      <td>13179</td>\n",
       "      <td>13173</td>\n",
       "      <td>13183</td>\n",
       "      <td>13171</td>\n",
       "      <td>13165</td>\n",
       "      <td>12841</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                               wt_emb  wt_l1  wt_l2  wt_l3  wt_l4  wt_ya  \\\n",
       "coding_promoter_jump_scap        3292   3615   4240   3716   3592   3186   \n",
       "coding_promoter_incr_scap         339    284    293    354    351    434   \n",
       "coding_promoter_jump_incr          72     75     88     97    114    112   \n",
       "coding_promoter_lowconf             0      0      0      0      0      0   \n",
       "pseudogene_promoter_jump_scap      55     54     62     52     48     46   \n",
       "pseudogene_promoter_incr_scap      13     11      6     11     10     15   \n",
       "pseudogene_promoter_jump_incr       7      8     11      8     11     11   \n",
       "pseudogene_promoter_lowconf         0      0      0      0      0      0   \n",
       "non-coding_RNA                    498    498    494    498    500    495   \n",
       "unknown_promoter_jump_scap        443    617    862    493    366    245   \n",
       "transcription_initiation        24337  23899  23019  23853  24111  24554   \n",
       "no_transcription                13189  13184  13170  13163  13142  13147   \n",
       "\n",
       "                               glp1_d1  glp1_d2  glp1_d6  glp1_d9  glp1_d13  \\\n",
       "coding_promoter_jump_scap         2469     2155     2457     2366      2553   \n",
       "coding_promoter_incr_scap          379      342      471      365       455   \n",
       "coding_promoter_jump_incr           81       89       77       92        95   \n",
       "coding_promoter_lowconf              0        0        0        0         0   \n",
       "pseudogene_promoter_jump_scap       22       11       27       23        26   \n",
       "pseudogene_promoter_incr_scap        6        9       10        6         7   \n",
       "pseudogene_promoter_jump_incr        5        3        5        6         6   \n",
       "pseudogene_promoter_lowconf          0        0        0        0         0   \n",
       "non-coding_RNA                     505      507      505      499       502   \n",
       "unknown_promoter_jump_scap         179      172      208      270       378   \n",
       "transcription_initiation         25420    25784    25302    25447     25058   \n",
       "no_transcription                 13179    13173    13183    13171     13165   \n",
       "\n",
       "                               summary  \n",
       "coding_promoter_jump_scap         6027  \n",
       "coding_promoter_incr_scap          701  \n",
       "coding_promoter_jump_incr          404  \n",
       "coding_promoter_lowconf              0  \n",
       "pseudogene_promoter_jump_scap      107  \n",
       "pseudogene_promoter_incr_scap       19  \n",
       "pseudogene_promoter_jump_incr       29  \n",
       "pseudogene_promoter_lowconf          0  \n",
       "non-coding_RNA                     482  \n",
       "unknown_promoter_jump_scap        1611  \n",
       "transcription_initiation         20024  \n",
       "no_transcription                 12841  "
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Reverse strand annotation summary table (excl. lowconf)\n",
    "df_rev_ = pd.concat([df_prom_rev['annot_detailed_%(stage)s' % locals()].value_counts() for stage in config['stages'] + ['summary']], axis=1).loc[d_detailed_annot.keys()]\n",
    "df_rev_.columns = config['stages'] + ['summary']\n",
    "#df_rev_.to_latex('%(dm_step)s_fig/prom_summary_rev.latex' % locals())\n",
    "df_rev_.fillna(0, downcast='infer')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2018-05-29T12:53:13.155139Z",
     "start_time": "2018-05-29T12:53:09.111540Z"
    },
    "code_folding": [
     0
    ]
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "wt_emb fwd\n",
      "wt_emb rev\n",
      "wt_l1 fwd\n",
      "wt_l1 rev\n",
      "wt_l2 fwd\n",
      "wt_l2 rev\n",
      "wt_l3 fwd\n",
      "wt_l3 rev\n",
      "wt_l4 fwd\n",
      "wt_l4 rev\n",
      "wt_ya fwd\n",
      "wt_ya rev\n",
      "glp1_d1 fwd\n",
      "glp1_d1 rev\n",
      "glp1_d2 fwd\n",
      "glp1_d2 rev\n",
      "glp1_d6 fwd\n",
      "glp1_d6 rev\n",
      "glp1_d9 fwd\n",
      "glp1_d9 rev\n",
      "glp1_d13 fwd\n",
      "glp1_d13 rev\n"
     ]
    }
   ],
   "source": [
    "# df_lowconf_%strand[annot_lowconf_raw_%stage] <- candidate \"raw\" low-confidence annotation (with aggregating)\n",
    "def lowconf_raw_stage(l_chrom, l_start, l_end, l_jump, l_incr, l_log2FoldChange, l_maxgap, l_exon1_dist, l_exon2_dist, l_pass2_exon2_dist, l_has_scap, l_scap_diff, l_gene_biotype, l_is_exon1_first, l_is_nearly_intergenic):\n",
    "    for (chrom, start, end, jump, incr, log2FoldChange, maxgap, exon1_dist, exon2_dist, pass2_exon1_dist, scap_mode_count, scap_diff, gene_biotype, is_exon1_first, is_nearly_intergenic) in \\\n",
    "            zip(l_chrom, l_start, l_end, l_jump, l_incr, l_log2FoldChange, l_maxgap, l_exon1_dist, l_exon2_dist, l_pass2_exon2_dist, l_has_scap, l_scap_diff, l_gene_biotype, l_is_exon1_first, l_is_nearly_intergenic):\n",
    "\n",
    "        near_exon1 = abs(exon1_dist) < 250\n",
    "        near_exon2 = abs(exon2_dist) < 250\n",
    "        near_pass2_exon1 = abs(pass2_exon1_dist) < 250\n",
    "        nearly_upstream_of_tss = exon1_dist > -250\n",
    "        has_scap = scap_mode_count >= 2\n",
    "        has_txn_evidence = (log2FoldChange > 1) #incr or (log2FoldChange > 1 and (has_scap))\n",
    "\n",
    "        v = False\n",
    "        #if chrom == \"chrI\" and start == 183421 and end == 183568: v = True\n",
    "        if (near_exon1 or not(near_exon2)) and nearly_upstream_of_tss and is_exon1_first and\\\n",
    "            is_nearly_intergenic and (not (maxgap > 0)) and has_txn_evidence and (scap_diff >= 0): #-50\n",
    "            #if pass1_exon1_locus_id == \"F59F4.3\":\n",
    "            #    print(chrom, start, end)\n",
    "            yield 'coding_promoter_lowconf' if (gene_biotype == 'protein_coding') else 'pseudogene_promoter_lowconf'\n",
    "        else:\n",
    "            yield 'no_lowconf_annotation'\n",
    "\n",
    "df_lowconf_fwd = df_atac[['chrom', 'start', 'end']].copy()\n",
    "df_lowconf_rev = df_atac[['chrom', 'start', 'end']].copy()\n",
    "df_lowconf_fwd['strand'] = '+'\n",
    "df_lowconf_rev['strand'] = '-'\n",
    "\n",
    "for stage in itertools.islice(config['stages'], None):\n",
    "    print(stage, 'fwd')\n",
    "    df_lowconf_fwd['annot_lowconf_raw_%s' % (stage,)] = list(lowconf_raw_stage(\n",
    "        l_chrom=df_atac['chrom'], l_start=df_atac['start'], l_end=df_atac['end'],\n",
    "        l_jump=df_lcap['lcap_%s_fwd_passed_jump' % (stage,)],\n",
    "        l_incr=df_lcap['lcap_%s_fwd_passed_incr' % (stage,)],\n",
    "        l_log2FoldChange=df_lcap['lcap_%s_fwd_log2FoldChange' % (stage,)],\n",
    "        l_maxgap=df_maxgap_fwd['maxgap_%s_fwd' % (stage,)],\n",
    "        #l_exon1_dist=df_exon_fwd['pass1_exon1_distance'],\n",
    "        l_exon1_dist=df_exon_fwd['pass1_exon1_start'] - l_atac_peak_pos,\n",
    "        l_exon2_dist=df_exon_fwd['pass1_exon2_start'] - l_atac_peak_pos,\n",
    "        l_pass2_exon2_dist=df_exon_fwd['pass2_exon1_start'] - l_atac_peak_pos,\n",
    "        l_has_scap=df_scap_fwd['scap_mode_count'],\n",
    "        l_scap_diff=pd.to_numeric(df_exon_fwd['pass1_exon1_init_cutoff_pos'], errors='coerce') - df_scap_fwd['scap_mode'],\n",
    "        l_gene_biotype=df_exon_fwd['pass1_exon1_gene_biotype'],\n",
    "        #l_pass1_exon1_locus_id=df_exon_fwd['pass1_exon1_locus_id'],\n",
    "        l_is_exon1_first=df_exon_fwd['pass1_exon1_distance'] <= df_exon_fwd['pass1_exon2_downstream_distance'],\n",
    "        l_is_nearly_intergenic=df_exon_fwd['is_nearly_intergenic'],\n",
    "    ))\n",
    "    print(stage, 'rev')\n",
    "    df_lowconf_rev['annot_lowconf_raw_%s' % (stage,)] = list(lowconf_raw_stage(\n",
    "        l_chrom=df_atac['chrom'], l_start=df_atac['start'], l_end=df_atac['end'],\n",
    "        l_jump=df_lcap['lcap_%s_rev_passed_jump' % (stage,)],\n",
    "        l_incr=df_lcap['lcap_%s_rev_passed_incr' % (stage,)],\n",
    "        l_log2FoldChange=df_lcap['lcap_%s_rev_log2FoldChange' % (stage,)],\n",
    "        l_maxgap=df_maxgap_rev['maxgap_%s_rev' % (stage,)],\n",
    "        #l_exon1_dist=-df_exon_rev['pass1_exon1_distance'], \n",
    "        l_exon1_dist=l_atac_peak_pos - (df_exon_rev['pass1_exon1_end'] - 1),\n",
    "        l_exon2_dist=l_atac_peak_pos - (df_exon_rev['pass1_exon2_end'] - 1),\n",
    "        l_pass2_exon2_dist=l_atac_peak_pos - (df_exon_rev['pass2_exon1_end'] - 1),\n",
    "        l_has_scap=df_scap_rev['scap_mode_count'],\n",
    "        l_scap_diff=df_scap_rev['scap_mode'] - pd.to_numeric(df_exon_rev['pass1_exon1_init_cutoff_pos'], errors='coerce'),\n",
    "        l_gene_biotype=df_exon_rev['pass1_exon1_gene_biotype'],\n",
    "        #l_pass1_exon1_locus_id=df_exon_rev['pass1_exon1_locus_id'],\n",
    "        l_is_exon1_first=df_exon_rev['pass1_exon2_downstream_distance'] <= df_exon_rev['pass1_exon1_distance'],\n",
    "        l_is_nearly_intergenic=df_exon_rev['is_nearly_intergenic'],\n",
    "    ))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2018-05-29T12:53:19.288842Z",
     "start_time": "2018-05-29T12:53:13.158176Z"
    },
    "code_folding": [
     0
    ],
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# df_lowconf_%strand[annot_lowconf_raw_summary] <- candidate \"raw\" low-confidence annotation across stages\n",
    "d_annot_lowconf_raw_legend = collections.OrderedDict([\n",
    "    ('coding_promoter_lowconf', yp.RED),\n",
    "    ('pseudogene_promoter_lowconf', yp.ORANGE),\n",
    "    ('no_lowconf_annotation', yp.BLACK),\n",
    "])\n",
    "\n",
    "def annot_lowconf_raw_summary_(l_annot):\n",
    "    l_ranking = list(d_annot_lowconf_raw_legend.keys())\n",
    "    l_annot_i = [l_ranking.index(annot) for annot in l_annot]\n",
    "    return l_ranking[min(l_annot_i)]\n",
    "\n",
    "df_lowconf_fwd['annot_lowconf_raw_summary'] = list(map(lambda ir: annot_lowconf_raw_summary_(ir[1].tolist()),\n",
    "    df_lowconf_fwd[['annot_lowconf_raw_%(stage)s' % locals() for stage in config['stages']]].iterrows()))\n",
    "df_lowconf_rev['annot_lowconf_raw_summary'] = list(map(lambda ir: annot_lowconf_raw_summary_(ir[1].tolist()),\n",
    "    df_lowconf_rev[['annot_lowconf_raw_%(stage)s' % locals() for stage in config['stages']]].iterrows()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2018-05-29T12:53:23.465298Z",
     "start_time": "2018-05-29T12:53:19.292170Z"
    },
    "code_folding": [
     0
    ],
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# df_lowconf_%strand[lcap_log2FoldChange_max] <- max log2FoldChange over all candidates (lowconf_raw) by gene_id\n",
    "\n",
    "# copy fields over in order to group by\n",
    "df_lowconf_fwd['pass1_exon1_gene_id'] = df_exon_fwd['pass1_exon1_gene_id'] \n",
    "df_lowconf_rev['pass1_exon1_gene_id'] = df_exon_rev['pass1_exon1_gene_id']\n",
    "\n",
    "q_ = \"(annot_lowconf_raw_summary == 'coding_promoter_lowconf') | (annot_lowconf_raw_summary == 'pseudogene_promoter_lowconf')\"\n",
    "df_lowconf_fwd['lcap_log2FoldChange'] = df_lcap_fwd[['lcap_%s_fwd_log2FoldChange' % (s_,) for s_ in config['stages']]].max(axis=1)\n",
    "df_lowconf_rev['lcap_log2FoldChange'] = df_lcap_rev[['lcap_%s_rev_log2FoldChange' % (s_,) for s_ in config['stages']]].max(axis=1)\n",
    "\n",
    "df_lowconf_fwd['lcap_log2FoldChange_max'] = df_lowconf_fwd.query(q_)\\\n",
    "    .groupby('pass1_exon1_gene_id')['lcap_log2FoldChange'].transform(np.nanmax)\n",
    "df_lowconf_rev['lcap_log2FoldChange_max'] = df_lowconf_rev.query(q_)\\\n",
    "    .groupby('pass1_exon1_gene_id')['lcap_log2FoldChange'].transform(np.nanmax)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2018-05-29T12:53:31.206685Z",
     "start_time": "2018-05-29T12:53:23.468513Z"
    },
    "code_folding": [
     0
    ]
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fwd strand:\n",
      ".                              41551\n",
      "coding_promoter_lowconf          642\n",
      "pseudogene_promoter_lowconf       52\n",
      "Name: annot_lowconf_summary, dtype: int64\n",
      "Rev strand:\n",
      ".                              41598\n",
      "coding_promoter_lowconf          598\n",
      "pseudogene_promoter_lowconf       49\n",
      "Name: annot_lowconf_summary, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "# df_lowconf_%strand[annot_lowconf_summary] <- low-confidence annotations candidates (=only for \"promoterless genes\")\n",
    "df_lowconf_fwd['annot_detailed_summary'] = df_prom_fwd['annot_detailed_summary']\n",
    "df_lowconf_rev['annot_detailed_summary'] = df_prom_rev['annot_detailed_summary']\n",
    "\n",
    "# list of gene_id without jump/incr promoters, but at least one candidate lowconf promoter\n",
    "l1_ = ['coding_promoter_jump_scap', 'coding_promoter_incr_scap', 'coding_promoter_jump_incr',\n",
    "    'pseudogene_promoter_jump_scap', 'pseudogene_promoter_incr_scap', 'pseudogene_promoter_jump_incr',\n",
    "]\n",
    "l2_ = ['coding_promoter_lowconf', 'pseudogene_promoter_lowconf']\n",
    "l_prom_lowconf_fwd_gene_id = df_lowconf_fwd.groupby('pass1_exon1_gene_id').agg({\n",
    "        'annot_detailed_summary': lambda a: len([a_i for a_i in a if a_i in l1_]),\n",
    "        'annot_lowconf_raw_summary': lambda a: len([a_i for a_i in a if a_i in l2_]),\n",
    "}).query('(annot_detailed_summary == 0) & (annot_lowconf_raw_summary > 0)').index.tolist()\n",
    "\n",
    "l_prom_lowconf_rev_gene_id = df_lowconf_rev.groupby('pass1_exon1_gene_id').agg({\n",
    "        'annot_detailed_summary': lambda a: len([a_i for a_i in a if a_i in l1_]),\n",
    "        'annot_lowconf_raw_summary': lambda a: len([a_i for a_i in a if a_i in l2_]),\n",
    "}).query('(annot_detailed_summary == 0) & (annot_lowconf_raw_summary > 0)').index.tolist()\n",
    "\n",
    "l3_ = [\"coding_promoter_lowconf\", \"pseudogene_promoter_lowconf\"]\n",
    "def annot_lowconf_fwd_summary(ir):\n",
    "    (i, r) = (ir[0], ir[1])\n",
    "    if (r.annot_lowconf_raw_summary in l3_) and (r.pass1_exon1_gene_id in l_prom_lowconf_fwd_gene_id) and (r.lcap_log2FoldChange_max == r.lcap_log2FoldChange):\n",
    "        return r.annot_lowconf_raw_summary\n",
    "    else:\n",
    "        return '.'\n",
    "\n",
    "def annot_lowconf_rev_summary(ir):\n",
    "    (i, r) = (ir[0], ir[1])\n",
    "    if (r.annot_lowconf_raw_summary in l3_) and (r.pass1_exon1_gene_id in l_prom_lowconf_rev_gene_id) and (r.lcap_log2FoldChange_max == r.lcap_log2FoldChange):\n",
    "        return r.annot_lowconf_raw_summary\n",
    "    else:\n",
    "        return '.'\n",
    "\n",
    "df_lowconf_fwd['annot_lowconf_summary'] = list(map(annot_lowconf_fwd_summary, df_lowconf_fwd.iterrows()))\n",
    "df_lowconf_rev['annot_lowconf_summary'] = list(map(annot_lowconf_rev_summary, df_lowconf_rev.iterrows()))\n",
    "print('Fwd strand:')\n",
    "print(df_lowconf_fwd['annot_lowconf_summary'].value_counts())\n",
    "print('Rev strand:')\n",
    "print(df_lowconf_rev['annot_lowconf_summary'].value_counts())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2018-05-29T12:53:38.582513Z",
     "start_time": "2018-05-29T12:53:31.210077Z"
    },
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "# df_lowconf_%strand[annot_lowconf_summary] => .bed-files of the low-confidence annotation\n",
    "d_ = collections.OrderedDict([\n",
    "    ('coding_promoter_lowconf', yp.RED),\n",
    "    ('pseudogene_promoter_lowconf', yp.ORANGE),\n",
    "    ('.', yp.BLACK),\n",
    "])\n",
    "\n",
    "write_gffbed(mp('regulatory_annotation_lowconf_fwd.bed'),\n",
    "        chrom = df_atac['chrom'],\n",
    "        start = df_atac['start'],\n",
    "        end = df_atac['end'],\n",
    "        name = df_exon_fwd['pass1_exon1_locus_id'],\n",
    "        strand = '+',\n",
    "        itemRgb = list(map(lambda lowconf: d_[lowconf], df_lowconf_fwd['annot_lowconf_summary'])),\n",
    "        attr = df_lowconf_fwd['annot_lowconf_summary'],\n",
    ")\n",
    "write_gffbed(mp('regulatory_annotation_lowconf_rev.bed'),\n",
    "        chrom = df_atac['chrom'],\n",
    "        start = df_atac['start'],\n",
    "        end = df_atac['end'],\n",
    "        name = df_exon_rev['pass1_exon1_locus_id'],\n",
    "        strand = '-',\n",
    "        itemRgb = list(map(lambda lowconf: d_[lowconf], df_lowconf_rev['annot_lowconf_summary'])),\n",
    "        attr = df_lowconf_rev['annot_lowconf_summary'],\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2018-05-29T12:53:38.873125Z",
     "start_time": "2018-05-29T12:53:38.587340Z"
    },
    "code_folding": [
     0
    ]
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "694 fwd strand lowconf (pseudo)gene promoters added\n",
      "647 rev strand lowconf (pseudo)gene promoters added\n"
     ]
    }
   ],
   "source": [
    "# df_prom_%strand[annot_detailed_all] <- update with low-confidence annotations\n",
    "m_fwd_ = df_lowconf_fwd['annot_lowconf_summary'] != \".\"\n",
    "m_rev_ = df_lowconf_rev['annot_lowconf_summary'] != \".\"\n",
    "df_prom_fwd.loc[m_fwd_, 'annot_detailed_summary'] = df_lowconf_fwd.loc[m_fwd_, 'annot_lowconf_summary']\n",
    "df_prom_rev.loc[m_rev_, 'annot_detailed_summary'] = df_lowconf_rev.loc[m_rev_, 'annot_lowconf_summary']\n",
    "print('%d fwd strand lowconf (pseudo)gene promoters added' % (sum(m_fwd_),))\n",
    "print('%d rev strand lowconf (pseudo)gene promoters added' % (sum(m_rev_),))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2018-05-29T12:53:39.162720Z",
     "start_time": "2018-05-29T12:53:38.874981Z"
    },
    "code_folding": [
     0
    ]
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>wt_emb</th>\n",
       "      <th>wt_l1</th>\n",
       "      <th>wt_l2</th>\n",
       "      <th>wt_l3</th>\n",
       "      <th>wt_l4</th>\n",
       "      <th>wt_ya</th>\n",
       "      <th>glp1_d1</th>\n",
       "      <th>glp1_d2</th>\n",
       "      <th>glp1_d6</th>\n",
       "      <th>glp1_d9</th>\n",
       "      <th>glp1_d13</th>\n",
       "      <th>summary</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>coding_promoter_jump_scap</th>\n",
       "      <td>3228</td>\n",
       "      <td>3612</td>\n",
       "      <td>4238</td>\n",
       "      <td>3502</td>\n",
       "      <td>3598</td>\n",
       "      <td>3163</td>\n",
       "      <td>2472</td>\n",
       "      <td>2109</td>\n",
       "      <td>2459</td>\n",
       "      <td>2413</td>\n",
       "      <td>2444</td>\n",
       "      <td>6042</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>coding_promoter_incr_scap</th>\n",
       "      <td>389</td>\n",
       "      <td>281</td>\n",
       "      <td>292</td>\n",
       "      <td>438</td>\n",
       "      <td>336</td>\n",
       "      <td>450</td>\n",
       "      <td>402</td>\n",
       "      <td>318</td>\n",
       "      <td>482</td>\n",
       "      <td>307</td>\n",
       "      <td>518</td>\n",
       "      <td>784</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>coding_promoter_jump_incr</th>\n",
       "      <td>52</td>\n",
       "      <td>73</td>\n",
       "      <td>94</td>\n",
       "      <td>73</td>\n",
       "      <td>101</td>\n",
       "      <td>104</td>\n",
       "      <td>74</td>\n",
       "      <td>97</td>\n",
       "      <td>71</td>\n",
       "      <td>94</td>\n",
       "      <td>78</td>\n",
       "      <td>368</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>coding_promoter_lowconf</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>642</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pseudogene_promoter_jump_scap</th>\n",
       "      <td>49</td>\n",
       "      <td>43</td>\n",
       "      <td>43</td>\n",
       "      <td>32</td>\n",
       "      <td>41</td>\n",
       "      <td>49</td>\n",
       "      <td>24</td>\n",
       "      <td>13</td>\n",
       "      <td>25</td>\n",
       "      <td>19</td>\n",
       "      <td>21</td>\n",
       "      <td>90</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pseudogene_promoter_incr_scap</th>\n",
       "      <td>11</td>\n",
       "      <td>8</td>\n",
       "      <td>10</td>\n",
       "      <td>13</td>\n",
       "      <td>9</td>\n",
       "      <td>3</td>\n",
       "      <td>9</td>\n",
       "      <td>4</td>\n",
       "      <td>9</td>\n",
       "      <td>10</td>\n",
       "      <td>11</td>\n",
       "      <td>20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pseudogene_promoter_jump_incr</th>\n",
       "      <td>8</td>\n",
       "      <td>8</td>\n",
       "      <td>11</td>\n",
       "      <td>7</td>\n",
       "      <td>11</td>\n",
       "      <td>8</td>\n",
       "      <td>6</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pseudogene_promoter_lowconf</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>52</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>non-coding_RNA</th>\n",
       "      <td>498</td>\n",
       "      <td>492</td>\n",
       "      <td>487</td>\n",
       "      <td>493</td>\n",
       "      <td>496</td>\n",
       "      <td>493</td>\n",
       "      <td>499</td>\n",
       "      <td>501</td>\n",
       "      <td>496</td>\n",
       "      <td>494</td>\n",
       "      <td>492</td>\n",
       "      <td>460</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>unknown_promoter_jump_scap</th>\n",
       "      <td>468</td>\n",
       "      <td>600</td>\n",
       "      <td>798</td>\n",
       "      <td>406</td>\n",
       "      <td>328</td>\n",
       "      <td>263</td>\n",
       "      <td>189</td>\n",
       "      <td>150</td>\n",
       "      <td>214</td>\n",
       "      <td>285</td>\n",
       "      <td>362</td>\n",
       "      <td>1508</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>transcription_initiation</th>\n",
       "      <td>24432</td>\n",
       "      <td>24038</td>\n",
       "      <td>23203</td>\n",
       "      <td>24190</td>\n",
       "      <td>24264</td>\n",
       "      <td>24650</td>\n",
       "      <td>25477</td>\n",
       "      <td>25978</td>\n",
       "      <td>25387</td>\n",
       "      <td>25544</td>\n",
       "      <td>25225</td>\n",
       "      <td>19799</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>no_transcription</th>\n",
       "      <td>13110</td>\n",
       "      <td>13090</td>\n",
       "      <td>13069</td>\n",
       "      <td>13091</td>\n",
       "      <td>13061</td>\n",
       "      <td>13062</td>\n",
       "      <td>13093</td>\n",
       "      <td>13074</td>\n",
       "      <td>13097</td>\n",
       "      <td>13075</td>\n",
       "      <td>13092</td>\n",
       "      <td>12457</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                               wt_emb  wt_l1  wt_l2  wt_l3  wt_l4  wt_ya  \\\n",
       "coding_promoter_jump_scap        3228   3612   4238   3502   3598   3163   \n",
       "coding_promoter_incr_scap         389    281    292    438    336    450   \n",
       "coding_promoter_jump_incr          52     73     94     73    101    104   \n",
       "coding_promoter_lowconf             0      0      0      0      0      0   \n",
       "pseudogene_promoter_jump_scap      49     43     43     32     41     49   \n",
       "pseudogene_promoter_incr_scap      11      8     10     13      9      3   \n",
       "pseudogene_promoter_jump_incr       8      8     11      7     11      8   \n",
       "pseudogene_promoter_lowconf         0      0      0      0      0      0   \n",
       "non-coding_RNA                    498    492    487    493    496    493   \n",
       "unknown_promoter_jump_scap        468    600    798    406    328    263   \n",
       "transcription_initiation        24432  24038  23203  24190  24264  24650   \n",
       "no_transcription                13110  13090  13069  13091  13061  13062   \n",
       "\n",
       "                               glp1_d1  glp1_d2  glp1_d6  glp1_d9  glp1_d13  \\\n",
       "coding_promoter_jump_scap         2472     2109     2459     2413      2444   \n",
       "coding_promoter_incr_scap          402      318      482      307       518   \n",
       "coding_promoter_jump_incr           74       97       71       94        78   \n",
       "coding_promoter_lowconf              0        0        0        0         0   \n",
       "pseudogene_promoter_jump_scap       24       13       25       19        21   \n",
       "pseudogene_promoter_incr_scap        9        4        9       10        11   \n",
       "pseudogene_promoter_jump_incr        6        1        5        4         2   \n",
       "pseudogene_promoter_lowconf          0        0        0        0         0   \n",
       "non-coding_RNA                     499      501      496      494       492   \n",
       "unknown_promoter_jump_scap         189      150      214      285       362   \n",
       "transcription_initiation         25477    25978    25387    25544     25225   \n",
       "no_transcription                 13093    13074    13097    13075     13092   \n",
       "\n",
       "                               summary  \n",
       "coding_promoter_jump_scap         6042  \n",
       "coding_promoter_incr_scap          784  \n",
       "coding_promoter_jump_incr          368  \n",
       "coding_promoter_lowconf            642  \n",
       "pseudogene_promoter_jump_scap       90  \n",
       "pseudogene_promoter_incr_scap       20  \n",
       "pseudogene_promoter_jump_incr       23  \n",
       "pseudogene_promoter_lowconf         52  \n",
       "non-coding_RNA                     460  \n",
       "unknown_promoter_jump_scap        1508  \n",
       "transcription_initiation         19799  \n",
       "no_transcription                 12457  "
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Forward strand annotation summary table (incl. lowconf across all stages but not the origin stage)\n",
    "df_fwd_ = pd.concat([df_prom_fwd['annot_detailed_%(stage)s' % locals()].value_counts() for stage in config['stages'] + ['summary']], axis=1).loc[d_detailed_annot.keys()]\n",
    "df_fwd_.columns = config['stages'] + ['summary']\n",
    "#df_fwd_.to_latex('%(dm_step)s_fig/prom_summary_fwd.latex' % locals())\n",
    "df_fwd_.fillna(0, downcast='infer')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2018-05-29T12:53:39.405318Z",
     "start_time": "2018-05-29T12:53:39.164958Z"
    },
    "code_folding": [
     0
    ]
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>wt_emb</th>\n",
       "      <th>wt_l1</th>\n",
       "      <th>wt_l2</th>\n",
       "      <th>wt_l3</th>\n",
       "      <th>wt_l4</th>\n",
       "      <th>wt_ya</th>\n",
       "      <th>glp1_d1</th>\n",
       "      <th>glp1_d2</th>\n",
       "      <th>glp1_d6</th>\n",
       "      <th>glp1_d9</th>\n",
       "      <th>glp1_d13</th>\n",
       "      <th>summary</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>coding_promoter_jump_scap</th>\n",
       "      <td>3292</td>\n",
       "      <td>3615</td>\n",
       "      <td>4240</td>\n",
       "      <td>3716</td>\n",
       "      <td>3592</td>\n",
       "      <td>3186</td>\n",
       "      <td>2469</td>\n",
       "      <td>2155</td>\n",
       "      <td>2457</td>\n",
       "      <td>2366</td>\n",
       "      <td>2553</td>\n",
       "      <td>6027</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>coding_promoter_incr_scap</th>\n",
       "      <td>339</td>\n",
       "      <td>284</td>\n",
       "      <td>293</td>\n",
       "      <td>354</td>\n",
       "      <td>351</td>\n",
       "      <td>434</td>\n",
       "      <td>379</td>\n",
       "      <td>342</td>\n",
       "      <td>471</td>\n",
       "      <td>365</td>\n",
       "      <td>455</td>\n",
       "      <td>701</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>coding_promoter_jump_incr</th>\n",
       "      <td>72</td>\n",
       "      <td>75</td>\n",
       "      <td>88</td>\n",
       "      <td>97</td>\n",
       "      <td>114</td>\n",
       "      <td>112</td>\n",
       "      <td>81</td>\n",
       "      <td>89</td>\n",
       "      <td>77</td>\n",
       "      <td>92</td>\n",
       "      <td>95</td>\n",
       "      <td>404</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>coding_promoter_lowconf</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>598</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pseudogene_promoter_jump_scap</th>\n",
       "      <td>55</td>\n",
       "      <td>54</td>\n",
       "      <td>62</td>\n",
       "      <td>52</td>\n",
       "      <td>48</td>\n",
       "      <td>46</td>\n",
       "      <td>22</td>\n",
       "      <td>11</td>\n",
       "      <td>27</td>\n",
       "      <td>23</td>\n",
       "      <td>26</td>\n",
       "      <td>107</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pseudogene_promoter_incr_scap</th>\n",
       "      <td>13</td>\n",
       "      <td>11</td>\n",
       "      <td>6</td>\n",
       "      <td>11</td>\n",
       "      <td>10</td>\n",
       "      <td>15</td>\n",
       "      <td>6</td>\n",
       "      <td>9</td>\n",
       "      <td>10</td>\n",
       "      <td>6</td>\n",
       "      <td>7</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pseudogene_promoter_jump_incr</th>\n",
       "      <td>7</td>\n",
       "      <td>8</td>\n",
       "      <td>11</td>\n",
       "      <td>8</td>\n",
       "      <td>11</td>\n",
       "      <td>11</td>\n",
       "      <td>5</td>\n",
       "      <td>3</td>\n",
       "      <td>5</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pseudogene_promoter_lowconf</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>49</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>non-coding_RNA</th>\n",
       "      <td>498</td>\n",
       "      <td>498</td>\n",
       "      <td>494</td>\n",
       "      <td>498</td>\n",
       "      <td>500</td>\n",
       "      <td>495</td>\n",
       "      <td>505</td>\n",
       "      <td>507</td>\n",
       "      <td>505</td>\n",
       "      <td>499</td>\n",
       "      <td>502</td>\n",
       "      <td>471</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>unknown_promoter_jump_scap</th>\n",
       "      <td>443</td>\n",
       "      <td>617</td>\n",
       "      <td>862</td>\n",
       "      <td>493</td>\n",
       "      <td>366</td>\n",
       "      <td>245</td>\n",
       "      <td>179</td>\n",
       "      <td>172</td>\n",
       "      <td>208</td>\n",
       "      <td>270</td>\n",
       "      <td>378</td>\n",
       "      <td>1596</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>transcription_initiation</th>\n",
       "      <td>24337</td>\n",
       "      <td>23899</td>\n",
       "      <td>23019</td>\n",
       "      <td>23853</td>\n",
       "      <td>24111</td>\n",
       "      <td>24554</td>\n",
       "      <td>25420</td>\n",
       "      <td>25784</td>\n",
       "      <td>25302</td>\n",
       "      <td>25447</td>\n",
       "      <td>25058</td>\n",
       "      <td>19713</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>no_transcription</th>\n",
       "      <td>13189</td>\n",
       "      <td>13184</td>\n",
       "      <td>13170</td>\n",
       "      <td>13163</td>\n",
       "      <td>13142</td>\n",
       "      <td>13147</td>\n",
       "      <td>13179</td>\n",
       "      <td>13173</td>\n",
       "      <td>13183</td>\n",
       "      <td>13171</td>\n",
       "      <td>13165</td>\n",
       "      <td>12531</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                               wt_emb  wt_l1  wt_l2  wt_l3  wt_l4  wt_ya  \\\n",
       "coding_promoter_jump_scap        3292   3615   4240   3716   3592   3186   \n",
       "coding_promoter_incr_scap         339    284    293    354    351    434   \n",
       "coding_promoter_jump_incr          72     75     88     97    114    112   \n",
       "coding_promoter_lowconf             0      0      0      0      0      0   \n",
       "pseudogene_promoter_jump_scap      55     54     62     52     48     46   \n",
       "pseudogene_promoter_incr_scap      13     11      6     11     10     15   \n",
       "pseudogene_promoter_jump_incr       7      8     11      8     11     11   \n",
       "pseudogene_promoter_lowconf         0      0      0      0      0      0   \n",
       "non-coding_RNA                    498    498    494    498    500    495   \n",
       "unknown_promoter_jump_scap        443    617    862    493    366    245   \n",
       "transcription_initiation        24337  23899  23019  23853  24111  24554   \n",
       "no_transcription                13189  13184  13170  13163  13142  13147   \n",
       "\n",
       "                               glp1_d1  glp1_d2  glp1_d6  glp1_d9  glp1_d13  \\\n",
       "coding_promoter_jump_scap         2469     2155     2457     2366      2553   \n",
       "coding_promoter_incr_scap          379      342      471      365       455   \n",
       "coding_promoter_jump_incr           81       89       77       92        95   \n",
       "coding_promoter_lowconf              0        0        0        0         0   \n",
       "pseudogene_promoter_jump_scap       22       11       27       23        26   \n",
       "pseudogene_promoter_incr_scap        6        9       10        6         7   \n",
       "pseudogene_promoter_jump_incr        5        3        5        6         6   \n",
       "pseudogene_promoter_lowconf          0        0        0        0         0   \n",
       "non-coding_RNA                     505      507      505      499       502   \n",
       "unknown_promoter_jump_scap         179      172      208      270       378   \n",
       "transcription_initiation         25420    25784    25302    25447     25058   \n",
       "no_transcription                 13179    13173    13183    13171     13165   \n",
       "\n",
       "                               summary  \n",
       "coding_promoter_jump_scap         6027  \n",
       "coding_promoter_incr_scap          701  \n",
       "coding_promoter_jump_incr          404  \n",
       "coding_promoter_lowconf            598  \n",
       "pseudogene_promoter_jump_scap      107  \n",
       "pseudogene_promoter_incr_scap       19  \n",
       "pseudogene_promoter_jump_incr       29  \n",
       "pseudogene_promoter_lowconf         49  \n",
       "non-coding_RNA                     471  \n",
       "unknown_promoter_jump_scap        1596  \n",
       "transcription_initiation         19713  \n",
       "no_transcription                 12531  "
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Reverse strand annotation summary table (incl. lowconf across all stages but not the origin stage)\n",
    "df_rev_ = pd.concat([df_prom_rev['annot_detailed_%(stage)s' % locals()].value_counts() for stage in config['stages'] + ['summary']], axis=1).loc[d_detailed_annot.keys()]\n",
    "df_rev_.columns = config['stages'] + ['summary']\n",
    "#df_rev_.to_latex('%(dm_step)s_fig/prom_summary_rev.latex' % locals())\n",
    "df_rev_.fillna(0, downcast='infer')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2018-05-29T12:53:39.640879Z",
     "start_time": "2018-05-29T12:53:39.407427Z"
    },
    "code_folding": [
     0
    ],
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# annot_detailed => collapse by method\n",
    "for s_ in config['stages'] + ['summary']:\n",
    "    df_prom_fwd['annot_%s' % (s_,)] = list(map(lambda annot_detailed: d_detailed_annot[annot_detailed], \n",
    "                                               df_prom_fwd['annot_detailed_%s' % (s_,)]))\n",
    "    df_prom_rev['annot_%s' % (s_,)] = list(map(lambda annot_detailed: d_detailed_annot[annot_detailed], \n",
    "                                               df_prom_rev['annot_detailed_%s' % (s_,)]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2018-05-29T12:53:39.834343Z",
     "start_time": "2018-05-29T12:53:39.642599Z"
    },
    "code_folding": [
     0
    ]
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>wt_emb</th>\n",
       "      <th>wt_l1</th>\n",
       "      <th>wt_l2</th>\n",
       "      <th>wt_l3</th>\n",
       "      <th>wt_l4</th>\n",
       "      <th>wt_ya</th>\n",
       "      <th>glp1_d1</th>\n",
       "      <th>glp1_d2</th>\n",
       "      <th>glp1_d6</th>\n",
       "      <th>glp1_d9</th>\n",
       "      <th>glp1_d13</th>\n",
       "      <th>summary</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>coding_promoter</th>\n",
       "      <td>3669</td>\n",
       "      <td>3966</td>\n",
       "      <td>4624</td>\n",
       "      <td>4013</td>\n",
       "      <td>4035</td>\n",
       "      <td>3717</td>\n",
       "      <td>2948</td>\n",
       "      <td>2524</td>\n",
       "      <td>3012</td>\n",
       "      <td>2814</td>\n",
       "      <td>3040</td>\n",
       "      <td>7836</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pseudogene_promoter</th>\n",
       "      <td>68</td>\n",
       "      <td>59</td>\n",
       "      <td>64</td>\n",
       "      <td>52</td>\n",
       "      <td>61</td>\n",
       "      <td>60</td>\n",
       "      <td>39</td>\n",
       "      <td>18</td>\n",
       "      <td>39</td>\n",
       "      <td>33</td>\n",
       "      <td>34</td>\n",
       "      <td>185</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>non-coding_RNA</th>\n",
       "      <td>498</td>\n",
       "      <td>492</td>\n",
       "      <td>487</td>\n",
       "      <td>493</td>\n",
       "      <td>496</td>\n",
       "      <td>493</td>\n",
       "      <td>499</td>\n",
       "      <td>501</td>\n",
       "      <td>496</td>\n",
       "      <td>494</td>\n",
       "      <td>492</td>\n",
       "      <td>460</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>unknown_promoter</th>\n",
       "      <td>468</td>\n",
       "      <td>600</td>\n",
       "      <td>798</td>\n",
       "      <td>406</td>\n",
       "      <td>328</td>\n",
       "      <td>263</td>\n",
       "      <td>189</td>\n",
       "      <td>150</td>\n",
       "      <td>214</td>\n",
       "      <td>285</td>\n",
       "      <td>362</td>\n",
       "      <td>1508</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>transcription_initiation</th>\n",
       "      <td>24432</td>\n",
       "      <td>24038</td>\n",
       "      <td>23203</td>\n",
       "      <td>24190</td>\n",
       "      <td>24264</td>\n",
       "      <td>24650</td>\n",
       "      <td>25477</td>\n",
       "      <td>25978</td>\n",
       "      <td>25387</td>\n",
       "      <td>25544</td>\n",
       "      <td>25225</td>\n",
       "      <td>19799</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>no_transcription</th>\n",
       "      <td>13110</td>\n",
       "      <td>13090</td>\n",
       "      <td>13069</td>\n",
       "      <td>13091</td>\n",
       "      <td>13061</td>\n",
       "      <td>13062</td>\n",
       "      <td>13093</td>\n",
       "      <td>13074</td>\n",
       "      <td>13097</td>\n",
       "      <td>13075</td>\n",
       "      <td>13092</td>\n",
       "      <td>12457</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                          wt_emb  wt_l1  wt_l2  wt_l3  wt_l4  wt_ya  glp1_d1  \\\n",
       "coding_promoter             3669   3966   4624   4013   4035   3717     2948   \n",
       "pseudogene_promoter           68     59     64     52     61     60       39   \n",
       "non-coding_RNA               498    492    487    493    496    493      499   \n",
       "unknown_promoter             468    600    798    406    328    263      189   \n",
       "transcription_initiation   24432  24038  23203  24190  24264  24650    25477   \n",
       "no_transcription           13110  13090  13069  13091  13061  13062    13093   \n",
       "\n",
       "                          glp1_d2  glp1_d6  glp1_d9  glp1_d13  summary  \n",
       "coding_promoter              2524     3012     2814      3040     7836  \n",
       "pseudogene_promoter            18       39       33        34      185  \n",
       "non-coding_RNA                501      496      494       492      460  \n",
       "unknown_promoter              150      214      285       362     1508  \n",
       "transcription_initiation    25978    25387    25544     25225    19799  \n",
       "no_transcription            13074    13097    13075     13092    12457  "
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Forward strand annotation (collapsed across methods)\n",
    "df_fwd_ = pd.concat([df_prom_fwd['annot_%(stage)s' % locals()].value_counts() for stage in config['stages'] + ['summary']], axis=1).loc[d_annot_strand_legend.keys()]\n",
    "df_fwd_.columns = config['stages'] + ['summary']\n",
    "#df_fwd_.to_latex('%(dm_step)s_fig/prom_summary_rev.latex' % locals())\n",
    "#plt.subplot(111, frame_on=False)\n",
    "#plt.gca().xaxis.set_visible(False)\n",
    "#plt.gca().yaxis.set_visible(False)\n",
    "#pd.plotting.table(data=df_fwd_, ax=plt.gca(), loc='center left')\n",
    "#plt.savefig('annot/FigA_mapping/annot_crosstab_fwd.pdf', bbox_inches='tight')\n",
    "df_fwd_"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2018-05-29T12:53:40.223049Z",
     "start_time": "2018-05-29T12:53:39.837406Z"
    },
    "code_folding": [
     0
    ],
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>wt_emb</th>\n",
       "      <th>wt_l1</th>\n",
       "      <th>wt_l2</th>\n",
       "      <th>wt_l3</th>\n",
       "      <th>wt_l4</th>\n",
       "      <th>wt_ya</th>\n",
       "      <th>glp1_d1</th>\n",
       "      <th>glp1_d2</th>\n",
       "      <th>glp1_d6</th>\n",
       "      <th>glp1_d9</th>\n",
       "      <th>glp1_d13</th>\n",
       "      <th>summary</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>coding_promoter</th>\n",
       "      <td>3703</td>\n",
       "      <td>3974</td>\n",
       "      <td>4621</td>\n",
       "      <td>4167</td>\n",
       "      <td>4057</td>\n",
       "      <td>3732</td>\n",
       "      <td>2929</td>\n",
       "      <td>2586</td>\n",
       "      <td>3005</td>\n",
       "      <td>2823</td>\n",
       "      <td>3103</td>\n",
       "      <td>7730</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pseudogene_promoter</th>\n",
       "      <td>75</td>\n",
       "      <td>73</td>\n",
       "      <td>79</td>\n",
       "      <td>71</td>\n",
       "      <td>69</td>\n",
       "      <td>72</td>\n",
       "      <td>33</td>\n",
       "      <td>23</td>\n",
       "      <td>42</td>\n",
       "      <td>35</td>\n",
       "      <td>39</td>\n",
       "      <td>204</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>non-coding_RNA</th>\n",
       "      <td>498</td>\n",
       "      <td>498</td>\n",
       "      <td>494</td>\n",
       "      <td>498</td>\n",
       "      <td>500</td>\n",
       "      <td>495</td>\n",
       "      <td>505</td>\n",
       "      <td>507</td>\n",
       "      <td>505</td>\n",
       "      <td>499</td>\n",
       "      <td>502</td>\n",
       "      <td>471</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>unknown_promoter</th>\n",
       "      <td>443</td>\n",
       "      <td>617</td>\n",
       "      <td>862</td>\n",
       "      <td>493</td>\n",
       "      <td>366</td>\n",
       "      <td>245</td>\n",
       "      <td>179</td>\n",
       "      <td>172</td>\n",
       "      <td>208</td>\n",
       "      <td>270</td>\n",
       "      <td>378</td>\n",
       "      <td>1596</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>transcription_initiation</th>\n",
       "      <td>24337</td>\n",
       "      <td>23899</td>\n",
       "      <td>23019</td>\n",
       "      <td>23853</td>\n",
       "      <td>24111</td>\n",
       "      <td>24554</td>\n",
       "      <td>25420</td>\n",
       "      <td>25784</td>\n",
       "      <td>25302</td>\n",
       "      <td>25447</td>\n",
       "      <td>25058</td>\n",
       "      <td>19713</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>no_transcription</th>\n",
       "      <td>13189</td>\n",
       "      <td>13184</td>\n",
       "      <td>13170</td>\n",
       "      <td>13163</td>\n",
       "      <td>13142</td>\n",
       "      <td>13147</td>\n",
       "      <td>13179</td>\n",
       "      <td>13173</td>\n",
       "      <td>13183</td>\n",
       "      <td>13171</td>\n",
       "      <td>13165</td>\n",
       "      <td>12531</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                          wt_emb  wt_l1  wt_l2  wt_l3  wt_l4  wt_ya  glp1_d1  \\\n",
       "coding_promoter             3703   3974   4621   4167   4057   3732     2929   \n",
       "pseudogene_promoter           75     73     79     71     69     72       33   \n",
       "non-coding_RNA               498    498    494    498    500    495      505   \n",
       "unknown_promoter             443    617    862    493    366    245      179   \n",
       "transcription_initiation   24337  23899  23019  23853  24111  24554    25420   \n",
       "no_transcription           13189  13184  13170  13163  13142  13147    13179   \n",
       "\n",
       "                          glp1_d2  glp1_d6  glp1_d9  glp1_d13  summary  \n",
       "coding_promoter              2586     3005     2823      3103     7730  \n",
       "pseudogene_promoter            23       42       35        39      204  \n",
       "non-coding_RNA                507      505      499       502      471  \n",
       "unknown_promoter              172      208      270       378     1596  \n",
       "transcription_initiation    25784    25302    25447     25058    19713  \n",
       "no_transcription            13173    13183    13171     13165    12531  "
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Reverse strand annotation summary (collapsed across methods)\n",
    "df_rev_ = pd.concat([df_prom_rev['annot_%(stage)s' % locals()].value_counts() for stage in config['stages'] + ['summary']], axis=1).loc[d_annot_strand_legend.keys()]\n",
    "df_rev_.columns = config['stages'] + ['summary']\n",
    "#df_rev_.to_latex('%(dm_step)s_fig/prom_summary_rev.latex' % locals())\n",
    "#plt.subplot(111, frame_on=False)\n",
    "#plt.gca().xaxis.set_visible(False)\n",
    "#plt.gca().yaxis.set_visible(False)\n",
    "#pd.plotting.table(data=df_rev_, ax=plt.gca(), loc='center left')\n",
    "#plt.savefig('annot/FigA_mapping/annot_crosstab_rev.pdf', bbox_inches='tight')\n",
    "df_rev_"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2018-05-29T12:53:40.324355Z",
     "start_time": "2018-05-29T12:53:40.225739Z"
    },
    "code_folding": [
     0
    ]
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>col_0</th>\n",
       "      <th>coding_promoter</th>\n",
       "      <th>pseudogene_promoter</th>\n",
       "      <th>non-coding_RNA</th>\n",
       "      <th>unknown_promoter</th>\n",
       "      <th>transcription_initiation</th>\n",
       "      <th>no_transcription</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>row_0</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>coding_promoter</th>\n",
       "      <td>1976</td>\n",
       "      <td>46</td>\n",
       "      <td>28</td>\n",
       "      <td>637</td>\n",
       "      <td>3696</td>\n",
       "      <td>1453</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pseudogene_promoter</th>\n",
       "      <td>43</td>\n",
       "      <td>9</td>\n",
       "      <td>1</td>\n",
       "      <td>9</td>\n",
       "      <td>71</td>\n",
       "      <td>52</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>non-coding_RNA</th>\n",
       "      <td>25</td>\n",
       "      <td>0</td>\n",
       "      <td>47</td>\n",
       "      <td>6</td>\n",
       "      <td>98</td>\n",
       "      <td>284</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>unknown_promoter</th>\n",
       "      <td>606</td>\n",
       "      <td>11</td>\n",
       "      <td>5</td>\n",
       "      <td>40</td>\n",
       "      <td>713</td>\n",
       "      <td>133</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>transcription_initiation</th>\n",
       "      <td>3660</td>\n",
       "      <td>70</td>\n",
       "      <td>92</td>\n",
       "      <td>737</td>\n",
       "      <td>11150</td>\n",
       "      <td>4090</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>no_transcription</th>\n",
       "      <td>1420</td>\n",
       "      <td>68</td>\n",
       "      <td>298</td>\n",
       "      <td>167</td>\n",
       "      <td>3985</td>\n",
       "      <td>6519</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "col_0                     coding_promoter  pseudogene_promoter  \\\n",
       "row_0                                                            \n",
       "coding_promoter                      1976                   46   \n",
       "pseudogene_promoter                    43                    9   \n",
       "non-coding_RNA                         25                    0   \n",
       "unknown_promoter                      606                   11   \n",
       "transcription_initiation             3660                   70   \n",
       "no_transcription                     1420                   68   \n",
       "\n",
       "col_0                     non-coding_RNA  unknown_promoter  \\\n",
       "row_0                                                        \n",
       "coding_promoter                       28               637   \n",
       "pseudogene_promoter                    1                 9   \n",
       "non-coding_RNA                        47                 6   \n",
       "unknown_promoter                       5                40   \n",
       "transcription_initiation              92               737   \n",
       "no_transcription                     298               167   \n",
       "\n",
       "col_0                     transcription_initiation  no_transcription  \n",
       "row_0                                                                 \n",
       "coding_promoter                               3696              1453  \n",
       "pseudogene_promoter                             71                52  \n",
       "non-coding_RNA                                  98               284  \n",
       "unknown_promoter                               713               133  \n",
       "transcription_initiation                     11150              4090  \n",
       "no_transcription                              3985              6519  "
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Cross-tabulate strand-specific summary annotation\n",
    "pd.crosstab(pd.Categorical(df_prom_fwd['annot_summary']), pd.Categorical(df_prom_rev['annot_summary'])).loc[list(d_annot_strand_legend.keys()), list(d_annot_strand_legend.keys())]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2018-05-29T12:53:40.706936Z",
     "start_time": "2018-05-29T12:53:40.326291Z"
    },
    "code_folding": [
     0
    ]
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "13590 bidirect, 22913 with scap, 10023 both, 73.75\n"
     ]
    }
   ],
   "source": [
    "# Promoter short cap statistics for text\n",
    "m_prom = (df_prom_fwd['annot_summary'] == 'coding_promoter') | (df_prom_rev['annot_summary'] == 'coding_promoter')\n",
    "m_both_scap = (df_scap_fwd['scap_mode_count'] >= 2) & (df_scap_rev['scap_mode_count'] >= 2)\n",
    "f_ = sum(m_prom & m_both_scap) / sum(m_prom)\n",
    "print('%d bidirect, %d with scap, %d both, %.2f' % (sum(m_prom), sum(m_both_scap), sum(m_prom & m_both_scap), 100*f_))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2018-05-29T12:54:13.088993Z",
     "start_time": "2018-05-29T12:53:40.710058Z"
    },
    "code_folding": [
     0
    ]
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "5927 fwd 5867 rev 11794 pooled\n",
      "11794 annot_ce10/metrics_type/_df_prom_bounds.bed\r\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/mnt/home1/ahringer/jj374/relmapping/scripts/yarp/yarp.py:411: FutureWarning: convert_objects is deprecated.  Use the data-type specific converters pd.to_datetime, pd.to_timedelta and pd.to_numeric.\n",
      "  df_name = df_name.convert_objects(convert_numeric=True)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "23155 annot_ce10/metrics_type/_df_gene_bounds.bed\n",
      "23156 annot_ce10/metrics_type/outron-extended_genes.bed\n"
     ]
    }
   ],
   "source": [
    "# S2_outron-extended_genes.bed <- genes extended by outron regions as defined by (aggregated) promoter annotations\n",
    "def df_outron_extended_genes_(df_prom_fwd_, df_prom_rev_):\n",
    "\n",
    "    # df_prom_bounds <- prom min/max boundaries, based on promoter annotation\n",
    "    df_prom_bounds_fwd = df_prom_fwd_.groupby(['pass1_exon1_gene_id', 'pass1_exon1_locus_id']).agg({\n",
    "        'chrom': lambda s: list(set(s))[0],\n",
    "        'start': np.min,\n",
    "        'end': np.max,\n",
    "    }).reset_index()[['chrom', 'start', 'end', 'pass1_exon1_gene_id', 'pass1_exon1_locus_id']]\n",
    "    df_prom_bounds_fwd.columns = ['chrom', 'start', 'end', 'gene_id', 'locus_id']\n",
    "\n",
    "    df_prom_bounds_rev = df_prom_rev_.groupby(['pass1_exon1_gene_id', 'pass1_exon1_locus_id']).agg({\n",
    "        'chrom': lambda s: list(set(s))[0],\n",
    "        'start': np.min,\n",
    "        'end': np.max,\n",
    "    }).reset_index()[['chrom', 'start', 'end', 'pass1_exon1_gene_id', 'pass1_exon1_locus_id']]\n",
    "    df_prom_bounds_rev.columns = ['chrom', 'start', 'end', 'gene_id', 'locus_id']\n",
    "\n",
    "    df_prom_bounds = pd.concat([df_prom_bounds_fwd, df_prom_bounds_rev], axis=0, ignore_index=False)\\\n",
    "        .sort_values(['chrom', 'start', 'end']).reset_index(drop=True)\n",
    "    print('%d fwd %d rev %d pooled' % (len(df_prom_bounds_fwd), len(df_prom_bounds_rev), len(df_prom_bounds)))\n",
    "\n",
    "    # Write a .bed-file of all the regions that changed\n",
    "    df_prom_bounds.to_csv(mp('_df_prom_bounds.bed'), index=False, header=False, sep='\\t')\n",
    "    !wc -l {mp('_df_prom_bounds.bed')}\n",
    "\n",
    "    # df_gene_bounds <- gene min/max boundaries, based on annotation\n",
    "    df_transcripts_ = yp.read_wbgtf('WS260_ce10/WS260_ce10.transcripts.annot.gtf.gz', parse_attr=False)\n",
    "    df_transcripts = yp.df_gfftags_unpack(df_transcripts_, name='attribute')\n",
    "    \n",
    "    # Add display_id\n",
    "    #fp_geneIDs = 'wget/ftp.wormbase.org/pub/wormbase/releases/WS260/species/c_elegans/PRJNA13758/annotation/c_elegans.PRJNA13758.WS260.geneIDs.txt.gz'\n",
    "    #df_geneIDs = pd.read_csv(fp_geneIDs, sep=',', names=('na', 'gene_id', 'locus', 'sequence_id', 'status'))[['gene_id', 'locus', 'sequence_id', 'status']]\n",
    "    #\n",
    "    #def display_id_(locus, sequence_id, gene_id):\n",
    "    #    if locus == locus:\n",
    "    #        return locus\n",
    "    #    elif sequence_id == sequence_id:\n",
    "    #        return sequence_id\n",
    "    #    else:\n",
    "    #        return gene_id\n",
    "    #\n",
    "    #df_geneIDs['locus_id'] = list(map(display_id_, df_geneIDs['locus'], df_geneIDs['sequence_id'], df_geneIDs['gene_id']))\n",
    "    #df_transcripts = pd.merge(df_transcripts, df_geneIDs[['gene_id', 'locus_id']], left_on='gene_id', right_on='gene_id')\n",
    "\n",
    "    df_gene_bounds = df_transcripts.groupby(['gene_id', 'locus_id']).agg({\n",
    "        'chrom': lambda s: list(set(s))[0],\n",
    "        'start': np.min,\n",
    "        'end': np.max,\n",
    "    }).reset_index()[['chrom', 'start', 'end', 'gene_id', 'locus_id']].sort_values(['chrom', 'start', 'end']).reset_index(drop=True)\n",
    "\n",
    "    df_gene_bounds.to_csv(mp('_df_gene_bounds.bed'), index=False, header=False, sep='\\t')\n",
    "    !wc -l {mp('_df_gene_bounds.bed')}\n",
    "    \n",
    "    df_ = df_prom_bounds.merge(df_gene_bounds, how='outer', on=['gene_id', 'locus_id'], suffixes=('_prom', '_gene'))\n",
    "    df_['start'] = list(map(int, map(np.nanmin, zip(df_['start_prom'], df_['start_gene']))))\n",
    "    df_['end'] = list(map(int, map(np.nanmax, zip(df_['end_prom'], df_['end_gene']))))\n",
    "    df_ = df_[['chrom_gene', 'start', 'end', 'gene_id', 'locus_id']].sort_values(['chrom_gene', 'start', 'end']).reset_index(drop=True)\n",
    "    df_.columns = [['chrom', 'start', 'end', 'gene_id', 'locus_id']]\n",
    "    return df_\n",
    "\n",
    "q_ = '(annot_summary == \"coding_promoter\") | (annot_summary == \"pseudogene_promoter\") | (annot_summary == \"non-coding_RNA\")'\n",
    "df_outron_extended_genes = df_outron_extended_genes_(\n",
    "    pd.concat([df_prom_fwd, df_exon_fwd['pass1_exon1_gene_id'], df_exon_fwd['pass1_exon1_locus_id']], axis=1).query(q_)[['chrom', 'start', 'end', 'annot_summary', 'pass1_exon1_gene_id', 'pass1_exon1_locus_id']],\n",
    "    pd.concat([df_prom_rev, df_exon_rev['pass1_exon1_gene_id'], df_exon_rev['pass1_exon1_locus_id']], axis=1).query(q_)[['chrom', 'start', 'end', 'annot_summary', 'pass1_exon1_gene_id', 'pass1_exon1_locus_id']],\n",
    ")\n",
    "df_outron_extended_genes.to_csv(mp('outron-extended_genes.bed'), index=False, header=False, sep='\\t')\n",
    "!wc -l {mp('outron-extended_genes.bed')}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2018-05-29T12:54:13.858448Z",
     "start_time": "2018-05-29T12:54:13.092914Z"
    },
    "code_folding": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "***** WARNING: File /tmp/pybedtools.q109h903.tmp has inconsistent naming convention for record:\n",
      "\t3826\t3977\t.\t.\n",
      "\n",
      "***** WARNING: File /tmp/pybedtools.q109h903.tmp has inconsistent naming convention for record:\n",
      "\t3826\t3977\t.\t.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# df_associated_gene <- associated gene assignments\n",
    "df_associated_gene = pd.read_csv(BedTool.from_dataframe(df_atac[['chrom', 'start', 'end']]).map(\n",
    "    BedTool.from_dataframe(df_outron_extended_genes).fn, c='4,5', o='distinct,distinct',).fn, sep='\\t',\n",
    "    names=yp.NAMES_BED9[:5])\n",
    "df_associated_gene.columns = ['chrom', 'start', 'end', 'associated_gene_id', 'associated_locus_id']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "start_time": "2018-05-29T12:50:06.660Z"
    },
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "# Table S2 -- unstranded summary annotation\n",
    "def name_(annot_fwd, annot_rev, locus_id_fwd, locus_id_rev, enhancer_locus_id):\n",
    "    if (annot_fwd in config['annot_with_gene_id']) and not(annot_rev in config['annot_with_gene_id']):\n",
    "        return locus_id_fwd\n",
    "    elif not(annot_fwd in config['annot_with_gene_id']) and (annot_rev in config['annot_with_gene_id']):\n",
    "        return locus_id_rev\n",
    "    elif (annot_fwd in config['annot_with_gene_id']) and (annot_rev in config['annot_with_gene_id']):\n",
    "        return '%s / %s' % (locus_id_rev, locus_id_fwd)\n",
    "    elif enhancer_locus_id != '.': # show \"likeliest target gene for an enhancer\" in brackets for all remaining (=not coding_promoter or non_coding_promoter)\n",
    "        return '(%s)' % (enhancer_locus_id,)\n",
    "    else:\n",
    "        return ''\n",
    "\n",
    "def annot_merge_(annot_fwd, annot_rev):\n",
    "    if annot_fwd == 'coding_promoter' or annot_rev == 'coding_promoter':\n",
    "        return 'coding_promoter'\n",
    "    elif annot_fwd == 'pseudogene_promoter' or annot_rev == 'pseudogene_promoter':\n",
    "        return 'pseudogene_promoter'\n",
    "    elif annot_fwd == 'non-coding_RNA' or annot_rev == 'non-coding_RNA':\n",
    "        return 'non-coding_RNA'\n",
    "    elif annot_fwd == 'unknown_promoter' or annot_rev == 'unknown_promoter':\n",
    "        return 'unknown_promoter'\n",
    "    elif annot_fwd == 'transcription_initiation' or annot_rev == 'transcription_initiation':\n",
    "        return 'putative_enhancer'\n",
    "    elif annot_fwd == 'no_transcription' or annot_rev == 'no_transcription':\n",
    "        return 'other_element'\n",
    "    assert False\n",
    "\n",
    "def strand_merge_(annot_fwd, annot_rev):\n",
    "    promoter_fwd = (annot_fwd in config['annot_with_gene_id'])\n",
    "    promoter_rev = (annot_rev in config['annot_with_gene_id'])\n",
    "    lcap_fwd = annot_fwd == 'unknown_promoter'\n",
    "    lcap_rev = annot_rev == 'unknown_promoter'\n",
    "    if promoter_fwd and promoter_rev:\n",
    "        return '.'\n",
    "    elif promoter_fwd and (not promoter_rev):\n",
    "        return '+'\n",
    "    elif (not promoter_fwd) and promoter_rev:\n",
    "        return '-'\n",
    "    # in the absence of promoter annotations with an \"annotated\" target, use unknown_promoter annotations for strand\n",
    "    elif lcap_fwd and lcap_rev:\n",
    "        return '.'\n",
    "    elif lcap_fwd and (not lcap_rev):\n",
    "        return '+'\n",
    "    elif (not lcap_fwd) and lcap_rev:\n",
    "        return '-'\n",
    "    else:\n",
    "        return '.'\n",
    "\n",
    "df_regl = df_atac[['chrom', 'start', 'end']].copy()\n",
    "df_regl['tss_fwd'] = df_scap_fwd['scap_mode']\n",
    "df_regl['tss_rev'] = df_scap_rev['scap_mode']\n",
    "df_regl['scap_mode_fwd'] = df_scap_fwd['scap_mode']\n",
    "df_regl['scap_mode_rev'] = df_scap_rev['scap_mode']\n",
    "df_regl['scap_mode_count_fwd'] = df_scap_fwd['scap_mode_count']\n",
    "df_regl['scap_mode_count_rev'] = df_scap_rev['scap_mode_count']\n",
    "df_regl['scap_count_fwd'] = df_scap_fwd['scap_count']\n",
    "df_regl['scap_count_rev'] = df_scap_rev['scap_count']\n",
    "df_regl['annot_fwd'] = df_prom_fwd['annot_summary']\n",
    "df_regl['annot_rev'] = df_prom_rev['annot_summary']\n",
    "df_regl['annot_detailed_fwd'] = df_prom_fwd['annot_detailed_summary']\n",
    "df_regl['annot_detailed_rev'] = df_prom_rev['annot_detailed_summary']\n",
    "df_regl['annot'] = list(map(annot_merge_, df_prom_fwd['annot_summary'], df_prom_rev['annot_summary']))\n",
    "df_regl['strand'] = list(map(strand_merge_, df_prom_fwd['annot_summary'], df_prom_rev['annot_summary']))\n",
    "\n",
    "def agg_pass1_pass2(annot, pass1_val, pass2_val):\n",
    "    if annot in ['coding_promoter', 'pseudogene_promoter']:\n",
    "        return pass1_val\n",
    "    elif annot == 'non-coding_RNA':\n",
    "        return pass2_val\n",
    "    else:\n",
    "        return ''\n",
    "\n",
    "df_regl['promoter_gene_id_fwd'] = list(map(agg_pass1_pass2, df_regl['annot_fwd'],\n",
    "                                    df_exon_fwd['pass1_exon1_gene_id'], df_exon_fwd['pass2_exon1_gene_id']))\n",
    "df_regl['promoter_gene_id_rev'] = list(map(agg_pass1_pass2, df_regl['annot_rev'],\n",
    "                                    df_exon_rev['pass1_exon1_gene_id'], df_exon_rev['pass2_exon1_gene_id']))\n",
    "\n",
    "df_regl['promoter_locus_id_fwd'] = list(map(agg_pass1_pass2, df_regl['annot_fwd'],\n",
    "                                    df_exon_fwd['pass1_exon1_locus_id'], df_exon_fwd['pass2_exon1_locus_id']))\n",
    "df_regl['promoter_locus_id_rev'] = list(map(agg_pass1_pass2, df_regl['annot_rev'],\n",
    "                                    df_exon_rev['pass1_exon1_locus_id'], df_exon_rev['pass2_exon1_locus_id']))\n",
    "\n",
    "df_regl['promoter_gene_biotype_fwd'] = list(map(agg_pass1_pass2, df_regl['annot_fwd'],\n",
    "                                    df_exon_fwd['pass1_exon1_gene_biotype'], df_exon_fwd['pass2_exon1_gene_biotype']))\n",
    "df_regl['promoter_gene_biotype_rev'] = list(map(agg_pass1_pass2, df_regl['annot_rev'],\n",
    "                                    df_exon_rev['pass1_exon1_gene_biotype'], df_exon_rev['pass2_exon1_gene_biotype']))\n",
    "\n",
    "df_regl['associated_gene_id'] = df_associated_gene['associated_gene_id']\n",
    "df_regl['associated_locus_id'] = df_associated_gene['associated_locus_id']\n",
    "\n",
    "df_regl['label'] = list(map(name_,\n",
    "    df_regl['annot_fwd'], df_regl['annot_rev'],                     \n",
    "    df_regl['promoter_locus_id_fwd'], df_regl['promoter_locus_id_rev'],\n",
    "    df_regl['associated_locus_id']\n",
    "))\n",
    "\n",
    "for stage in config['stages_wt'] + config['stages_glp1']:\n",
    "    df_regl['atac_%s_height' % (stage,)] = df_atac['atac_%s_height' % (stage,)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "start_time": "2018-05-29T12:50:06.662Z"
    },
    "code_folding": [
     0
    ]
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Input file  = annot_ce10/metrics_type/_liftover_fp_inp.tsv \r\n",
      "Output file = annot_ce10/metrics_type/_liftover_fp_out.tsv \r\n",
      "Lines processed from input file              = 42245\r\n",
      "Comment / blank lines written to output file = 0\r\n",
      "Data lines written to output file            = 42245\r\n",
      "Number of records changed                    = 40852\r\n"
     ]
    }
   ],
   "source": [
    "# Liftover coordinates to ce11\n",
    "def liftover_(chroms, starts):#, ends, strands): # Liftover wrapper\n",
    "    \"\"\"\n",
    "    # http://www.ncbi.nlm.nih.gov/pmc/articles/PMC4512556/\n",
    "    # note that while annotations changed, the N2 genome sequence remained identical from WS215 through WS234\n",
    "    Test: \n",
    "    list(liftover_WS230_WS220(\n",
    "    chroms=['chrI', 'chrX', 'chrIII', 'chrV', 'chrIV', 'chrII'],\n",
    "    starts=[1,1,4,1,1, 1],\n",
    "    ends=[12507921,13713648,471811,15465174,11718826, 11718826],\n",
    "    strands=['+', '-', '+', '-', '+']))\n",
    "    \"\"\"\n",
    "    fp_inp = mp('_liftover_fp_inp.tsv')\n",
    "    fp_out = mp('_liftover_fp_out.tsv')\n",
    "    with open(fp_inp, 'w') as fh_inp:\n",
    "        #for (chrom, start, end, strand) in zip(chroms, starts, ends, strands):        \n",
    "        for (chrom, start) in zip(chroms, starts):\n",
    "            print('%s\\t%s' % (chrom, start), file=fh_inp)\n",
    "\n",
    "    liftover_bin = '/mnt/home1/ahringer/jj374/lab/raw_data/datastore/Kenneth_Evans/ConvertWormAssembly-1-0-0/ConvertWormAssembly.1.0.0.pl'\n",
    "    liftover_dir = '/mnt/home1/ahringer/jj374/lab/raw_data/datastore/Kenneth_Evans/ConvertWormAssembly-1-0-0/CHROMOSOME-DIFFERENCES-to-WS244'\n",
    "    !{liftover_bin} \\\n",
    "        --input={fp_inp} \\\n",
    "        --output={fp_out} \\\n",
    "        --release1=WS220 \\\n",
    "        --release2=WS235 \\\n",
    "        --chrfield=1 \\\n",
    "        --posfield=2 \\\n",
    "        --postype=1 \\\n",
    "        --refdir={liftover_dir}\n",
    "    \n",
    "    df_out = pd.read_csv(fp_out, sep='\\t', names=['chrom', 'start'])\n",
    "    for i, r in df_out.iterrows():\n",
    "        yield((r['chrom'], r['start']))\n",
    "\n",
    "df_ce11_pos = pd.DataFrame.from_records(liftover_(df_regl['chrom'], summit_pos(df_regl)), columns=['chrom', 'pos'])\n",
    "df_regl['chrom_ce11'] = df_ce11_pos['chrom']\n",
    "df_regl['start_ce11'] = df_ce11_pos['pos'] - 75\n",
    "df_regl['end_ce11'] = df_ce11_pos['pos'] + 75 + 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "start_time": "2018-05-29T12:50:06.687Z"
    }
   },
   "outputs": [],
   "source": [
    "fp_ = os.path.join(annot_, 'regulatory_annotation.tsv')\n",
    "l_ = [\n",
    "    'chrom', 'start', 'end', \n",
    "    'annot', 'annot_fwd', 'annot_rev',\n",
    "    'promoter_gene_id_fwd', 'promoter_locus_id_fwd', 'promoter_gene_biotype_fwd',\n",
    "    'promoter_gene_id_rev', 'promoter_locus_id_rev', 'promoter_gene_biotype_rev',\n",
    "    'associated_gene_id', 'associated_locus_id',\n",
    "] + ['atac_%s_height' % (stage,) for stage in config['stages_wt'] + config['stages_glp1']] + [\\\n",
    "    'tss_fwd', 'tss_rev',\n",
    "    #'scap_mode_count_fwd', 'scap_mode_count_rev',\n",
    "    #'scap_count_fwd', 'scap_count_rev',\n",
    "    'chrom_ce11', 'start_ce11', 'end_ce11',\n",
    "]\n",
    "\n",
    "pd.concat([df_regl[l_], df_lcap], axis=1).to_csv(fp_, header=True, index=False, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "start_time": "2018-05-29T12:50:06.689Z"
    },
    "code_folding": [
     0,
     1
    ]
   },
   "outputs": [],
   "source": [
    "# Short cap bidirectionality statistics\n",
    "def scap_bidirectionality(mode_count_fwd, mode_count_rev):\n",
    "    if mode_count_fwd >= 2 and mode_count_rev >= 2:\n",
    "        return 'bidirectional'\n",
    "    elif mode_count_fwd >= 2:\n",
    "        return 'forward'\n",
    "    elif mode_count_rev >= 2:\n",
    "        return 'reverse'\n",
    "    else:\n",
    "        return 'none/weak'\n",
    "\n",
    "df_has_scap = pd.DataFrame()\n",
    "df_has_scap['annot'] = df_regl['annot']\n",
    "df_has_scap['scap_bidirectionality'] = list(map(scap_bidirectionality, \n",
    "    df_scap_fwd['scap_mode_count'], df_scap_rev['scap_mode_count']))\n",
    "df_has_scap['scap_bidirectionality'].value_counts()\n",
    "\n",
    "df_ = pd.crosstab(pd.Categorical(df_has_scap['annot']), pd.Categorical(df_has_scap['scap_bidirectionality']),\n",
    "                 margins=True)\\\n",
    ".loc[config['annot'] + ['All'], ['bidirectional', 'forward', 'reverse', 'none/weak']]\n",
    "\n",
    "fig = plt.figure(figsize=(4,2))\n",
    "plt.subplot(111, frame_on=False)\n",
    "plt.gca().xaxis.set_visible(False)\n",
    "plt.gca().yaxis.set_visible(False)\n",
    "pd.plotting.table(data=df_, ax=plt.gca(), loc='center left')\n",
    "#plt.savefig('annot/FigA_mapping/annot_crosstab_scap.pdf', bbox_inches='tight')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "start_time": "2018-05-29T12:50:06.691Z"
    },
    "code_folding": [
     0
    ],
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Strand/stage-specific annotations as .bed-tracks\n",
    "write_gffbed(os.path.join(annot_, 'regulatory_annotation_fwd.bed'),\n",
    "        chrom = df_regl['chrom'],\n",
    "        start = df_regl['start'],\n",
    "        end = df_regl['end'],\n",
    "        name = df_regl['promoter_locus_id_fwd'],\n",
    "        strand = '+',\n",
    "        itemRgb = list(map(lambda annot: d_annot_strand_legend[annot], df_regl['annot_fwd'])),\n",
    "        attr = df_prom_fwd[['annot_%s' % (stage,) for stage in config['stages']]],\n",
    ")\n",
    "\n",
    "write_gffbed(os.path.join(annot_, 'regulatory_annotation_rev.bed'),\n",
    "        chrom = df_regl['chrom'],\n",
    "        start = df_regl['start'],\n",
    "        end = df_regl['end'],\n",
    "        name = df_regl['promoter_locus_id_rev'],\n",
    "        strand = '-',\n",
    "        itemRgb = list(map(lambda annot: d_annot_strand_legend[annot], df_regl['annot_rev'])),\n",
    "        attr = df_prom_rev[['annot_%s' % (stage,) for stage in config['stages']]],\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "start_time": "2018-05-29T12:50:06.693Z"
    },
    "code_folding": [],
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Visualise annotation as a single, unstranded .bed-file\n",
    "def write_regl_bed9(fp_, df_regl_):\n",
    "    write_gffbed(fp_,\n",
    "        chrom = df_regl_['chrom'],\n",
    "        start = df_regl_['start'],\n",
    "        end = df_regl_['end'],\n",
    "        name = df_regl_['label'],\n",
    "        strand = df_regl_['strand'],\n",
    "        itemRgb = list(map(lambda annot: d_annot_legend[annot], df_regl_['annot'])),\n",
    "        attr = df_regl_[['annot', 'annot_fwd', 'annot_rev', 'annot_detailed_fwd', 'annot_detailed_rev']],\n",
    "    )\n",
    "\n",
    "write_regl_bed9(os.path.join(annot_, 'regulatory_annotation_ce10.bed'), df_regl)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "start_time": "2018-05-29T12:50:06.694Z"
    },
    "code_folding": [
     0
    ],
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Visualise annotation as a single, unstranded .bed-file (ce11)\n",
    "def write_regl_bed9(fp_, df_regl_):\n",
    "    write_gffbed(fp_,\n",
    "        chrom = df_regl_['chrom_ce11'],\n",
    "        start = df_regl_['start_ce11'],\n",
    "        end = df_regl_['end_ce11'],\n",
    "        name = df_regl_['label'],\n",
    "        strand = df_regl_['strand'],\n",
    "        itemRgb = list(map(lambda annot: d_annot_legend[annot], df_regl_['annot'])),\n",
    "        attr = df_regl_[['annot', 'annot_fwd', 'annot_rev', 'annot_detailed_fwd', 'annot_detailed_rev']],\n",
    "    )\n",
    "\n",
    "write_regl_bed9(os.path.join(annot_, 'regulatory_annotation_ce11.bed'), df_regl)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "start_time": "2018-05-29T12:50:06.696Z"
    },
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "# .bed-files of specific types of elements\n",
    "for annot in config['annot']:\n",
    "    fp_ = mp('regulatory_annotation.%s_only.bed' % (annot,))\n",
    "    write_regl_bed9(fp_, df_regl.query('annot == \"%s\"' % (annot,)).reset_index(drop=True))\n",
    "    !wc -l {fp_}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "start_time": "2018-05-29T12:50:06.698Z"
    },
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "# Visualise number of promoters called\n",
    "n_bid = len(df_regl.query('(annot == \"coding_promoter\") & (strand == \".\")'))\n",
    "n_fwd = len(df_regl.query('(annot == \"coding_promoter\") & (strand == \"+\")'))\n",
    "n_rev = len(df_regl.query('(annot == \"coding_promoter\") & (strand == \"-\")'))\n",
    "print('%d bidirectional promoters' % (n_bid))\n",
    "print('%d forward-only promoters' % (n_fwd,))\n",
    "print('%d reverse-only promoters' % (n_rev,))\n",
    "plt.figure(figsize=(16,6)).subplots_adjust(wspace=0.4)\n",
    "plt.subplot(1,2,1, aspect=1)\n",
    "plt.title('%d of %d sites (%.1f%%) annotated\\nas protein-coding promoters' % \\\n",
    "    (n_fwd + n_rev + n_bid, len(df_regl), 100 * (n_fwd + n_rev + n_bid) / float(len(df_regl)),))\n",
    "plt.pie([n_fwd, n_rev, n_bid], labels=[\n",
    "        '%(n_fwd)d forward\\npromoters' % locals(),\n",
    "        '%(n_rev)d reverse\\npromoters' % locals(), \n",
    "        '%(n_bid)d bidirectional\\npromoters' % locals(),\n",
    "])[-1][-1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "anaconda-cloud": {},
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
