{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2018-05-30T18:16:33.954504Z",
     "start_time": "2018-05-30T18:16:28.721743Z"
    },
    "code_folding": [
     0
    ]
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/mnt/home3/jj374/anaconda36/lib/python3.6/site-packages/statsmodels/compat/pandas.py:56: FutureWarning: The pandas.core.datetools module is deprecated and will be removed in a future version. Please use the pandas.tseries module instead.\n",
      "  from pandas.core import datetools\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "os.getcwd(): /mnt/beegfs/scratch_copy/ahringer/jj374/lab/relmapping\n"
     ]
    }
   ],
   "source": [
    "# Set annotation to annot_ce10\n",
    "%run ~/relmapping/annot/notebooks/annot__init__.ipynb\n",
    "annot_ = 'annot_ce10'\n",
    "def mp(fp, annot_=annot_): return os.path.join(annot_, 'metrics_type', fp)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2018-05-30T18:16:39.534932Z",
     "start_time": "2018-05-30T18:16:33.957486Z"
    },
    "code_folding": [
     0
    ],
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Load pre-computed ATAC-seq, exon, long/short cap metrics\n",
    "df_atac = pd.read_csv(os.path.join(annot_, 'accessible_sites.tsv'), sep='\\t')\n",
    "l_atac_peak_pos = df_atac[['start', 'end']].mean(axis=1).map(int)\n",
    "\n",
    "df_lcap_fwd = pd.read_csv(os.path.join(annot_, 'metrics_lcap', 'lcap_all_fwd.tsv'), sep='\\t', low_memory=False)\n",
    "df_lcap_rev = pd.read_csv(os.path.join(annot_, 'metrics_lcap', 'lcap_all_rev.tsv'), sep='\\t', low_memory=False)\n",
    "\n",
    "df_exon_fwd = pd.read_csv(os.path.join(annot_, 'metrics_exon', 'closest_exon_fwd.tsv'), sep='\\t', low_memory=False)\n",
    "df_exon_rev = pd.read_csv(os.path.join(annot_, 'metrics_exon', 'closest_exon_rev.tsv'), sep='\\t', low_memory=False)\n",
    "\n",
    "df_maxgap_fwd = pd.read_csv(os.path.join(annot_, 'metrics_maxgap', 'maxgap_fwd.tsv'), sep='\\t')\n",
    "df_maxgap_rev = pd.read_csv(os.path.join(annot_, 'metrics_maxgap', 'maxgap_rev.tsv'), sep='\\t')\n",
    "\n",
    "df_scap_fwd = pd.read_csv(os.path.join(annot_, 'metrics_scap', 'scap_fwd.tsv'), sep='\\t')\n",
    "df_scap_rev = pd.read_csv(os.path.join(annot_, 'metrics_scap', 'scap_rev.tsv'), sep='\\t')\n",
    "\n",
    "df_prom_fwd = df_atac[['chrom', 'start', 'end']].copy()\n",
    "df_prom_rev = df_atac[['chrom', 'start', 'end']].copy()\n",
    "df_prom_fwd['strand'] = '+'\n",
    "df_prom_rev['strand'] = '-'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2018-05-30T18:16:40.513769Z",
     "start_time": "2018-05-30T18:16:39.538415Z"
    },
    "code_folding": [
     0
    ]
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "11073 10928 fwd/rev genes\n",
      "10829 10668 fwd/rev genes longer than 250bp kept\n"
     ]
    }
   ],
   "source": [
    "# df_pass1_gene_%strand: WS260_ce10 gene coordinates, excluding 250bp downstream of TSS\n",
    "df_ = pd.concat([\n",
    "    yp.read_wbgtf('WS260_ce10/WS260_ce10.genes.pseudogene.gtf.gz', parse_attr=True, coords_adj=True),\n",
    "    yp.read_wbgtf('WS260_ce10/WS260_ce10.genes.protein_coding.gtf.gz', parse_attr=True, coords_adj=True),\n",
    "], axis=0)\n",
    "\n",
    "df_pass1_gene_fwd = df_.query('strand == \"+\"').reset_index(drop=True)[['chrom', 'start', 'end', 'gene_id']]\n",
    "df_pass1_gene_rev = df_.query('strand == \"-\"').reset_index(drop=True)[['chrom', 'start', 'end', 'gene_id']]\n",
    "\n",
    "print(len(df_pass1_gene_fwd), len(df_pass1_gene_rev), 'fwd/rev genes')\n",
    "\n",
    "# do not mask 250bp downstream of TSS\n",
    "df_pass1_gene_fwd['start'] = df_pass1_gene_fwd['start'] + 250\n",
    "df_pass1_gene_rev['end'] = df_pass1_gene_rev['end'] - 250\n",
    "\n",
    "df_pass1_gene_fwd = df_pass1_gene_fwd.query('start < end')\n",
    "df_pass1_gene_rev = df_pass1_gene_rev.query('start < end')\n",
    "print(len(df_pass1_gene_fwd), len(df_pass1_gene_rev), 'fwd/rev genes longer than 250bp kept')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2018-05-30T18:16:41.437664Z",
     "start_time": "2018-05-30T18:16:40.515625Z"
    },
    "code_folding": [
     0
    ]
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "42265 42309 fwd/rev gene body hits\n"
     ]
    }
   ],
   "source": [
    "# df_pass1_gpos_%strand[gene_chrom/start/end]: closest gene; if overlaps multiple genes, choose one with the more distal TSS\n",
    "def df_pos():\n",
    "    l_atac_peak_pos = df_atac[['start', 'end']].mean(axis=1).map(int)\n",
    "    df_ = pd.concat([df_atac['chrom'], l_atac_peak_pos, l_atac_peak_pos + 1], axis=1).copy()\n",
    "    df_.columns = ['chrom', 'start', 'end']\n",
    "    return df_\n",
    "\n",
    "def df_closest_(df_a, df_b, b_prefix, **kwargs):\n",
    "    df_a_sort = df_a\n",
    "    df_b_sort = df_b.sort_values(list(df_b.columns[:3]))\n",
    "    fn_ = BedTool.from_dataframe(df_a).closest(BedTool.from_dataframe(df_b_sort).fn, **kwargs).fn\n",
    "    names_ = list(itertools.chain(df_a.columns.values,\n",
    "        ['%s_%s' % (b_prefix, col) for col in df_b.columns.values],\n",
    "    ))\n",
    "    df_ = pd.read_csv(fn_, sep='\\t', names=names_)\n",
    "    return df_#[names_[len(df_a.columns):]]\n",
    "\n",
    "df_pass1_gpos_fwd_ = df_closest_(df_pos(), df_pass1_gene_fwd, 'pass1_gpos', t='all')\n",
    "df_pass1_gpos_rev_ = df_closest_(df_pos(), df_pass1_gene_rev, 'pass1_gpos', t='all')\n",
    "print(len(df_pass1_gpos_fwd_), len(df_pass1_gpos_rev_), 'fwd/rev gene body hits')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2018-05-30T18:16:50.655711Z",
     "start_time": "2018-05-30T18:16:41.442204Z"
    },
    "code_folding": [
     0
    ]
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "42245 42245\n",
      "33548\n",
      "33808\n"
     ]
    }
   ],
   "source": [
    "# df_pass1_gpos_%strand[is_nearly_intergenic]: site is outside of gene bodies, or up to 250bp from the TSS\n",
    "df_pass1_gpos_fwd_['is_nearly_intergenic'] = ~((df_pass1_gpos_fwd_['pass1_gpos_start'] <= df_pass1_gpos_fwd_['start'])\\\n",
    "                                           & (df_pass1_gpos_fwd_['end'] <= df_pass1_gpos_fwd_['pass1_gpos_end']))\n",
    "df_pass1_gpos_rev_['is_nearly_intergenic'] = ~((df_pass1_gpos_rev_['pass1_gpos_start'] <= df_pass1_gpos_rev_['start'])\\\n",
    "                                           & (df_pass1_gpos_rev_['end'] <= df_pass1_gpos_rev_['pass1_gpos_end']))\n",
    "\n",
    "l_ = 'is_nearly_intergenic'\n",
    "d_ = {'is_nearly_intergenic': np.any}\n",
    "df_exon_fwd[l_] = df_pass1_gpos_fwd_.groupby(['chrom', 'start', 'end']).agg(d_).reset_index()[l_]\n",
    "df_exon_rev[l_] = df_pass1_gpos_rev_.groupby(['chrom', 'start', 'end']).agg(d_).reset_index()[l_]\n",
    "print(len(df_exon_fwd), len(df_exon_rev))\n",
    "print(len(df_exon_fwd.query('is_nearly_intergenic')))\n",
    "print(len(df_exon_rev.query('is_nearly_intergenic')))\n",
    "\n",
    "# Visualise maxgap as .bed-files\n",
    "def itemRgb_(is_nearly_intergenic_):\n",
    "    if is_nearly_intergenic_:\n",
    "        return yp.ORANGE\n",
    "    else:\n",
    "        return yp.BLUE\n",
    "\n",
    "write_gffbed(mp('is_nearly_intergenic_fwd.bed'),\n",
    "    chrom = df_atac['chrom'],\n",
    "    start = df_atac['start'],\n",
    "    end = df_atac['end'],\n",
    "    #name = df_maxgap_fwd.min(axis=1),\n",
    "    strand = '+',\n",
    "    #attr = df_pass1_gpos_fwd_,\n",
    "    itemRgb = map(itemRgb_, df_exon_fwd['is_nearly_intergenic']),\n",
    ")\n",
    "\n",
    "write_gffbed(mp('is_nearly_intergenic_rev.bed'),\n",
    "    chrom = df_atac['chrom'],\n",
    "    start = df_atac['start'],\n",
    "    end = df_atac['end'],\n",
    "    #name = df_maxgap_rev.min(axis=1),\n",
    "    strand = '-',\n",
    "    #attr = df_maxgap_rev,\n",
    "    itemRgb = map(itemRgb_, df_exon_rev['is_nearly_intergenic']),\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2018-05-30T18:16:54.790915Z",
     "start_time": "2018-05-30T18:16:50.660434Z"
    },
    "code_folding": [
     0
    ]
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "wt_emb fwd\n",
      "wt_emb rev\n",
      "wt_l1 fwd\n",
      "wt_l1 rev\n",
      "wt_l2 fwd\n",
      "wt_l2 rev\n",
      "wt_l3 fwd\n",
      "wt_l3 rev\n",
      "wt_l4 fwd\n",
      "wt_l4 rev\n",
      "wt_ya fwd\n",
      "wt_ya rev\n",
      "glp1_d1 fwd\n",
      "glp1_d1 rev\n",
      "glp1_d2 fwd\n",
      "glp1_d2 rev\n",
      "glp1_d6 fwd\n",
      "glp1_d6 rev\n",
      "glp1_d9 fwd\n",
      "glp1_d9 rev\n",
      "glp1_d13 fwd\n",
      "glp1_d13 rev\n"
     ]
    }
   ],
   "source": [
    "# df_prom_%strand[annot_detailed_%stage]: stage-specific detailed annotation, excluding lowconf\n",
    "d_detailed_annot = collections.OrderedDict([\n",
    "    ('coding_promoter_jump_scap', 'coding_promoter'),\n",
    "    ('coding_promoter_incr_scap', 'coding_promoter'),\n",
    "    ('coding_promoter_jump_incr', 'coding_promoter'),\n",
    "    ('coding_promoter_lowconf', 'coding_promoter'),\n",
    "    ('pseudogene_promoter_jump_scap', 'pseudogene_promoter'),\n",
    "    ('pseudogene_promoter_incr_scap', 'pseudogene_promoter'),\n",
    "    ('pseudogene_promoter_jump_incr', 'pseudogene_promoter'),\n",
    "    ('pseudogene_promoter_lowconf', 'pseudogene_promoter'),\n",
    "    ('non-coding_RNA', 'non-coding_RNA'),\n",
    "    ('unknown_promoter_jump_scap', 'unknown_promoter'),\n",
    "    #('unknown_promoter_jump_incr', 'unknown_promoter'),\n",
    "    ('transcription_initiation', 'transcription_initiation'),\n",
    "    ('no_transcription', 'no_transcription'),\n",
    "])\n",
    "\n",
    "d_annot_strand_legend = collections.OrderedDict([\n",
    "    ('coding_promoter', yp.RED),\n",
    "    ('pseudogene_promoter', yp.ORANGE),\n",
    "    ('non-coding_RNA', yp.BLACK),\n",
    "    ('unknown_promoter', yp.YELLOW),\n",
    "    ('transcription_initiation', yp.GREEN),\n",
    "    ('no_transcription', yp.BLUE),\n",
    "])\n",
    "\n",
    "d_annot_legend = collections.OrderedDict([\n",
    "    ('coding_promoter', yp.RED),\n",
    "    ('pseudogene_promoter', yp.ORANGE),\n",
    "    ('non-coding_RNA', yp.BLACK),\n",
    "    ('unknown_promoter', yp.YELLOW),\n",
    "    ('putative_enhancer', yp.GREEN),\n",
    "    ('other_element', yp.BLUE),\n",
    "])\n",
    "\n",
    "def annot_detailed_(chrom, start, end, jump, incr, maxgap, exon1_dist, exon2_dist, pass2_exon1_dist, \n",
    "                    scap_pass, scap_diff, gene_biotype, is_exon1_first, is_nearly_intergenic):\n",
    "    \"\"\"\n",
    "    #is_gene_biotype = (gene_biotype == 'protein_coding') | (gene_biotype == 'pseudogene')\n",
    "    #v = False\n",
    "    #if chrom == \"chrI\" and start == 183421 and end == 183568: v = True\n",
    "    #if (chrom == \"chrX\") and (start == 1125552) and (end == 1125703):\n",
    "    #    v = True\n",
    "    #    print(jump, maxgap <= 0, scap_diff >= 0, has_scap)\n",
    "    #if start == 11829601-1: print(incr, jump, has_scap, n_txn_evidence)\n",
    "    \"\"\"\n",
    "\n",
    "    near_exon1 = abs(exon1_dist) < 250\n",
    "    near_exon2 = abs(exon2_dist) < 250\n",
    "    near_pass2_exon1 = abs(pass2_exon1_dist) < 250\n",
    "    #nearly_upstream_of_tss = exon1_dist > -250\n",
    "\n",
    "    # We annotated accessible sites as coding_promoter or pseudogene_promoter if they fulfilled the following\n",
    "    # four criteria (p1 to p4).\n",
    "    # 1) The accessible site had transcription initiation, and passed at least one of the elongation tests \n",
    "    #    (jump or incr), or passed both elongation tests (jump and incr).\n",
    "    p1 = (int(scap_pass) + int(jump) + int(incr)) >= 2\n",
    "    # 2) Transcription initiation mode at the accessible site was either upstream of the closest first exon, \n",
    "    #    or, in the presence of a UTR, up to 250bp downstream within the UTR.\n",
    "    #p2 = nearly_upstream_of_tss and (scap_diff >= 0)\n",
    "    p2 = (scap_diff >= 0)\n",
    "    # 3) The region from peak accessibility to the closest first exon did not contain the 5' end of a non-first exon.\n",
    "    p3 = is_exon1_first\n",
    "    # 4) Distal sites (peak accessibility >250bp from the closest first exon) were additionally required to \n",
    "    # (a) have continuous long cap coverage from 250bp downstream of peak accessibility to the closest first exon, and \n",
    "    # (b) be further than 250bp away from any non-first exon.\n",
    "    p4 = near_exon1 or ((maxgap <= 0) and not near_exon2)\n",
    "    if p1 and p2 and p3 and p4:\n",
    "        #if v: print('coding_promoter')\n",
    "        if scap_pass and jump:\n",
    "            return 'coding_promoter_jump_scap' if (gene_biotype == 'protein_coding') else 'pseudogene_promoter_jump_scap'\n",
    "        elif scap_pass and incr:\n",
    "            return 'coding_promoter_incr_scap' if (gene_biotype == 'protein_coding') else 'pseudogene_promoter_incr_scap'\n",
    "        elif jump and incr:\n",
    "            return 'coding_promoter_jump_incr' if (gene_biotype == 'protein_coding') else 'pseudogene_promoter_jump_incr'\n",
    "        else:\n",
    "            assert False\n",
    "\n",
    "    # Sites within 250bp of the 5' end of an annotated tRNA, snoRNA, miRNA, snRNA or rRNA were annotated as \n",
    "    # non-coding_RNA.\n",
    "    elif near_pass2_exon1:\n",
    "        #if v: print('non-coding_RNA')\n",
    "        return 'non-coding_RNA'\n",
    "\n",
    "    # Next, intergenic sites more than 250bp away from annotated exons that had initiating transcription, \n",
    "    # and passed the jump test were annotated as unknown_promoter.\n",
    "    elif is_nearly_intergenic and not(near_exon1 or near_exon2) and scap_pass and jump:\n",
    "        return 'unknown_promoter_jump_scap'\n",
    "\n",
    "    # All remaining sites were annotated as transcription_initiation or no_transcription based on \n",
    "    # whether they had transcription initiation.\n",
    "    elif scap_pass:\n",
    "        #if v: print('transcription_initiation')\n",
    "        return 'transcription_initiation'\n",
    "    else:\n",
    "        #if v: print('no_transcription')\n",
    "        return 'no_transcription'\n",
    "\n",
    "for stage in itertools.islice(config['stages'], None):\n",
    "    print(stage, 'fwd')\n",
    "    df_prom_fwd['annot_detailed_%s' % (stage,)] = [*map(annot_detailed_,\n",
    "        # chrom, start, end\n",
    "        df_atac['chrom'], df_atac['start'], df_atac['end'],\n",
    "        # jump, incr\n",
    "        df_lcap_fwd['lcap_%s_fwd_passed_jump' % (stage,)], df_lcap_fwd['lcap_%s_fwd_passed_incr' % (stage,)],\n",
    "        # maxgap\n",
    "        df_maxgap_fwd['maxgap_%s_fwd' % (stage,)],\n",
    "        # exon1_dist\n",
    "        df_exon_fwd['pass1_exon1_start'] - l_atac_peak_pos,\n",
    "        # exon2_dist\n",
    "        df_exon_fwd['pass1_exon2_start'] - l_atac_peak_pos,\n",
    "        # pass2_exon2_dist\n",
    "        df_exon_fwd['pass2_exon1_start'] - l_atac_peak_pos,\n",
    "        # scap_pass\n",
    "        df_scap_fwd['scap_pass'],\n",
    "        # scap_diff\n",
    "        pd.to_numeric(df_exon_fwd['pass1_exon1_init_cutoff_pos'], errors='coerce') - df_scap_fwd['scap_mode'],\n",
    "        # gene_biotype\n",
    "        df_exon_fwd['pass1_exon1_gene_biotype'],\n",
    "        # is_exon1_first\n",
    "        df_exon_fwd['pass1_exon1_distance'] <= df_exon_fwd['pass1_exon2_downstream_distance'],\n",
    "        # is_nearly_intergenic\n",
    "        df_exon_fwd['is_nearly_intergenic'],\n",
    "    )]\n",
    "    print(stage, 'rev')\n",
    "    df_prom_rev['annot_detailed_%s' % (stage,)] = [*map(annot_detailed_,\n",
    "        # chrom, start, end\n",
    "        df_atac['chrom'], df_atac['start'], df_atac['end'],\n",
    "        # jump, incr\n",
    "        df_lcap_rev['lcap_%s_rev_passed_jump' % (stage,)], df_lcap_rev['lcap_%s_rev_passed_incr' % (stage,)],\n",
    "        # maxgap\n",
    "        df_maxgap_rev['maxgap_%s_rev' % (stage,)],\n",
    "        # exon1_dist\n",
    "        l_atac_peak_pos - (df_exon_rev['pass1_exon1_end'] - 1),\n",
    "        # exon2_dist\n",
    "        l_atac_peak_pos - (df_exon_rev['pass1_exon2_end'] - 1),\n",
    "        # pass2_exon2_dist\n",
    "        l_atac_peak_pos - (df_exon_rev['pass2_exon1_end'] - 1),\n",
    "        # scap_pass\n",
    "        df_scap_rev['scap_pass'],\n",
    "        # scap_diff\n",
    "        df_scap_rev['scap_mode'] - pd.to_numeric(df_exon_rev['pass1_exon1_init_cutoff_pos'], errors='coerce'),\n",
    "        # gene_biotype\n",
    "        df_exon_rev['pass1_exon1_gene_biotype'],\n",
    "        # is_exon1_first\n",
    "        df_exon_rev['pass1_exon2_downstream_distance'] <= df_exon_rev['pass1_exon1_distance'],\n",
    "        # is_nearly_intergenic\n",
    "        df_exon_rev['is_nearly_intergenic'],\n",
    "    )]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2018-05-30T18:17:00.999464Z",
     "start_time": "2018-05-30T18:16:54.793169Z"
    },
    "code_folding": [
     0
    ],
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# df_prom_%strand[annot_detailed_all]: stage-aggregated detailed annotation, excluding lowconf\n",
    "def annot_detailed_summary_(l_annot):\n",
    "    l_ranking = list(d_detailed_annot.keys())\n",
    "    l_annot_i = [l_ranking.index(annot) for annot in l_annot]\n",
    "    return l_ranking[min(l_annot_i)]\n",
    "\n",
    "df_prom_fwd['annot_detailed_summary'] = list(map(lambda ir: annot_detailed_summary_(ir[1].tolist()),\n",
    "    df_prom_fwd[['annot_detailed_%(stage)s' % locals() for stage in config['stages']]].iterrows()))\n",
    "\n",
    "df_prom_rev['annot_detailed_summary'] = list(map(lambda ir: annot_detailed_summary_(ir[1].tolist()),\n",
    "    df_prom_rev[['annot_detailed_%(stage)s' % locals() for stage in config['stages']]].iterrows()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2018-05-30T18:17:01.236831Z",
     "start_time": "2018-05-30T18:17:01.001733Z"
    },
    "code_folding": [
     0
    ]
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>wt_emb</th>\n",
       "      <th>wt_l1</th>\n",
       "      <th>wt_l2</th>\n",
       "      <th>wt_l3</th>\n",
       "      <th>wt_l4</th>\n",
       "      <th>wt_ya</th>\n",
       "      <th>glp1_d1</th>\n",
       "      <th>glp1_d2</th>\n",
       "      <th>glp1_d6</th>\n",
       "      <th>glp1_d9</th>\n",
       "      <th>glp1_d13</th>\n",
       "      <th>summary</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>coding_promoter_jump_scap</th>\n",
       "      <td>3231</td>\n",
       "      <td>3617</td>\n",
       "      <td>4245</td>\n",
       "      <td>3508</td>\n",
       "      <td>3604</td>\n",
       "      <td>3168</td>\n",
       "      <td>2475</td>\n",
       "      <td>2114</td>\n",
       "      <td>2463</td>\n",
       "      <td>2418</td>\n",
       "      <td>2449</td>\n",
       "      <td>6055</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>coding_promoter_incr_scap</th>\n",
       "      <td>390</td>\n",
       "      <td>281</td>\n",
       "      <td>293</td>\n",
       "      <td>439</td>\n",
       "      <td>336</td>\n",
       "      <td>452</td>\n",
       "      <td>404</td>\n",
       "      <td>318</td>\n",
       "      <td>486</td>\n",
       "      <td>307</td>\n",
       "      <td>521</td>\n",
       "      <td>788</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>coding_promoter_jump_incr</th>\n",
       "      <td>52</td>\n",
       "      <td>71</td>\n",
       "      <td>89</td>\n",
       "      <td>71</td>\n",
       "      <td>97</td>\n",
       "      <td>99</td>\n",
       "      <td>71</td>\n",
       "      <td>95</td>\n",
       "      <td>69</td>\n",
       "      <td>90</td>\n",
       "      <td>76</td>\n",
       "      <td>357</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>coding_promoter_lowconf</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pseudogene_promoter_jump_scap</th>\n",
       "      <td>49</td>\n",
       "      <td>43</td>\n",
       "      <td>43</td>\n",
       "      <td>32</td>\n",
       "      <td>41</td>\n",
       "      <td>49</td>\n",
       "      <td>24</td>\n",
       "      <td>13</td>\n",
       "      <td>25</td>\n",
       "      <td>19</td>\n",
       "      <td>21</td>\n",
       "      <td>90</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pseudogene_promoter_incr_scap</th>\n",
       "      <td>11</td>\n",
       "      <td>8</td>\n",
       "      <td>10</td>\n",
       "      <td>13</td>\n",
       "      <td>9</td>\n",
       "      <td>3</td>\n",
       "      <td>9</td>\n",
       "      <td>4</td>\n",
       "      <td>9</td>\n",
       "      <td>10</td>\n",
       "      <td>11</td>\n",
       "      <td>20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pseudogene_promoter_jump_incr</th>\n",
       "      <td>8</td>\n",
       "      <td>8</td>\n",
       "      <td>11</td>\n",
       "      <td>7</td>\n",
       "      <td>11</td>\n",
       "      <td>8</td>\n",
       "      <td>6</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pseudogene_promoter_lowconf</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>non-coding_RNA</th>\n",
       "      <td>495</td>\n",
       "      <td>489</td>\n",
       "      <td>485</td>\n",
       "      <td>488</td>\n",
       "      <td>494</td>\n",
       "      <td>491</td>\n",
       "      <td>498</td>\n",
       "      <td>498</td>\n",
       "      <td>490</td>\n",
       "      <td>493</td>\n",
       "      <td>486</td>\n",
       "      <td>463</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>unknown_promoter_jump_scap</th>\n",
       "      <td>468</td>\n",
       "      <td>600</td>\n",
       "      <td>798</td>\n",
       "      <td>406</td>\n",
       "      <td>328</td>\n",
       "      <td>263</td>\n",
       "      <td>189</td>\n",
       "      <td>150</td>\n",
       "      <td>214</td>\n",
       "      <td>285</td>\n",
       "      <td>362</td>\n",
       "      <td>1521</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>transcription_initiation</th>\n",
       "      <td>24456</td>\n",
       "      <td>24062</td>\n",
       "      <td>23226</td>\n",
       "      <td>24214</td>\n",
       "      <td>24288</td>\n",
       "      <td>24674</td>\n",
       "      <td>25501</td>\n",
       "      <td>26003</td>\n",
       "      <td>25412</td>\n",
       "      <td>25568</td>\n",
       "      <td>25250</td>\n",
       "      <td>20162</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>no_transcription</th>\n",
       "      <td>13085</td>\n",
       "      <td>13066</td>\n",
       "      <td>13045</td>\n",
       "      <td>13067</td>\n",
       "      <td>13037</td>\n",
       "      <td>13038</td>\n",
       "      <td>13068</td>\n",
       "      <td>13049</td>\n",
       "      <td>13072</td>\n",
       "      <td>13051</td>\n",
       "      <td>13067</td>\n",
       "      <td>12766</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                               wt_emb  wt_l1  wt_l2  wt_l3  wt_l4  wt_ya  \\\n",
       "coding_promoter_jump_scap        3231   3617   4245   3508   3604   3168   \n",
       "coding_promoter_incr_scap         390    281    293    439    336    452   \n",
       "coding_promoter_jump_incr          52     71     89     71     97     99   \n",
       "coding_promoter_lowconf             0      0      0      0      0      0   \n",
       "pseudogene_promoter_jump_scap      49     43     43     32     41     49   \n",
       "pseudogene_promoter_incr_scap      11      8     10     13      9      3   \n",
       "pseudogene_promoter_jump_incr       8      8     11      7     11      8   \n",
       "pseudogene_promoter_lowconf         0      0      0      0      0      0   \n",
       "non-coding_RNA                    495    489    485    488    494    491   \n",
       "unknown_promoter_jump_scap        468    600    798    406    328    263   \n",
       "transcription_initiation        24456  24062  23226  24214  24288  24674   \n",
       "no_transcription                13085  13066  13045  13067  13037  13038   \n",
       "\n",
       "                               glp1_d1  glp1_d2  glp1_d6  glp1_d9  glp1_d13  \\\n",
       "coding_promoter_jump_scap         2475     2114     2463     2418      2449   \n",
       "coding_promoter_incr_scap          404      318      486      307       521   \n",
       "coding_promoter_jump_incr           71       95       69       90        76   \n",
       "coding_promoter_lowconf              0        0        0        0         0   \n",
       "pseudogene_promoter_jump_scap       24       13       25       19        21   \n",
       "pseudogene_promoter_incr_scap        9        4        9       10        11   \n",
       "pseudogene_promoter_jump_incr        6        1        5        4         2   \n",
       "pseudogene_promoter_lowconf          0        0        0        0         0   \n",
       "non-coding_RNA                     498      498      490      493       486   \n",
       "unknown_promoter_jump_scap         189      150      214      285       362   \n",
       "transcription_initiation         25501    26003    25412    25568     25250   \n",
       "no_transcription                 13068    13049    13072    13051     13067   \n",
       "\n",
       "                               summary  \n",
       "coding_promoter_jump_scap         6055  \n",
       "coding_promoter_incr_scap          788  \n",
       "coding_promoter_jump_incr          357  \n",
       "coding_promoter_lowconf              0  \n",
       "pseudogene_promoter_jump_scap       90  \n",
       "pseudogene_promoter_incr_scap       20  \n",
       "pseudogene_promoter_jump_incr       23  \n",
       "pseudogene_promoter_lowconf          0  \n",
       "non-coding_RNA                     463  \n",
       "unknown_promoter_jump_scap        1521  \n",
       "transcription_initiation         20162  \n",
       "no_transcription                 12766  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Forward strand annotation summary table (excl. lowconf)\n",
    "df_fwd_ = pd.concat([df_prom_fwd['annot_detailed_%(stage)s' % locals()].value_counts() for stage in config['stages'] + ['summary']], axis=1).loc[d_detailed_annot.keys()]\n",
    "df_fwd_.columns = config['stages'] + ['summary']\n",
    "#df_fwd_.to_latex('%(dm_step)s_fig/prom_summary_fwd.latex' % locals())\n",
    "df_fwd_.fillna(0, downcast='infer')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2018-05-30T18:17:01.525043Z",
     "start_time": "2018-05-30T18:17:01.239207Z"
    },
    "code_folding": [
     0
    ]
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>wt_emb</th>\n",
       "      <th>wt_l1</th>\n",
       "      <th>wt_l2</th>\n",
       "      <th>wt_l3</th>\n",
       "      <th>wt_l4</th>\n",
       "      <th>wt_ya</th>\n",
       "      <th>glp1_d1</th>\n",
       "      <th>glp1_d2</th>\n",
       "      <th>glp1_d6</th>\n",
       "      <th>glp1_d9</th>\n",
       "      <th>glp1_d13</th>\n",
       "      <th>summary</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>coding_promoter_jump_scap</th>\n",
       "      <td>3297</td>\n",
       "      <td>3618</td>\n",
       "      <td>4246</td>\n",
       "      <td>3720</td>\n",
       "      <td>3597</td>\n",
       "      <td>3194</td>\n",
       "      <td>2471</td>\n",
       "      <td>2156</td>\n",
       "      <td>2459</td>\n",
       "      <td>2370</td>\n",
       "      <td>2559</td>\n",
       "      <td>6040</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>coding_promoter_incr_scap</th>\n",
       "      <td>340</td>\n",
       "      <td>288</td>\n",
       "      <td>295</td>\n",
       "      <td>356</td>\n",
       "      <td>352</td>\n",
       "      <td>434</td>\n",
       "      <td>382</td>\n",
       "      <td>343</td>\n",
       "      <td>474</td>\n",
       "      <td>366</td>\n",
       "      <td>460</td>\n",
       "      <td>703</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>coding_promoter_jump_incr</th>\n",
       "      <td>69</td>\n",
       "      <td>73</td>\n",
       "      <td>83</td>\n",
       "      <td>93</td>\n",
       "      <td>110</td>\n",
       "      <td>106</td>\n",
       "      <td>79</td>\n",
       "      <td>88</td>\n",
       "      <td>76</td>\n",
       "      <td>88</td>\n",
       "      <td>93</td>\n",
       "      <td>393</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>coding_promoter_lowconf</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pseudogene_promoter_jump_scap</th>\n",
       "      <td>55</td>\n",
       "      <td>54</td>\n",
       "      <td>62</td>\n",
       "      <td>52</td>\n",
       "      <td>48</td>\n",
       "      <td>46</td>\n",
       "      <td>22</td>\n",
       "      <td>11</td>\n",
       "      <td>27</td>\n",
       "      <td>23</td>\n",
       "      <td>26</td>\n",
       "      <td>107</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pseudogene_promoter_incr_scap</th>\n",
       "      <td>13</td>\n",
       "      <td>11</td>\n",
       "      <td>6</td>\n",
       "      <td>11</td>\n",
       "      <td>10</td>\n",
       "      <td>15</td>\n",
       "      <td>6</td>\n",
       "      <td>9</td>\n",
       "      <td>10</td>\n",
       "      <td>6</td>\n",
       "      <td>7</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pseudogene_promoter_jump_incr</th>\n",
       "      <td>7</td>\n",
       "      <td>8</td>\n",
       "      <td>11</td>\n",
       "      <td>8</td>\n",
       "      <td>11</td>\n",
       "      <td>11</td>\n",
       "      <td>5</td>\n",
       "      <td>3</td>\n",
       "      <td>5</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pseudogene_promoter_lowconf</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>non-coding_RNA</th>\n",
       "      <td>496</td>\n",
       "      <td>494</td>\n",
       "      <td>492</td>\n",
       "      <td>496</td>\n",
       "      <td>498</td>\n",
       "      <td>493</td>\n",
       "      <td>502</td>\n",
       "      <td>507</td>\n",
       "      <td>502</td>\n",
       "      <td>499</td>\n",
       "      <td>494</td>\n",
       "      <td>479</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>unknown_promoter_jump_scap</th>\n",
       "      <td>443</td>\n",
       "      <td>617</td>\n",
       "      <td>863</td>\n",
       "      <td>493</td>\n",
       "      <td>366</td>\n",
       "      <td>245</td>\n",
       "      <td>179</td>\n",
       "      <td>172</td>\n",
       "      <td>208</td>\n",
       "      <td>270</td>\n",
       "      <td>379</td>\n",
       "      <td>1613</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>transcription_initiation</th>\n",
       "      <td>24385</td>\n",
       "      <td>23947</td>\n",
       "      <td>23066</td>\n",
       "      <td>23901</td>\n",
       "      <td>24159</td>\n",
       "      <td>24602</td>\n",
       "      <td>25468</td>\n",
       "      <td>25832</td>\n",
       "      <td>25350</td>\n",
       "      <td>25495</td>\n",
       "      <td>25104</td>\n",
       "      <td>20068</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>no_transcription</th>\n",
       "      <td>13140</td>\n",
       "      <td>13135</td>\n",
       "      <td>13121</td>\n",
       "      <td>13115</td>\n",
       "      <td>13094</td>\n",
       "      <td>13099</td>\n",
       "      <td>13131</td>\n",
       "      <td>13124</td>\n",
       "      <td>13134</td>\n",
       "      <td>13122</td>\n",
       "      <td>13117</td>\n",
       "      <td>12794</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                               wt_emb  wt_l1  wt_l2  wt_l3  wt_l4  wt_ya  \\\n",
       "coding_promoter_jump_scap        3297   3618   4246   3720   3597   3194   \n",
       "coding_promoter_incr_scap         340    288    295    356    352    434   \n",
       "coding_promoter_jump_incr          69     73     83     93    110    106   \n",
       "coding_promoter_lowconf             0      0      0      0      0      0   \n",
       "pseudogene_promoter_jump_scap      55     54     62     52     48     46   \n",
       "pseudogene_promoter_incr_scap      13     11      6     11     10     15   \n",
       "pseudogene_promoter_jump_incr       7      8     11      8     11     11   \n",
       "pseudogene_promoter_lowconf         0      0      0      0      0      0   \n",
       "non-coding_RNA                    496    494    492    496    498    493   \n",
       "unknown_promoter_jump_scap        443    617    863    493    366    245   \n",
       "transcription_initiation        24385  23947  23066  23901  24159  24602   \n",
       "no_transcription                13140  13135  13121  13115  13094  13099   \n",
       "\n",
       "                               glp1_d1  glp1_d2  glp1_d6  glp1_d9  glp1_d13  \\\n",
       "coding_promoter_jump_scap         2471     2156     2459     2370      2559   \n",
       "coding_promoter_incr_scap          382      343      474      366       460   \n",
       "coding_promoter_jump_incr           79       88       76       88        93   \n",
       "coding_promoter_lowconf              0        0        0        0         0   \n",
       "pseudogene_promoter_jump_scap       22       11       27       23        26   \n",
       "pseudogene_promoter_incr_scap        6        9       10        6         7   \n",
       "pseudogene_promoter_jump_incr        5        3        5        6         6   \n",
       "pseudogene_promoter_lowconf          0        0        0        0         0   \n",
       "non-coding_RNA                     502      507      502      499       494   \n",
       "unknown_promoter_jump_scap         179      172      208      270       379   \n",
       "transcription_initiation         25468    25832    25350    25495     25104   \n",
       "no_transcription                 13131    13124    13134    13122     13117   \n",
       "\n",
       "                               summary  \n",
       "coding_promoter_jump_scap         6040  \n",
       "coding_promoter_incr_scap          703  \n",
       "coding_promoter_jump_incr          393  \n",
       "coding_promoter_lowconf              0  \n",
       "pseudogene_promoter_jump_scap      107  \n",
       "pseudogene_promoter_incr_scap       19  \n",
       "pseudogene_promoter_jump_incr       29  \n",
       "pseudogene_promoter_lowconf          0  \n",
       "non-coding_RNA                     479  \n",
       "unknown_promoter_jump_scap        1613  \n",
       "transcription_initiation         20068  \n",
       "no_transcription                 12794  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Reverse strand annotation summary table (excl. lowconf)\n",
    "df_rev_ = pd.concat([df_prom_rev['annot_detailed_%(stage)s' % locals()].value_counts() for stage in config['stages'] + ['summary']], axis=1).loc[d_detailed_annot.keys()]\n",
    "df_rev_.columns = config['stages'] + ['summary']\n",
    "#df_rev_.to_latex('%(dm_step)s_fig/prom_summary_rev.latex' % locals())\n",
    "df_rev_.fillna(0, downcast='infer')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2018-05-30T18:17:03.988135Z",
     "start_time": "2018-05-30T18:17:01.527483Z"
    },
    "code_folding": [
     0
    ]
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "wt_emb fwd\n",
      "wt_emb rev\n",
      "wt_l1 fwd\n",
      "wt_l1 rev\n",
      "wt_l2 fwd\n",
      "wt_l2 rev\n",
      "wt_l3 fwd\n",
      "wt_l3 rev\n",
      "wt_l4 fwd\n",
      "wt_l4 rev\n",
      "wt_ya fwd\n",
      "wt_ya rev\n",
      "glp1_d1 fwd\n",
      "glp1_d1 rev\n",
      "glp1_d2 fwd\n",
      "glp1_d2 rev\n",
      "glp1_d6 fwd\n",
      "glp1_d6 rev\n",
      "glp1_d9 fwd\n",
      "glp1_d9 rev\n",
      "glp1_d13 fwd\n",
      "glp1_d13 rev\n"
     ]
    }
   ],
   "source": [
    "# df_lowconf_%strand[annot_lowconf_raw_%stage] <- candidate \"raw\" low-confidence annotation (with aggregating)\n",
    "def lowconf_raw_stage(chrom, start, end, jump, incr, log2FoldChange, maxgap, exon1_dist, exon2_dist,\n",
    "                      pass2_exon1_dist, scap_pass, scap_diff, gene_biotype, is_exon1_first, is_nearly_intergenic):\n",
    "    near_exon1 = abs(exon1_dist) < 250\n",
    "    near_exon2 = abs(exon2_dist) < 250\n",
    "    near_pass2_exon1 = abs(pass2_exon1_dist) < 250\n",
    "    nearly_upstream_of_tss = exon1_dist > -250\n",
    "    has_txn_evidence = (log2FoldChange > 1) #incr or (log2FoldChange > 1 and (has_scap))\n",
    "\n",
    "    v = False\n",
    "    #if chrom == \"chrI\" and start == 183421 and end == 183568: v = True\n",
    "    if (near_exon1 or not(near_exon2)) and nearly_upstream_of_tss and is_exon1_first and\\\n",
    "        is_nearly_intergenic and (not (maxgap > 0)) and has_txn_evidence and (scap_diff >= 0): #-50\n",
    "        #if pass1_exon1_locus_id == \"F59F4.3\":\n",
    "        #    print(chrom, start, end)\n",
    "        return 'coding_promoter_lowconf' if (gene_biotype == 'protein_coding') else 'pseudogene_promoter_lowconf'\n",
    "    else:\n",
    "        return 'no_lowconf_annotation'\n",
    "\n",
    "df_lowconf_fwd = df_atac[['chrom', 'start', 'end']].copy()\n",
    "df_lowconf_rev = df_atac[['chrom', 'start', 'end']].copy()\n",
    "df_lowconf_fwd['strand'] = '+'\n",
    "df_lowconf_rev['strand'] = '-'\n",
    "\n",
    "for stage in itertools.islice(config['stages'], None):\n",
    "    print(stage, 'fwd')\n",
    "    df_lowconf_fwd['annot_lowconf_raw_%s' % (stage,)] = [*map(lowconf_raw_stage,\n",
    "        #chrom, start, end\n",
    "        df_atac['chrom'], df_atac['start'], df_atac['end'],\n",
    "        # jump, incr\n",
    "        df_lcap_fwd['lcap_%s_fwd_passed_jump' % (stage,)], df_lcap_fwd['lcap_%s_fwd_passed_incr' % (stage,)],\n",
    "        # log2FoldChange\n",
    "        df_lcap_fwd['lcap_%s_fwd_log2FoldChange' % (stage,)],\n",
    "        # maxgap\n",
    "        df_maxgap_fwd['maxgap_%s_fwd' % (stage,)],\n",
    "        # exon1_dist\n",
    "        df_exon_fwd['pass1_exon1_start'] - l_atac_peak_pos,\n",
    "        # exon2_dist\n",
    "        df_exon_fwd['pass1_exon2_start'] - l_atac_peak_pos,\n",
    "        # pass2_exon1_dist\n",
    "        df_exon_fwd['pass2_exon1_start'] - l_atac_peak_pos,\n",
    "        # scap_pass\n",
    "        df_scap_fwd['scap_pass'],\n",
    "        # scap_diff\n",
    "        pd.to_numeric(df_exon_fwd['pass1_exon1_init_cutoff_pos'], errors='coerce') - df_scap_fwd['scap_mode'],\n",
    "        # gene_biotype\n",
    "        df_exon_fwd['pass1_exon1_gene_biotype'],\n",
    "        # is_exon1_first\n",
    "        df_exon_fwd['pass1_exon1_distance'] <= df_exon_fwd['pass1_exon2_downstream_distance'],\n",
    "        # is_nearly_intergenic\n",
    "        df_exon_fwd['is_nearly_intergenic'],\n",
    "    )]\n",
    "\n",
    "    print(stage, 'rev')\n",
    "    df_lowconf_rev['annot_lowconf_raw_%s' % (stage,)] = [*map(lowconf_raw_stage,\n",
    "        # chrom, start, end\n",
    "        df_atac['chrom'], df_atac['start'], df_atac['end'],\n",
    "        # jump, incr\n",
    "        df_lcap_rev['lcap_%s_rev_passed_jump' % (stage,)], df_lcap_rev['lcap_%s_rev_passed_incr' % (stage,)],\n",
    "        # log2FoldChange\n",
    "        df_lcap_rev['lcap_%s_rev_log2FoldChange' % (stage,)],\n",
    "        # maxgap\n",
    "        df_maxgap_rev['maxgap_%s_rev' % (stage,)],\n",
    "        # exon1_dist\n",
    "        l_atac_peak_pos - (df_exon_rev['pass1_exon1_end'] - 1),\n",
    "        # exon2_dist\n",
    "        l_atac_peak_pos - (df_exon_rev['pass1_exon2_end'] - 1),\n",
    "        # pass2_exon1_dist\n",
    "        l_atac_peak_pos - (df_exon_rev['pass2_exon1_end'] - 1),\n",
    "        # scap_pass\n",
    "        df_scap_fwd['scap_pass'],\n",
    "        # scap_diff\n",
    "        df_scap_rev['scap_mode'] - pd.to_numeric(df_exon_rev['pass1_exon1_init_cutoff_pos'], errors='coerce'),\n",
    "        # gene_biotype\n",
    "        df_exon_rev['pass1_exon1_gene_biotype'],\n",
    "        # is_exon1_first\n",
    "        df_exon_rev['pass1_exon2_downstream_distance'] <= df_exon_rev['pass1_exon1_distance'],\n",
    "        #is_nearly_intergenic\n",
    "        df_exon_rev['is_nearly_intergenic'],\n",
    "    )]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2018-05-30T18:17:09.928819Z",
     "start_time": "2018-05-30T18:17:03.990175Z"
    },
    "code_folding": [
     0
    ],
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# df_lowconf_%strand[annot_lowconf_raw_summary] <- candidate \"raw\" low-confidence annotation across stages\n",
    "d_annot_lowconf_raw_legend = collections.OrderedDict([\n",
    "    ('coding_promoter_lowconf', yp.RED),\n",
    "    ('pseudogene_promoter_lowconf', yp.ORANGE),\n",
    "    ('no_lowconf_annotation', yp.BLACK),\n",
    "])\n",
    "\n",
    "def annot_lowconf_raw_summary_(l_annot):\n",
    "    l_ranking = list(d_annot_lowconf_raw_legend.keys())\n",
    "    l_annot_i = [l_ranking.index(annot) for annot in l_annot]\n",
    "    return l_ranking[min(l_annot_i)]\n",
    "\n",
    "df_lowconf_fwd['annot_lowconf_raw_summary'] = list(map(lambda ir: annot_lowconf_raw_summary_(ir[1].tolist()),\n",
    "    df_lowconf_fwd[['annot_lowconf_raw_%(stage)s' % locals() for stage in config['stages']]].iterrows()))\n",
    "df_lowconf_rev['annot_lowconf_raw_summary'] = list(map(lambda ir: annot_lowconf_raw_summary_(ir[1].tolist()),\n",
    "    df_lowconf_rev[['annot_lowconf_raw_%(stage)s' % locals() for stage in config['stages']]].iterrows()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2018-05-30T18:17:13.848114Z",
     "start_time": "2018-05-30T18:17:09.930972Z"
    },
    "code_folding": [
     0
    ],
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# df_lowconf_%strand[lcap_log2FoldChange_max] <- max log2FoldChange over all candidates (lowconf_raw) by gene_id\n",
    "\n",
    "# copy fields over in order to group by\n",
    "df_lowconf_fwd['pass1_exon1_gene_id'] = df_exon_fwd['pass1_exon1_gene_id'] \n",
    "df_lowconf_rev['pass1_exon1_gene_id'] = df_exon_rev['pass1_exon1_gene_id']\n",
    "\n",
    "q_ = \"(annot_lowconf_raw_summary == 'coding_promoter_lowconf') | (annot_lowconf_raw_summary == 'pseudogene_promoter_lowconf')\"\n",
    "df_lowconf_fwd['lcap_log2FoldChange'] = df_lcap_fwd[['lcap_%s_fwd_log2FoldChange' % (s_,) for s_ in config['stages']]].max(axis=1)\n",
    "df_lowconf_rev['lcap_log2FoldChange'] = df_lcap_rev[['lcap_%s_rev_log2FoldChange' % (s_,) for s_ in config['stages']]].max(axis=1)\n",
    "\n",
    "df_lowconf_fwd['lcap_log2FoldChange_max'] = df_lowconf_fwd.query(q_)\\\n",
    "    .groupby('pass1_exon1_gene_id')['lcap_log2FoldChange'].transform(np.nanmax)\n",
    "df_lowconf_rev['lcap_log2FoldChange_max'] = df_lowconf_rev.query(q_)\\\n",
    "    .groupby('pass1_exon1_gene_id')['lcap_log2FoldChange'].transform(np.nanmax)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2018-05-30T18:17:21.668416Z",
     "start_time": "2018-05-30T18:17:13.850108Z"
    },
    "code_folding": [
     0
    ]
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fwd strand:\n",
      ".                              41554\n",
      "coding_promoter_lowconf          639\n",
      "pseudogene_promoter_lowconf       52\n",
      "Name: annot_lowconf_summary, dtype: int64\n",
      "Rev strand:\n",
      ".                              41599\n",
      "coding_promoter_lowconf          597\n",
      "pseudogene_promoter_lowconf       49\n",
      "Name: annot_lowconf_summary, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "# df_lowconf_%strand[annot_lowconf_summary] <- low-confidence annotations candidates (=only for \"promoterless genes\")\n",
    "df_lowconf_fwd['annot_detailed_summary'] = df_prom_fwd['annot_detailed_summary']\n",
    "df_lowconf_rev['annot_detailed_summary'] = df_prom_rev['annot_detailed_summary']\n",
    "\n",
    "# list of gene_id without jump/incr promoters, but at least one candidate lowconf promoter\n",
    "l1_ = ['coding_promoter_jump_scap', 'coding_promoter_incr_scap', 'coding_promoter_jump_incr',\n",
    "    'pseudogene_promoter_jump_scap', 'pseudogene_promoter_incr_scap', 'pseudogene_promoter_jump_incr',\n",
    "]\n",
    "l2_ = ['coding_promoter_lowconf', 'pseudogene_promoter_lowconf']\n",
    "l_prom_lowconf_fwd_gene_id = df_lowconf_fwd.groupby('pass1_exon1_gene_id').agg({\n",
    "        'annot_detailed_summary': lambda a: len([a_i for a_i in a if a_i in l1_]),\n",
    "        'annot_lowconf_raw_summary': lambda a: len([a_i for a_i in a if a_i in l2_]),\n",
    "}).query('(annot_detailed_summary == 0) & (annot_lowconf_raw_summary > 0)').index.tolist()\n",
    "\n",
    "l_prom_lowconf_rev_gene_id = df_lowconf_rev.groupby('pass1_exon1_gene_id').agg({\n",
    "        'annot_detailed_summary': lambda a: len([a_i for a_i in a if a_i in l1_]),\n",
    "        'annot_lowconf_raw_summary': lambda a: len([a_i for a_i in a if a_i in l2_]),\n",
    "}).query('(annot_detailed_summary == 0) & (annot_lowconf_raw_summary > 0)').index.tolist()\n",
    "\n",
    "l3_ = [\"coding_promoter_lowconf\", \"pseudogene_promoter_lowconf\"]\n",
    "def annot_lowconf_fwd_summary(ir):\n",
    "    (i, r) = (ir[0], ir[1])\n",
    "    if (r.annot_lowconf_raw_summary in l3_) and (r.pass1_exon1_gene_id in l_prom_lowconf_fwd_gene_id) and (r.lcap_log2FoldChange_max == r.lcap_log2FoldChange):\n",
    "        return r.annot_lowconf_raw_summary\n",
    "    else:\n",
    "        return '.'\n",
    "\n",
    "def annot_lowconf_rev_summary(ir):\n",
    "    (i, r) = (ir[0], ir[1])\n",
    "    if (r.annot_lowconf_raw_summary in l3_) and (r.pass1_exon1_gene_id in l_prom_lowconf_rev_gene_id) and (r.lcap_log2FoldChange_max == r.lcap_log2FoldChange):\n",
    "        return r.annot_lowconf_raw_summary\n",
    "    else:\n",
    "        return '.'\n",
    "\n",
    "df_lowconf_fwd['annot_lowconf_summary'] = list(map(annot_lowconf_fwd_summary, df_lowconf_fwd.iterrows()))\n",
    "df_lowconf_rev['annot_lowconf_summary'] = list(map(annot_lowconf_rev_summary, df_lowconf_rev.iterrows()))\n",
    "print('Fwd strand:')\n",
    "print(df_lowconf_fwd['annot_lowconf_summary'].value_counts())\n",
    "print('Rev strand:')\n",
    "print(df_lowconf_rev['annot_lowconf_summary'].value_counts())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2018-05-30T18:17:28.936145Z",
     "start_time": "2018-05-30T18:17:21.670507Z"
    },
    "code_folding": [
     0
    ],
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# df_lowconf_%strand[annot_lowconf_summary] => .bed-files of the low-confidence annotation\n",
    "d_ = collections.OrderedDict([\n",
    "    ('coding_promoter_lowconf', yp.RED),\n",
    "    ('pseudogene_promoter_lowconf', yp.ORANGE),\n",
    "    ('.', yp.BLACK),\n",
    "])\n",
    "\n",
    "write_gffbed(mp('regulatory_annotation_lowconf_fwd.bed'),\n",
    "        chrom = df_atac['chrom'],\n",
    "        start = df_atac['start'],\n",
    "        end = df_atac['end'],\n",
    "        name = df_exon_fwd['pass1_exon1_locus_id'],\n",
    "        strand = '+',\n",
    "        itemRgb = list(map(lambda lowconf: d_[lowconf], df_lowconf_fwd['annot_lowconf_summary'])),\n",
    "        attr = df_lowconf_fwd['annot_lowconf_summary'],\n",
    ")\n",
    "write_gffbed(mp('regulatory_annotation_lowconf_rev.bed'),\n",
    "        chrom = df_atac['chrom'],\n",
    "        start = df_atac['start'],\n",
    "        end = df_atac['end'],\n",
    "        name = df_exon_rev['pass1_exon1_locus_id'],\n",
    "        strand = '-',\n",
    "        itemRgb = list(map(lambda lowconf: d_[lowconf], df_lowconf_rev['annot_lowconf_summary'])),\n",
    "        attr = df_lowconf_rev['annot_lowconf_summary'],\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2018-05-30T18:17:29.277318Z",
     "start_time": "2018-05-30T18:17:28.939203Z"
    },
    "code_folding": [
     0
    ]
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "691 fwd strand lowconf (pseudo)gene promoters added\n",
      "646 rev strand lowconf (pseudo)gene promoters added\n"
     ]
    }
   ],
   "source": [
    "# df_prom_%strand[annot_detailed_all] <- update with low-confidence annotations\n",
    "m_fwd_ = df_lowconf_fwd['annot_lowconf_summary'] != \".\"\n",
    "m_rev_ = df_lowconf_rev['annot_lowconf_summary'] != \".\"\n",
    "df_prom_fwd.loc[m_fwd_, 'annot_detailed_summary'] = df_lowconf_fwd.loc[m_fwd_, 'annot_lowconf_summary']\n",
    "df_prom_rev.loc[m_rev_, 'annot_detailed_summary'] = df_lowconf_rev.loc[m_rev_, 'annot_lowconf_summary']\n",
    "print('%d fwd strand lowconf (pseudo)gene promoters added' % (sum(m_fwd_),))\n",
    "print('%d rev strand lowconf (pseudo)gene promoters added' % (sum(m_rev_),))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2018-05-30T18:17:29.816754Z",
     "start_time": "2018-05-30T18:17:29.279142Z"
    },
    "code_folding": [
     0
    ]
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>wt_emb</th>\n",
       "      <th>wt_l1</th>\n",
       "      <th>wt_l2</th>\n",
       "      <th>wt_l3</th>\n",
       "      <th>wt_l4</th>\n",
       "      <th>wt_ya</th>\n",
       "      <th>glp1_d1</th>\n",
       "      <th>glp1_d2</th>\n",
       "      <th>glp1_d6</th>\n",
       "      <th>glp1_d9</th>\n",
       "      <th>glp1_d13</th>\n",
       "      <th>summary</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>coding_promoter_jump_scap</th>\n",
       "      <td>3231</td>\n",
       "      <td>3617</td>\n",
       "      <td>4245</td>\n",
       "      <td>3508</td>\n",
       "      <td>3604</td>\n",
       "      <td>3168</td>\n",
       "      <td>2475</td>\n",
       "      <td>2114</td>\n",
       "      <td>2463</td>\n",
       "      <td>2418</td>\n",
       "      <td>2449</td>\n",
       "      <td>6055</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>coding_promoter_incr_scap</th>\n",
       "      <td>390</td>\n",
       "      <td>281</td>\n",
       "      <td>293</td>\n",
       "      <td>439</td>\n",
       "      <td>336</td>\n",
       "      <td>452</td>\n",
       "      <td>404</td>\n",
       "      <td>318</td>\n",
       "      <td>486</td>\n",
       "      <td>307</td>\n",
       "      <td>521</td>\n",
       "      <td>788</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>coding_promoter_jump_incr</th>\n",
       "      <td>52</td>\n",
       "      <td>71</td>\n",
       "      <td>89</td>\n",
       "      <td>71</td>\n",
       "      <td>97</td>\n",
       "      <td>99</td>\n",
       "      <td>71</td>\n",
       "      <td>95</td>\n",
       "      <td>69</td>\n",
       "      <td>90</td>\n",
       "      <td>76</td>\n",
       "      <td>357</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>coding_promoter_lowconf</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>639</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pseudogene_promoter_jump_scap</th>\n",
       "      <td>49</td>\n",
       "      <td>43</td>\n",
       "      <td>43</td>\n",
       "      <td>32</td>\n",
       "      <td>41</td>\n",
       "      <td>49</td>\n",
       "      <td>24</td>\n",
       "      <td>13</td>\n",
       "      <td>25</td>\n",
       "      <td>19</td>\n",
       "      <td>21</td>\n",
       "      <td>90</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pseudogene_promoter_incr_scap</th>\n",
       "      <td>11</td>\n",
       "      <td>8</td>\n",
       "      <td>10</td>\n",
       "      <td>13</td>\n",
       "      <td>9</td>\n",
       "      <td>3</td>\n",
       "      <td>9</td>\n",
       "      <td>4</td>\n",
       "      <td>9</td>\n",
       "      <td>10</td>\n",
       "      <td>11</td>\n",
       "      <td>20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pseudogene_promoter_jump_incr</th>\n",
       "      <td>8</td>\n",
       "      <td>8</td>\n",
       "      <td>11</td>\n",
       "      <td>7</td>\n",
       "      <td>11</td>\n",
       "      <td>8</td>\n",
       "      <td>6</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pseudogene_promoter_lowconf</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>52</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>non-coding_RNA</th>\n",
       "      <td>495</td>\n",
       "      <td>489</td>\n",
       "      <td>485</td>\n",
       "      <td>488</td>\n",
       "      <td>494</td>\n",
       "      <td>491</td>\n",
       "      <td>498</td>\n",
       "      <td>498</td>\n",
       "      <td>490</td>\n",
       "      <td>493</td>\n",
       "      <td>486</td>\n",
       "      <td>457</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>unknown_promoter_jump_scap</th>\n",
       "      <td>468</td>\n",
       "      <td>600</td>\n",
       "      <td>798</td>\n",
       "      <td>406</td>\n",
       "      <td>328</td>\n",
       "      <td>263</td>\n",
       "      <td>189</td>\n",
       "      <td>150</td>\n",
       "      <td>214</td>\n",
       "      <td>285</td>\n",
       "      <td>362</td>\n",
       "      <td>1508</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>transcription_initiation</th>\n",
       "      <td>24456</td>\n",
       "      <td>24062</td>\n",
       "      <td>23226</td>\n",
       "      <td>24214</td>\n",
       "      <td>24288</td>\n",
       "      <td>24674</td>\n",
       "      <td>25501</td>\n",
       "      <td>26003</td>\n",
       "      <td>25412</td>\n",
       "      <td>25568</td>\n",
       "      <td>25250</td>\n",
       "      <td>19822</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>no_transcription</th>\n",
       "      <td>13085</td>\n",
       "      <td>13066</td>\n",
       "      <td>13045</td>\n",
       "      <td>13067</td>\n",
       "      <td>13037</td>\n",
       "      <td>13038</td>\n",
       "      <td>13068</td>\n",
       "      <td>13049</td>\n",
       "      <td>13072</td>\n",
       "      <td>13051</td>\n",
       "      <td>13067</td>\n",
       "      <td>12434</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                               wt_emb  wt_l1  wt_l2  wt_l3  wt_l4  wt_ya  \\\n",
       "coding_promoter_jump_scap        3231   3617   4245   3508   3604   3168   \n",
       "coding_promoter_incr_scap         390    281    293    439    336    452   \n",
       "coding_promoter_jump_incr          52     71     89     71     97     99   \n",
       "coding_promoter_lowconf             0      0      0      0      0      0   \n",
       "pseudogene_promoter_jump_scap      49     43     43     32     41     49   \n",
       "pseudogene_promoter_incr_scap      11      8     10     13      9      3   \n",
       "pseudogene_promoter_jump_incr       8      8     11      7     11      8   \n",
       "pseudogene_promoter_lowconf         0      0      0      0      0      0   \n",
       "non-coding_RNA                    495    489    485    488    494    491   \n",
       "unknown_promoter_jump_scap        468    600    798    406    328    263   \n",
       "transcription_initiation        24456  24062  23226  24214  24288  24674   \n",
       "no_transcription                13085  13066  13045  13067  13037  13038   \n",
       "\n",
       "                               glp1_d1  glp1_d2  glp1_d6  glp1_d9  glp1_d13  \\\n",
       "coding_promoter_jump_scap         2475     2114     2463     2418      2449   \n",
       "coding_promoter_incr_scap          404      318      486      307       521   \n",
       "coding_promoter_jump_incr           71       95       69       90        76   \n",
       "coding_promoter_lowconf              0        0        0        0         0   \n",
       "pseudogene_promoter_jump_scap       24       13       25       19        21   \n",
       "pseudogene_promoter_incr_scap        9        4        9       10        11   \n",
       "pseudogene_promoter_jump_incr        6        1        5        4         2   \n",
       "pseudogene_promoter_lowconf          0        0        0        0         0   \n",
       "non-coding_RNA                     498      498      490      493       486   \n",
       "unknown_promoter_jump_scap         189      150      214      285       362   \n",
       "transcription_initiation         25501    26003    25412    25568     25250   \n",
       "no_transcription                 13068    13049    13072    13051     13067   \n",
       "\n",
       "                               summary  \n",
       "coding_promoter_jump_scap         6055  \n",
       "coding_promoter_incr_scap          788  \n",
       "coding_promoter_jump_incr          357  \n",
       "coding_promoter_lowconf            639  \n",
       "pseudogene_promoter_jump_scap       90  \n",
       "pseudogene_promoter_incr_scap       20  \n",
       "pseudogene_promoter_jump_incr       23  \n",
       "pseudogene_promoter_lowconf         52  \n",
       "non-coding_RNA                     457  \n",
       "unknown_promoter_jump_scap        1508  \n",
       "transcription_initiation         19822  \n",
       "no_transcription                 12434  "
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Forward strand annotation summary table (incl. lowconf across all stages but not the origin stage)\n",
    "df_fwd_ = pd.concat([df_prom_fwd['annot_detailed_%(stage)s' % locals()].value_counts() for stage in config['stages'] + ['summary']], axis=1).loc[d_detailed_annot.keys()]\n",
    "df_fwd_.columns = config['stages'] + ['summary']\n",
    "#df_fwd_.to_latex('%(dm_step)s_fig/prom_summary_fwd.latex' % locals())\n",
    "df_fwd_.fillna(0, downcast='infer')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2018-05-30T18:17:31.312302Z",
     "start_time": "2018-05-30T18:17:29.819405Z"
    },
    "code_folding": [
     0
    ]
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>wt_emb</th>\n",
       "      <th>wt_l1</th>\n",
       "      <th>wt_l2</th>\n",
       "      <th>wt_l3</th>\n",
       "      <th>wt_l4</th>\n",
       "      <th>wt_ya</th>\n",
       "      <th>glp1_d1</th>\n",
       "      <th>glp1_d2</th>\n",
       "      <th>glp1_d6</th>\n",
       "      <th>glp1_d9</th>\n",
       "      <th>glp1_d13</th>\n",
       "      <th>summary</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>coding_promoter_jump_scap</th>\n",
       "      <td>3297</td>\n",
       "      <td>3618</td>\n",
       "      <td>4246</td>\n",
       "      <td>3720</td>\n",
       "      <td>3597</td>\n",
       "      <td>3194</td>\n",
       "      <td>2471</td>\n",
       "      <td>2156</td>\n",
       "      <td>2459</td>\n",
       "      <td>2370</td>\n",
       "      <td>2559</td>\n",
       "      <td>6040</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>coding_promoter_incr_scap</th>\n",
       "      <td>340</td>\n",
       "      <td>288</td>\n",
       "      <td>295</td>\n",
       "      <td>356</td>\n",
       "      <td>352</td>\n",
       "      <td>434</td>\n",
       "      <td>382</td>\n",
       "      <td>343</td>\n",
       "      <td>474</td>\n",
       "      <td>366</td>\n",
       "      <td>460</td>\n",
       "      <td>703</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>coding_promoter_jump_incr</th>\n",
       "      <td>69</td>\n",
       "      <td>73</td>\n",
       "      <td>83</td>\n",
       "      <td>93</td>\n",
       "      <td>110</td>\n",
       "      <td>106</td>\n",
       "      <td>79</td>\n",
       "      <td>88</td>\n",
       "      <td>76</td>\n",
       "      <td>88</td>\n",
       "      <td>93</td>\n",
       "      <td>393</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>coding_promoter_lowconf</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>597</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pseudogene_promoter_jump_scap</th>\n",
       "      <td>55</td>\n",
       "      <td>54</td>\n",
       "      <td>62</td>\n",
       "      <td>52</td>\n",
       "      <td>48</td>\n",
       "      <td>46</td>\n",
       "      <td>22</td>\n",
       "      <td>11</td>\n",
       "      <td>27</td>\n",
       "      <td>23</td>\n",
       "      <td>26</td>\n",
       "      <td>107</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pseudogene_promoter_incr_scap</th>\n",
       "      <td>13</td>\n",
       "      <td>11</td>\n",
       "      <td>6</td>\n",
       "      <td>11</td>\n",
       "      <td>10</td>\n",
       "      <td>15</td>\n",
       "      <td>6</td>\n",
       "      <td>9</td>\n",
       "      <td>10</td>\n",
       "      <td>6</td>\n",
       "      <td>7</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pseudogene_promoter_jump_incr</th>\n",
       "      <td>7</td>\n",
       "      <td>8</td>\n",
       "      <td>11</td>\n",
       "      <td>8</td>\n",
       "      <td>11</td>\n",
       "      <td>11</td>\n",
       "      <td>5</td>\n",
       "      <td>3</td>\n",
       "      <td>5</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pseudogene_promoter_lowconf</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>49</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>non-coding_RNA</th>\n",
       "      <td>496</td>\n",
       "      <td>494</td>\n",
       "      <td>492</td>\n",
       "      <td>496</td>\n",
       "      <td>498</td>\n",
       "      <td>493</td>\n",
       "      <td>502</td>\n",
       "      <td>507</td>\n",
       "      <td>502</td>\n",
       "      <td>499</td>\n",
       "      <td>494</td>\n",
       "      <td>468</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>unknown_promoter_jump_scap</th>\n",
       "      <td>443</td>\n",
       "      <td>617</td>\n",
       "      <td>863</td>\n",
       "      <td>493</td>\n",
       "      <td>366</td>\n",
       "      <td>245</td>\n",
       "      <td>179</td>\n",
       "      <td>172</td>\n",
       "      <td>208</td>\n",
       "      <td>270</td>\n",
       "      <td>379</td>\n",
       "      <td>1598</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>transcription_initiation</th>\n",
       "      <td>24385</td>\n",
       "      <td>23947</td>\n",
       "      <td>23066</td>\n",
       "      <td>23901</td>\n",
       "      <td>24159</td>\n",
       "      <td>24602</td>\n",
       "      <td>25468</td>\n",
       "      <td>25832</td>\n",
       "      <td>25350</td>\n",
       "      <td>25495</td>\n",
       "      <td>25104</td>\n",
       "      <td>19756</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>no_transcription</th>\n",
       "      <td>13140</td>\n",
       "      <td>13135</td>\n",
       "      <td>13121</td>\n",
       "      <td>13115</td>\n",
       "      <td>13094</td>\n",
       "      <td>13099</td>\n",
       "      <td>13131</td>\n",
       "      <td>13124</td>\n",
       "      <td>13134</td>\n",
       "      <td>13122</td>\n",
       "      <td>13117</td>\n",
       "      <td>12486</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                               wt_emb  wt_l1  wt_l2  wt_l3  wt_l4  wt_ya  \\\n",
       "coding_promoter_jump_scap        3297   3618   4246   3720   3597   3194   \n",
       "coding_promoter_incr_scap         340    288    295    356    352    434   \n",
       "coding_promoter_jump_incr          69     73     83     93    110    106   \n",
       "coding_promoter_lowconf             0      0      0      0      0      0   \n",
       "pseudogene_promoter_jump_scap      55     54     62     52     48     46   \n",
       "pseudogene_promoter_incr_scap      13     11      6     11     10     15   \n",
       "pseudogene_promoter_jump_incr       7      8     11      8     11     11   \n",
       "pseudogene_promoter_lowconf         0      0      0      0      0      0   \n",
       "non-coding_RNA                    496    494    492    496    498    493   \n",
       "unknown_promoter_jump_scap        443    617    863    493    366    245   \n",
       "transcription_initiation        24385  23947  23066  23901  24159  24602   \n",
       "no_transcription                13140  13135  13121  13115  13094  13099   \n",
       "\n",
       "                               glp1_d1  glp1_d2  glp1_d6  glp1_d9  glp1_d13  \\\n",
       "coding_promoter_jump_scap         2471     2156     2459     2370      2559   \n",
       "coding_promoter_incr_scap          382      343      474      366       460   \n",
       "coding_promoter_jump_incr           79       88       76       88        93   \n",
       "coding_promoter_lowconf              0        0        0        0         0   \n",
       "pseudogene_promoter_jump_scap       22       11       27       23        26   \n",
       "pseudogene_promoter_incr_scap        6        9       10        6         7   \n",
       "pseudogene_promoter_jump_incr        5        3        5        6         6   \n",
       "pseudogene_promoter_lowconf          0        0        0        0         0   \n",
       "non-coding_RNA                     502      507      502      499       494   \n",
       "unknown_promoter_jump_scap         179      172      208      270       379   \n",
       "transcription_initiation         25468    25832    25350    25495     25104   \n",
       "no_transcription                 13131    13124    13134    13122     13117   \n",
       "\n",
       "                               summary  \n",
       "coding_promoter_jump_scap         6040  \n",
       "coding_promoter_incr_scap          703  \n",
       "coding_promoter_jump_incr          393  \n",
       "coding_promoter_lowconf            597  \n",
       "pseudogene_promoter_jump_scap      107  \n",
       "pseudogene_promoter_incr_scap       19  \n",
       "pseudogene_promoter_jump_incr       29  \n",
       "pseudogene_promoter_lowconf         49  \n",
       "non-coding_RNA                     468  \n",
       "unknown_promoter_jump_scap        1598  \n",
       "transcription_initiation         19756  \n",
       "no_transcription                 12486  "
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Reverse strand annotation summary table (incl. lowconf across all stages but not the origin stage)\n",
    "df_rev_ = pd.concat([df_prom_rev['annot_detailed_%(stage)s' % locals()].value_counts() for stage in config['stages'] + ['summary']], axis=1).loc[d_detailed_annot.keys()]\n",
    "df_rev_.columns = config['stages'] + ['summary']\n",
    "#df_rev_.to_latex('%(dm_step)s_fig/prom_summary_rev.latex' % locals())\n",
    "df_rev_.fillna(0, downcast='infer')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2018-05-30T18:17:31.656359Z",
     "start_time": "2018-05-30T18:17:31.314425Z"
    },
    "code_folding": [
     0
    ],
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# annot_detailed => collapse by method\n",
    "for s_ in config['stages'] + ['summary']:\n",
    "    df_prom_fwd['annot_%s' % (s_,)] = list(map(lambda annot_detailed: d_detailed_annot[annot_detailed], \n",
    "                                               df_prom_fwd['annot_detailed_%s' % (s_,)]))\n",
    "    df_prom_rev['annot_%s' % (s_,)] = list(map(lambda annot_detailed: d_detailed_annot[annot_detailed], \n",
    "                                               df_prom_rev['annot_detailed_%s' % (s_,)]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2018-05-30T18:17:31.782955Z",
     "start_time": "2018-05-30T18:17:31.658282Z"
    },
    "code_folding": [
     0
    ]
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>wt_emb</th>\n",
       "      <th>wt_l1</th>\n",
       "      <th>wt_l2</th>\n",
       "      <th>wt_l3</th>\n",
       "      <th>wt_l4</th>\n",
       "      <th>wt_ya</th>\n",
       "      <th>glp1_d1</th>\n",
       "      <th>glp1_d2</th>\n",
       "      <th>glp1_d6</th>\n",
       "      <th>glp1_d9</th>\n",
       "      <th>glp1_d13</th>\n",
       "      <th>summary</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>coding_promoter</th>\n",
       "      <td>3673</td>\n",
       "      <td>3969</td>\n",
       "      <td>4627</td>\n",
       "      <td>4018</td>\n",
       "      <td>4037</td>\n",
       "      <td>3719</td>\n",
       "      <td>2950</td>\n",
       "      <td>2527</td>\n",
       "      <td>3018</td>\n",
       "      <td>2815</td>\n",
       "      <td>3046</td>\n",
       "      <td>7839</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pseudogene_promoter</th>\n",
       "      <td>68</td>\n",
       "      <td>59</td>\n",
       "      <td>64</td>\n",
       "      <td>52</td>\n",
       "      <td>61</td>\n",
       "      <td>60</td>\n",
       "      <td>39</td>\n",
       "      <td>18</td>\n",
       "      <td>39</td>\n",
       "      <td>33</td>\n",
       "      <td>34</td>\n",
       "      <td>185</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>non-coding_RNA</th>\n",
       "      <td>495</td>\n",
       "      <td>489</td>\n",
       "      <td>485</td>\n",
       "      <td>488</td>\n",
       "      <td>494</td>\n",
       "      <td>491</td>\n",
       "      <td>498</td>\n",
       "      <td>498</td>\n",
       "      <td>490</td>\n",
       "      <td>493</td>\n",
       "      <td>486</td>\n",
       "      <td>457</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>unknown_promoter</th>\n",
       "      <td>468</td>\n",
       "      <td>600</td>\n",
       "      <td>798</td>\n",
       "      <td>406</td>\n",
       "      <td>328</td>\n",
       "      <td>263</td>\n",
       "      <td>189</td>\n",
       "      <td>150</td>\n",
       "      <td>214</td>\n",
       "      <td>285</td>\n",
       "      <td>362</td>\n",
       "      <td>1508</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>transcription_initiation</th>\n",
       "      <td>24456</td>\n",
       "      <td>24062</td>\n",
       "      <td>23226</td>\n",
       "      <td>24214</td>\n",
       "      <td>24288</td>\n",
       "      <td>24674</td>\n",
       "      <td>25501</td>\n",
       "      <td>26003</td>\n",
       "      <td>25412</td>\n",
       "      <td>25568</td>\n",
       "      <td>25250</td>\n",
       "      <td>19822</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>no_transcription</th>\n",
       "      <td>13085</td>\n",
       "      <td>13066</td>\n",
       "      <td>13045</td>\n",
       "      <td>13067</td>\n",
       "      <td>13037</td>\n",
       "      <td>13038</td>\n",
       "      <td>13068</td>\n",
       "      <td>13049</td>\n",
       "      <td>13072</td>\n",
       "      <td>13051</td>\n",
       "      <td>13067</td>\n",
       "      <td>12434</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                          wt_emb  wt_l1  wt_l2  wt_l3  wt_l4  wt_ya  glp1_d1  \\\n",
       "coding_promoter             3673   3969   4627   4018   4037   3719     2950   \n",
       "pseudogene_promoter           68     59     64     52     61     60       39   \n",
       "non-coding_RNA               495    489    485    488    494    491      498   \n",
       "unknown_promoter             468    600    798    406    328    263      189   \n",
       "transcription_initiation   24456  24062  23226  24214  24288  24674    25501   \n",
       "no_transcription           13085  13066  13045  13067  13037  13038    13068   \n",
       "\n",
       "                          glp1_d2  glp1_d6  glp1_d9  glp1_d13  summary  \n",
       "coding_promoter              2527     3018     2815      3046     7839  \n",
       "pseudogene_promoter            18       39       33        34      185  \n",
       "non-coding_RNA                498      490      493       486      457  \n",
       "unknown_promoter              150      214      285       362     1508  \n",
       "transcription_initiation    26003    25412    25568     25250    19822  \n",
       "no_transcription            13049    13072    13051     13067    12434  "
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Forward strand annotation (collapsed across methods)\n",
    "df_fwd_ = pd.concat([df_prom_fwd['annot_%(stage)s' % locals()].value_counts() for stage in config['stages'] + ['summary']], axis=1).loc[d_annot_strand_legend.keys()]\n",
    "df_fwd_.columns = config['stages'] + ['summary']\n",
    "#df_fwd_.to_latex('%(dm_step)s_fig/prom_summary_rev.latex' % locals())\n",
    "#plt.subplot(111, frame_on=False)\n",
    "#plt.gca().xaxis.set_visible(False)\n",
    "#plt.gca().yaxis.set_visible(False)\n",
    "#pd.plotting.table(data=df_fwd_, ax=plt.gca(), loc='center left')\n",
    "#plt.savefig('annot/FigA_mapping/annot_crosstab_fwd.pdf', bbox_inches='tight')\n",
    "df_fwd_"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2018-05-30T18:17:31.995141Z",
     "start_time": "2018-05-30T18:17:31.784506Z"
    },
    "code_folding": [
     0
    ],
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>wt_emb</th>\n",
       "      <th>wt_l1</th>\n",
       "      <th>wt_l2</th>\n",
       "      <th>wt_l3</th>\n",
       "      <th>wt_l4</th>\n",
       "      <th>wt_ya</th>\n",
       "      <th>glp1_d1</th>\n",
       "      <th>glp1_d2</th>\n",
       "      <th>glp1_d6</th>\n",
       "      <th>glp1_d9</th>\n",
       "      <th>glp1_d13</th>\n",
       "      <th>summary</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>coding_promoter</th>\n",
       "      <td>3706</td>\n",
       "      <td>3979</td>\n",
       "      <td>4624</td>\n",
       "      <td>4169</td>\n",
       "      <td>4059</td>\n",
       "      <td>3734</td>\n",
       "      <td>2932</td>\n",
       "      <td>2587</td>\n",
       "      <td>3009</td>\n",
       "      <td>2824</td>\n",
       "      <td>3112</td>\n",
       "      <td>7733</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pseudogene_promoter</th>\n",
       "      <td>75</td>\n",
       "      <td>73</td>\n",
       "      <td>79</td>\n",
       "      <td>71</td>\n",
       "      <td>69</td>\n",
       "      <td>72</td>\n",
       "      <td>33</td>\n",
       "      <td>23</td>\n",
       "      <td>42</td>\n",
       "      <td>35</td>\n",
       "      <td>39</td>\n",
       "      <td>204</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>non-coding_RNA</th>\n",
       "      <td>496</td>\n",
       "      <td>494</td>\n",
       "      <td>492</td>\n",
       "      <td>496</td>\n",
       "      <td>498</td>\n",
       "      <td>493</td>\n",
       "      <td>502</td>\n",
       "      <td>507</td>\n",
       "      <td>502</td>\n",
       "      <td>499</td>\n",
       "      <td>494</td>\n",
       "      <td>468</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>unknown_promoter</th>\n",
       "      <td>443</td>\n",
       "      <td>617</td>\n",
       "      <td>863</td>\n",
       "      <td>493</td>\n",
       "      <td>366</td>\n",
       "      <td>245</td>\n",
       "      <td>179</td>\n",
       "      <td>172</td>\n",
       "      <td>208</td>\n",
       "      <td>270</td>\n",
       "      <td>379</td>\n",
       "      <td>1598</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>transcription_initiation</th>\n",
       "      <td>24385</td>\n",
       "      <td>23947</td>\n",
       "      <td>23066</td>\n",
       "      <td>23901</td>\n",
       "      <td>24159</td>\n",
       "      <td>24602</td>\n",
       "      <td>25468</td>\n",
       "      <td>25832</td>\n",
       "      <td>25350</td>\n",
       "      <td>25495</td>\n",
       "      <td>25104</td>\n",
       "      <td>19756</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>no_transcription</th>\n",
       "      <td>13140</td>\n",
       "      <td>13135</td>\n",
       "      <td>13121</td>\n",
       "      <td>13115</td>\n",
       "      <td>13094</td>\n",
       "      <td>13099</td>\n",
       "      <td>13131</td>\n",
       "      <td>13124</td>\n",
       "      <td>13134</td>\n",
       "      <td>13122</td>\n",
       "      <td>13117</td>\n",
       "      <td>12486</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                          wt_emb  wt_l1  wt_l2  wt_l3  wt_l4  wt_ya  glp1_d1  \\\n",
       "coding_promoter             3706   3979   4624   4169   4059   3734     2932   \n",
       "pseudogene_promoter           75     73     79     71     69     72       33   \n",
       "non-coding_RNA               496    494    492    496    498    493      502   \n",
       "unknown_promoter             443    617    863    493    366    245      179   \n",
       "transcription_initiation   24385  23947  23066  23901  24159  24602    25468   \n",
       "no_transcription           13140  13135  13121  13115  13094  13099    13131   \n",
       "\n",
       "                          glp1_d2  glp1_d6  glp1_d9  glp1_d13  summary  \n",
       "coding_promoter              2587     3009     2824      3112     7733  \n",
       "pseudogene_promoter            23       42       35        39      204  \n",
       "non-coding_RNA                507      502      499       494      468  \n",
       "unknown_promoter              172      208      270       379     1598  \n",
       "transcription_initiation    25832    25350    25495     25104    19756  \n",
       "no_transcription            13124    13134    13122     13117    12486  "
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Reverse strand annotation summary (collapsed across methods)\n",
    "df_rev_ = pd.concat([df_prom_rev['annot_%(stage)s' % locals()].value_counts() for stage in config['stages'] + ['summary']], axis=1).loc[d_annot_strand_legend.keys()]\n",
    "df_rev_.columns = config['stages'] + ['summary']\n",
    "#df_rev_.to_latex('%(dm_step)s_fig/prom_summary_rev.latex' % locals())\n",
    "#plt.subplot(111, frame_on=False)\n",
    "#plt.gca().xaxis.set_visible(False)\n",
    "#plt.gca().yaxis.set_visible(False)\n",
    "#pd.plotting.table(data=df_rev_, ax=plt.gca(), loc='center left')\n",
    "#plt.savefig('annot/FigA_mapping/annot_crosstab_rev.pdf', bbox_inches='tight')\n",
    "df_rev_"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2018-05-30T18:17:32.088435Z",
     "start_time": "2018-05-30T18:17:31.997187Z"
    },
    "code_folding": [
     0
    ]
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>col_0</th>\n",
       "      <th>coding_promoter</th>\n",
       "      <th>pseudogene_promoter</th>\n",
       "      <th>non-coding_RNA</th>\n",
       "      <th>unknown_promoter</th>\n",
       "      <th>transcription_initiation</th>\n",
       "      <th>no_transcription</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>row_0</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>coding_promoter</th>\n",
       "      <td>1976</td>\n",
       "      <td>46</td>\n",
       "      <td>28</td>\n",
       "      <td>638</td>\n",
       "      <td>3703</td>\n",
       "      <td>1448</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pseudogene_promoter</th>\n",
       "      <td>43</td>\n",
       "      <td>9</td>\n",
       "      <td>1</td>\n",
       "      <td>9</td>\n",
       "      <td>71</td>\n",
       "      <td>52</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>non-coding_RNA</th>\n",
       "      <td>25</td>\n",
       "      <td>0</td>\n",
       "      <td>47</td>\n",
       "      <td>6</td>\n",
       "      <td>121</td>\n",
       "      <td>258</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>unknown_promoter</th>\n",
       "      <td>606</td>\n",
       "      <td>11</td>\n",
       "      <td>5</td>\n",
       "      <td>40</td>\n",
       "      <td>714</td>\n",
       "      <td>132</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>transcription_initiation</th>\n",
       "      <td>3663</td>\n",
       "      <td>70</td>\n",
       "      <td>107</td>\n",
       "      <td>737</td>\n",
       "      <td>11161</td>\n",
       "      <td>4084</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>no_transcription</th>\n",
       "      <td>1420</td>\n",
       "      <td>68</td>\n",
       "      <td>280</td>\n",
       "      <td>168</td>\n",
       "      <td>3986</td>\n",
       "      <td>6512</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "col_0                     coding_promoter  pseudogene_promoter  \\\n",
       "row_0                                                            \n",
       "coding_promoter                      1976                   46   \n",
       "pseudogene_promoter                    43                    9   \n",
       "non-coding_RNA                         25                    0   \n",
       "unknown_promoter                      606                   11   \n",
       "transcription_initiation             3663                   70   \n",
       "no_transcription                     1420                   68   \n",
       "\n",
       "col_0                     non-coding_RNA  unknown_promoter  \\\n",
       "row_0                                                        \n",
       "coding_promoter                       28               638   \n",
       "pseudogene_promoter                    1                 9   \n",
       "non-coding_RNA                        47                 6   \n",
       "unknown_promoter                       5                40   \n",
       "transcription_initiation             107               737   \n",
       "no_transcription                     280               168   \n",
       "\n",
       "col_0                     transcription_initiation  no_transcription  \n",
       "row_0                                                                 \n",
       "coding_promoter                               3703              1448  \n",
       "pseudogene_promoter                             71                52  \n",
       "non-coding_RNA                                 121               258  \n",
       "unknown_promoter                               714               132  \n",
       "transcription_initiation                     11161              4084  \n",
       "no_transcription                              3986              6512  "
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Cross-tabulation of strand-specific summary annotation\n",
    "pd.crosstab(pd.Categorical(df_prom_fwd['annot_summary']), pd.Categorical(df_prom_rev['annot_summary'])).loc[list(d_annot_strand_legend.keys()), list(d_annot_strand_legend.keys())]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2018-05-30T18:17:32.699579Z",
     "start_time": "2018-05-30T18:17:32.090713Z"
    },
    "code_folding": [
     0
    ]
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "13596 bidirect, 23059 with scap, 10056 both, 73.96\n"
     ]
    }
   ],
   "source": [
    "# Promoter short cap statistics for text\n",
    "m_prom = (df_prom_fwd['annot_summary'] == 'coding_promoter') | (df_prom_rev['annot_summary'] == 'coding_promoter')\n",
    "m_both_scap = (df_scap_fwd['scap_mode_count'] >= 2) & (df_scap_rev['scap_mode_count'] >= 2)\n",
    "f_ = sum(m_prom & m_both_scap) / sum(m_prom)\n",
    "print('%d bidirect, %d with scap, %d both, %.2f' % (sum(m_prom), sum(m_both_scap), sum(m_prom & m_both_scap), 100*f_))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2018-05-30T18:17:59.358653Z",
     "start_time": "2018-05-30T18:17:32.702601Z"
    },
    "code_folding": [
     0
    ]
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "5927 fwd 5867 rev 11794 pooled\n",
      "11794 annot_ce10/metrics_type/_df_prom_bounds.bed\n",
      "23155 annot_ce10/metrics_type/_df_gene_bounds.bed\n",
      "23156 annot_ce10/metrics_type/outron-extended_genes.bed\n"
     ]
    }
   ],
   "source": [
    "# S2_outron-extended_genes.bed <- genes extended by outron regions as defined by (aggregated) promoter annotations\n",
    "def df_outron_extended_genes_(df_prom_fwd_, df_prom_rev_):\n",
    "\n",
    "    # df_prom_bounds <- prom min/max boundaries, based on promoter annotation\n",
    "    df_prom_bounds_fwd = df_prom_fwd_.groupby(['pass1_exon1_gene_id', 'pass1_exon1_locus_id']).agg({\n",
    "        'chrom': lambda s: list(set(s))[0],\n",
    "        'start': np.min,\n",
    "        'end': np.max,\n",
    "    }).reset_index()[['chrom', 'start', 'end', 'pass1_exon1_gene_id', 'pass1_exon1_locus_id']]\n",
    "    df_prom_bounds_fwd.columns = ['chrom', 'start', 'end', 'gene_id', 'locus_id']\n",
    "\n",
    "    df_prom_bounds_rev = df_prom_rev_.groupby(['pass1_exon1_gene_id', 'pass1_exon1_locus_id']).agg({\n",
    "        'chrom': lambda s: list(set(s))[0],\n",
    "        'start': np.min,\n",
    "        'end': np.max,\n",
    "    }).reset_index()[['chrom', 'start', 'end', 'pass1_exon1_gene_id', 'pass1_exon1_locus_id']]\n",
    "    df_prom_bounds_rev.columns = ['chrom', 'start', 'end', 'gene_id', 'locus_id']\n",
    "\n",
    "    df_prom_bounds = pd.concat([df_prom_bounds_fwd, df_prom_bounds_rev], axis=0, ignore_index=False)\\\n",
    "        .sort_values(['chrom', 'start', 'end']).reset_index(drop=True)\n",
    "    print('%d fwd %d rev %d pooled' % (len(df_prom_bounds_fwd), len(df_prom_bounds_rev), len(df_prom_bounds)))\n",
    "\n",
    "    # Write a .bed-file of all the regions that changed\n",
    "    df_prom_bounds.to_csv(mp('_df_prom_bounds.bed'), index=False, header=False, sep='\\t')\n",
    "    !wc -l {mp('_df_prom_bounds.bed')}\n",
    "\n",
    "    # df_gene_bounds <- gene min/max boundaries, based on annotation\n",
    "    df_transcripts_ = yp.read_wbgtf('WS260_ce10/WS260_ce10.transcripts.annot.gtf.gz', parse_attr=False)\n",
    "    df_transcripts = yp.df_gfftags_unpack(df_transcripts_, name='attribute')\n",
    "    \n",
    "    # Add display_id\n",
    "    #fp_geneIDs = 'wget/ftp.wormbase.org/pub/wormbase/releases/WS260/species/c_elegans/PRJNA13758/annotation/c_elegans.PRJNA13758.WS260.geneIDs.txt.gz'\n",
    "    #df_geneIDs = pd.read_csv(fp_geneIDs, sep=',', names=('na', 'gene_id', 'locus', 'sequence_id', 'status'))[['gene_id', 'locus', 'sequence_id', 'status']]\n",
    "    #\n",
    "    #def display_id_(locus, sequence_id, gene_id):\n",
    "    #    if locus == locus:\n",
    "    #        return locus\n",
    "    #    elif sequence_id == sequence_id:\n",
    "    #        return sequence_id\n",
    "    #    else:\n",
    "    #        return gene_id\n",
    "    #\n",
    "    #df_geneIDs['locus_id'] = list(map(display_id_, df_geneIDs['locus'], df_geneIDs['sequence_id'], df_geneIDs['gene_id']))\n",
    "    #df_transcripts = pd.merge(df_transcripts, df_geneIDs[['gene_id', 'locus_id']], left_on='gene_id', right_on='gene_id')\n",
    "\n",
    "    df_gene_bounds = df_transcripts.groupby(['gene_id', 'locus_id']).agg({\n",
    "        'chrom': lambda s: list(set(s))[0],\n",
    "        'start': np.min,\n",
    "        'end': np.max,\n",
    "    }).reset_index()[['chrom', 'start', 'end', 'gene_id', 'locus_id']].sort_values(['chrom', 'start', 'end']).reset_index(drop=True)\n",
    "\n",
    "    df_gene_bounds.to_csv(mp('_df_gene_bounds.bed'), index=False, header=False, sep='\\t')\n",
    "    !wc -l {mp('_df_gene_bounds.bed')}\n",
    "    \n",
    "    df_ = df_prom_bounds.merge(df_gene_bounds, how='outer', on=['gene_id', 'locus_id'], suffixes=('_prom', '_gene'))\n",
    "    df_['start'] = list(map(int, map(np.nanmin, zip(df_['start_prom'], df_['start_gene']))))\n",
    "    df_['end'] = list(map(int, map(np.nanmax, zip(df_['end_prom'], df_['end_gene']))))\n",
    "    df_ = df_[['chrom_gene', 'start', 'end', 'gene_id', 'locus_id']].sort_values(['chrom_gene', 'start', 'end']).reset_index(drop=True)\n",
    "    df_.columns = [['chrom', 'start', 'end', 'gene_id', 'locus_id']]\n",
    "    return df_\n",
    "\n",
    "q_ = '(annot_summary == \"coding_promoter\") | (annot_summary == \"pseudogene_promoter\") | (annot_summary == \"non-coding_RNA\")'\n",
    "df_outron_extended_genes = df_outron_extended_genes_(\n",
    "    pd.concat([df_prom_fwd, df_exon_fwd['pass1_exon1_gene_id'], df_exon_fwd['pass1_exon1_locus_id']], axis=1).query(q_)[['chrom', 'start', 'end', 'annot_summary', 'pass1_exon1_gene_id', 'pass1_exon1_locus_id']],\n",
    "    pd.concat([df_prom_rev, df_exon_rev['pass1_exon1_gene_id'], df_exon_rev['pass1_exon1_locus_id']], axis=1).query(q_)[['chrom', 'start', 'end', 'annot_summary', 'pass1_exon1_gene_id', 'pass1_exon1_locus_id']],\n",
    ")\n",
    "df_outron_extended_genes.to_csv(mp('outron-extended_genes.bed'), index=False, header=False, sep='\\t')\n",
    "!wc -l {mp('outron-extended_genes.bed')}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2018-05-30T18:17:59.894721Z",
     "start_time": "2018-05-30T18:17:59.362424Z"
    },
    "code_folding": [
     0
    ]
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "***** WARNING: File /tmp/pybedtools.k6oxwiq3.tmp has inconsistent naming convention for record:\n",
      "\t3826\t3977\t.\t.\n",
      "\n",
      "***** WARNING: File /tmp/pybedtools.k6oxwiq3.tmp has inconsistent naming convention for record:\n",
      "\t3826\t3977\t.\t.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# df_associated_gene <- associated gene assignments\n",
    "df_associated_gene = pd.read_csv(BedTool.from_dataframe(df_atac[['chrom', 'start', 'end']]).map(\n",
    "    BedTool.from_dataframe(df_outron_extended_genes).fn, c='4,5', o='distinct,distinct',).fn, sep='\\t',\n",
    "    names=yp.NAMES_BED9[:5])\n",
    "df_associated_gene.columns = ['chrom', 'start', 'end', 'associated_gene_id', 'associated_locus_id']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "start_time": "2018-05-30T18:17:41.741Z"
    },
    "code_folding": [
     0
    ],
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Table S2 -- unstranded summary annotation\n",
    "def name_(annot_fwd, annot_rev, locus_id_fwd, locus_id_rev, enhancer_locus_id):\n",
    "    if (annot_fwd in config['annot_with_gene_id']) and not(annot_rev in config['annot_with_gene_id']):\n",
    "        return locus_id_fwd\n",
    "    elif not(annot_fwd in config['annot_with_gene_id']) and (annot_rev in config['annot_with_gene_id']):\n",
    "        return locus_id_rev\n",
    "    elif (annot_fwd in config['annot_with_gene_id']) and (annot_rev in config['annot_with_gene_id']):\n",
    "        return '%s / %s' % (locus_id_rev, locus_id_fwd)\n",
    "    elif enhancer_locus_id != '.': # show \"likeliest target gene for an enhancer\" in brackets for all remaining (=not coding_promoter or non_coding_promoter)\n",
    "        return '(%s)' % (enhancer_locus_id,)\n",
    "    else:\n",
    "        return ''\n",
    "\n",
    "def annot_merge_(annot_fwd, annot_rev):\n",
    "    if annot_fwd == 'coding_promoter' or annot_rev == 'coding_promoter':\n",
    "        return 'coding_promoter'\n",
    "    elif annot_fwd == 'pseudogene_promoter' or annot_rev == 'pseudogene_promoter':\n",
    "        return 'pseudogene_promoter'\n",
    "    elif annot_fwd == 'non-coding_RNA' or annot_rev == 'non-coding_RNA':\n",
    "        return 'non-coding_RNA'\n",
    "    elif annot_fwd == 'unknown_promoter' or annot_rev == 'unknown_promoter':\n",
    "        return 'unknown_promoter'\n",
    "    elif annot_fwd == 'transcription_initiation' or annot_rev == 'transcription_initiation':\n",
    "        return 'putative_enhancer'\n",
    "    elif annot_fwd == 'no_transcription' or annot_rev == 'no_transcription':\n",
    "        return 'other_element'\n",
    "    assert False\n",
    "\n",
    "def strand_merge_(annot_fwd, annot_rev):\n",
    "    promoter_fwd = (annot_fwd in config['annot_with_gene_id'])\n",
    "    promoter_rev = (annot_rev in config['annot_with_gene_id'])\n",
    "    lcap_fwd = annot_fwd == 'unknown_promoter'\n",
    "    lcap_rev = annot_rev == 'unknown_promoter'\n",
    "    if promoter_fwd and promoter_rev:\n",
    "        return '.'\n",
    "    elif promoter_fwd and (not promoter_rev):\n",
    "        return '+'\n",
    "    elif (not promoter_fwd) and promoter_rev:\n",
    "        return '-'\n",
    "    # in the absence of promoter annotations with an \"annotated\" target, use unknown_promoter annotations for strand\n",
    "    elif lcap_fwd and lcap_rev:\n",
    "        return '.'\n",
    "    elif lcap_fwd and (not lcap_rev):\n",
    "        return '+'\n",
    "    elif (not lcap_fwd) and lcap_rev:\n",
    "        return '-'\n",
    "    else:\n",
    "        return '.'\n",
    "\n",
    "df_regl = df_atac[['chrom', 'start', 'end']].copy()\n",
    "df_regl['tss_fwd'] = df_scap_fwd['scap_mode']\n",
    "df_regl['tss_rev'] = df_scap_rev['scap_mode']\n",
    "df_regl['scap_pass_fwd'] = df_scap_fwd['scap_pass']\n",
    "df_regl['scap_pass_rev'] = df_scap_rev['scap_pass']\n",
    "df_regl['scap_mode_fwd'] = df_scap_fwd['scap_mode']\n",
    "df_regl['scap_mode_rev'] = df_scap_rev['scap_mode']\n",
    "df_regl['scap_mode_count_fwd'] = df_scap_fwd['scap_mode_count']\n",
    "df_regl['scap_mode_count_rev'] = df_scap_rev['scap_mode_count']\n",
    "df_regl['scap_count_fwd'] = df_scap_fwd['scap_count']\n",
    "df_regl['scap_count_rev'] = df_scap_rev['scap_count']\n",
    "df_regl['annot_fwd'] = df_prom_fwd['annot_summary']\n",
    "df_regl['annot_rev'] = df_prom_rev['annot_summary']\n",
    "df_regl['annot_detailed_fwd'] = df_prom_fwd['annot_detailed_summary']\n",
    "df_regl['annot_detailed_rev'] = df_prom_rev['annot_detailed_summary']\n",
    "df_regl['annot'] = list(map(annot_merge_, df_prom_fwd['annot_summary'], df_prom_rev['annot_summary']))\n",
    "df_regl['strand'] = list(map(strand_merge_, df_prom_fwd['annot_summary'], df_prom_rev['annot_summary']))\n",
    "\n",
    "def agg_pass1_pass2(annot, pass1_val, pass2_val):\n",
    "    if annot in ['coding_promoter', 'pseudogene_promoter']:\n",
    "        return pass1_val\n",
    "    elif annot == 'non-coding_RNA':\n",
    "        return pass2_val\n",
    "    else:\n",
    "        return ''\n",
    "\n",
    "df_regl['promoter_gene_id_fwd'] = list(map(agg_pass1_pass2, df_regl['annot_fwd'],\n",
    "                                    df_exon_fwd['pass1_exon1_gene_id'], df_exon_fwd['pass2_exon1_gene_id']))\n",
    "df_regl['promoter_gene_id_rev'] = list(map(agg_pass1_pass2, df_regl['annot_rev'],\n",
    "                                    df_exon_rev['pass1_exon1_gene_id'], df_exon_rev['pass2_exon1_gene_id']))\n",
    "\n",
    "df_regl['promoter_locus_id_fwd'] = list(map(agg_pass1_pass2, df_regl['annot_fwd'],\n",
    "                                    df_exon_fwd['pass1_exon1_locus_id'], df_exon_fwd['pass2_exon1_locus_id']))\n",
    "df_regl['promoter_locus_id_rev'] = list(map(agg_pass1_pass2, df_regl['annot_rev'],\n",
    "                                    df_exon_rev['pass1_exon1_locus_id'], df_exon_rev['pass2_exon1_locus_id']))\n",
    "\n",
    "df_regl['promoter_gene_biotype_fwd'] = list(map(agg_pass1_pass2, df_regl['annot_fwd'],\n",
    "                                    df_exon_fwd['pass1_exon1_gene_biotype'], df_exon_fwd['pass2_exon1_gene_biotype']))\n",
    "df_regl['promoter_gene_biotype_rev'] = list(map(agg_pass1_pass2, df_regl['annot_rev'],\n",
    "                                    df_exon_rev['pass1_exon1_gene_biotype'], df_exon_rev['pass2_exon1_gene_biotype']))\n",
    "\n",
    "df_regl['associated_gene_id'] = df_associated_gene['associated_gene_id']\n",
    "df_regl['associated_locus_id'] = df_associated_gene['associated_locus_id']\n",
    "\n",
    "df_regl['label'] = list(map(name_,\n",
    "    df_regl['annot_fwd'], df_regl['annot_rev'],                     \n",
    "    df_regl['promoter_locus_id_fwd'], df_regl['promoter_locus_id_rev'],\n",
    "    df_regl['associated_locus_id']\n",
    "))\n",
    "\n",
    "for stage in config['stages_wt'] + config['stages_glp1']:\n",
    "    df_regl['atac_%s_height' % (stage,)] = df_atac['atac_%s_height' % (stage,)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "start_time": "2018-05-30T18:17:41.742Z"
    },
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "# chrom_,start_,end_ce11: site coordinates in ce11, obtained by lifting over the position with peak accessibility\n",
    "def liftover_(chroms, starts):#, ends, strands): # Liftover wrapper\n",
    "    \"\"\"\n",
    "    # http://www.ncbi.nlm.nih.gov/pmc/articles/PMC4512556/\n",
    "    # note that while annotations changed, the N2 genome sequence remained identical from WS215 through WS234\n",
    "    Test: \n",
    "    list(liftover_WS230_WS220(\n",
    "    chroms=['chrI', 'chrX', 'chrIII', 'chrV', 'chrIV', 'chrII'],\n",
    "    starts=[1,1,4,1,1, 1],\n",
    "    ends=[12507921,13713648,471811,15465174,11718826, 11718826],\n",
    "    strands=['+', '-', '+', '-', '+']))\n",
    "    \"\"\"\n",
    "    fp_inp = mp('_liftover_fp_inp.tsv')\n",
    "    fp_out = mp('_liftover_fp_out.tsv')\n",
    "    with open(fp_inp, 'w') as fh_inp:\n",
    "        #for (chrom, start, end, strand) in zip(chroms, starts, ends, strands):        \n",
    "        for (chrom, start) in zip(chroms, starts):\n",
    "            print('%s\\t%s' % (chrom, start), file=fh_inp)\n",
    "\n",
    "    liftover_bin = '/mnt/home1/ahringer/jj374/lab/raw_data/datastore/Kenneth_Evans/ConvertWormAssembly-1-0-0/ConvertWormAssembly.1.0.0.pl'\n",
    "    liftover_dir = '/mnt/home1/ahringer/jj374/lab/raw_data/datastore/Kenneth_Evans/ConvertWormAssembly-1-0-0/CHROMOSOME-DIFFERENCES-to-WS244'\n",
    "    !{liftover_bin} \\\n",
    "        --input={fp_inp} \\\n",
    "        --output={fp_out} \\\n",
    "        --release1=WS220 \\\n",
    "        --release2=WS235 \\\n",
    "        --chrfield=1 \\\n",
    "        --posfield=2 \\\n",
    "        --postype=1 \\\n",
    "        --refdir={liftover_dir}\n",
    "    \n",
    "    df_out = pd.read_csv(fp_out, sep='\\t', names=['chrom', 'start'])\n",
    "    for i, r in df_out.iterrows():\n",
    "        yield((r['chrom'], r['start']))\n",
    "        \n",
    "    !rm {fp_inp}\n",
    "    !rm {fp_out}\n",
    "\n",
    "df_ce11_pos = pd.DataFrame.from_records(liftover_(df_regl['chrom'], l_atac_peak_pos), columns=['chrom', 'pos'])\n",
    "df_regl['chrom_ce11'] = df_ce11_pos['chrom']\n",
    "df_regl['start_ce11'] = df_ce11_pos['pos'] - 75\n",
    "df_regl['end_ce11'] = df_ce11_pos['pos'] + 75 + 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "start_time": "2018-05-30T18:17:41.743Z"
    },
    "code_folding": [
     0
    ],
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# regulatory_annotation.tsv: subset of (most informative) annotation metrics\n",
    "fp_ = os.path.join(annot_, 'regulatory_annotation.tsv')\n",
    "l_cols_ = [\n",
    "    'chrom', 'start', 'end', \n",
    "    'chrom_ce11', 'start_ce11', 'end_ce11',\n",
    "    'annot', 'annot_fwd', 'annot_rev',\n",
    "    'promoter_gene_id_fwd', 'promoter_locus_id_fwd', 'promoter_gene_biotype_fwd',\n",
    "    'promoter_gene_id_rev', 'promoter_locus_id_rev', 'promoter_gene_biotype_rev',\n",
    "    'associated_gene_id', 'associated_locus_id',\n",
    "    'tss_fwd', 'tss_rev', 'scap_pass_fwd', 'scap_pass_rev',\n",
    "]\n",
    "\n",
    "l_cols_fwd_ = \\\n",
    "    ['lcap_%s_fwd_passed_jump' % (stage,) for stage in config['stages']] +\\\n",
    "    ['lcap_%s_fwd_passed_incr' % (stage,) for stage in config['stages']]\n",
    "\n",
    "l_cols_rev_ = \\\n",
    "    ['lcap_%s_rev_passed_jump' % (stage,) for stage in config['stages']] +\\\n",
    "    ['lcap_%s_rev_passed_incr' % (stage,) for stage in config['stages']]\n",
    "\n",
    "pd.concat([df_regl[l_cols_], df_lcap_fwd[l_cols_fwd_], df_lcap_rev[l_cols_rev_]], axis=1)\\\n",
    "    .to_csv(fp_, header=True, index=False, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "start_time": "2018-05-30T18:17:41.744Z"
    },
    "code_folding": [
     0
    ],
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# regulatory_annotation_%strand.bed: stranded .bed-files showing the annotation in individual stages\n",
    "write_gffbed(os.path.join(annot_, 'regulatory_annotation_fwd.bed'),\n",
    "        chrom = df_regl['chrom'],\n",
    "        start = df_regl['start'],\n",
    "        end = df_regl['end'],\n",
    "        name = df_regl['promoter_locus_id_fwd'],\n",
    "        strand = '+',\n",
    "        itemRgb = list(map(lambda annot: d_annot_strand_legend[annot], df_regl['annot_fwd'])),\n",
    "        attr = df_prom_fwd[['annot_%s' % (stage,) for stage in config['stages']]],\n",
    ")\n",
    "\n",
    "write_gffbed(os.path.join(annot_, 'regulatory_annotation_rev.bed'),\n",
    "        chrom = df_regl['chrom'],\n",
    "        start = df_regl['start'],\n",
    "        end = df_regl['end'],\n",
    "        name = df_regl['promoter_locus_id_rev'],\n",
    "        strand = '-',\n",
    "        itemRgb = list(map(lambda annot: d_annot_strand_legend[annot], df_regl['annot_rev'])),\n",
    "        attr = df_prom_rev[['annot_%s' % (stage,) for stage in config['stages']]],\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "start_time": "2018-05-30T18:17:41.745Z"
    },
    "code_folding": [
     0
    ],
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# regulatory_annotation_ce10.bed: final annotation as a single, unstranded .bed-file (ce10)\n",
    "def write_ce10_(fp_, df_regl_):\n",
    "    write_gffbed(fp_,\n",
    "        chrom = df_regl_['chrom'],\n",
    "        start = df_regl_['start'],\n",
    "        end = df_regl_['end'],\n",
    "        name = df_regl_['label'],\n",
    "        strand = df_regl_['strand'],\n",
    "        itemRgb = list(map(lambda annot: d_annot_legend[annot], df_regl_['annot'])),\n",
    "        attr = df_regl_[['annot', 'annot_fwd', 'annot_rev', 'annot_detailed_fwd', 'annot_detailed_rev']],\n",
    "    )\n",
    "\n",
    "write_ce10_(os.path.join(annot_, 'regulatory_annotation_ce10.bed'), df_regl)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "start_time": "2018-05-30T18:17:41.747Z"
    },
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "# metrics_type/regulatory_annotation.coding_promoter_only.bed.bed-files of specific types of elements\n",
    "for annot in config['annot']:\n",
    "    fp_ = mp('regulatory_annotation.%s.bed' % (annot,))\n",
    "    write_ce10_(fp_, df_regl.query('annot == \"%s\"' % (annot,)).reset_index(drop=True))\n",
    "    !wc -l {fp_}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "start_time": "2018-05-30T18:17:41.749Z"
    },
    "code_folding": [
     0
    ],
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# regulatory_annotation_ce11.bed: final annotation as a single, unstranded .bed-file (ce11)\n",
    "def write_ce11_(fp_, df_regl_):\n",
    "    write_gffbed(fp_,\n",
    "        chrom = df_regl_['chrom_ce11'],\n",
    "        start = df_regl_['start_ce11'],\n",
    "        end = df_regl_['end_ce11'],\n",
    "        name = df_regl_['label'],\n",
    "        strand = df_regl_['strand'],\n",
    "        itemRgb = list(map(lambda annot: d_annot_legend[annot], df_regl_['annot'])),\n",
    "        attr = df_regl_[['annot', 'annot_fwd', 'annot_rev', 'annot_detailed_fwd', 'annot_detailed_rev']],\n",
    "    )\n",
    "\n",
    "write_ce11_(os.path.join(annot_, 'regulatory_annotation_ce11.bed'), df_regl)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "start_time": "2018-05-30T18:17:41.749Z"
    },
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "# Table of short cap bidirectionality statistics by type of element\n",
    "def directionality_(pass_fwd, pass_rev):\n",
    "    if pass_fwd and pass_rev:\n",
    "        return 'bidirectional'\n",
    "    elif pass_fwd:\n",
    "        return 'forward'\n",
    "    elif pass_rev:\n",
    "        return 'reverse'\n",
    "    else:\n",
    "        return 'none/weak'\n",
    "\n",
    "df_has_scap = pd.DataFrame()\n",
    "df_has_scap['annot'] = df_regl['annot']\n",
    "df_has_scap['directionality'] = [*map(directionality_, df_scap_fwd['scap_pass'], df_scap_rev['scap_pass'])]\n",
    "df_has_scap['directionality'].value_counts()\n",
    "\n",
    "df_ = pd.crosstab(pd.Categorical(df_has_scap['annot']), pd.Categorical(df_has_scap['directionality']), margins=True)\\\n",
    ".loc[config['annot'] + ['All'], ['bidirectional', 'forward', 'reverse', 'none/weak']]\n",
    "\n",
    "fig = plt.figure(figsize=(4,2))\n",
    "plt.subplot(111, frame_on=False)\n",
    "plt.gca().xaxis.set_visible(False)\n",
    "plt.gca().yaxis.set_visible(False)\n",
    "pd.plotting.table(data=df_, ax=plt.gca(), loc='center left')\n",
    "#plt.savefig('annot/FigA_mapping/annot_crosstab_scap.pdf', bbox_inches='tight')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "start_time": "2018-05-30T18:17:41.751Z"
    },
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "# Pie chart of different types of promoters\n",
    "n_bid = len(df_regl.query('(annot == \"coding_promoter\") & (strand == \".\")'))\n",
    "n_fwd = len(df_regl.query('(annot == \"coding_promoter\") & (strand == \"+\")'))\n",
    "n_rev = len(df_regl.query('(annot == \"coding_promoter\") & (strand == \"-\")'))\n",
    "print('%d bidirectional promoters' % (n_bid))\n",
    "print('%d forward-only promoters' % (n_fwd,))\n",
    "print('%d reverse-only promoters' % (n_rev,))\n",
    "plt.figure(figsize=(16,6)).subplots_adjust(wspace=0.4)\n",
    "plt.subplot(1,2,1, aspect=1)\n",
    "plt.title('%d of %d sites (%.1f%%) annotated\\nas protein-coding promoters' % \\\n",
    "    (n_fwd + n_rev + n_bid, len(df_regl), 100 * (n_fwd + n_rev + n_bid) / float(len(df_regl)),))\n",
    "plt.pie([n_fwd, n_rev, n_bid], labels=[\n",
    "        '%(n_fwd)d forward\\npromoters' % locals(),\n",
    "        '%(n_rev)d reverse\\npromoters' % locals(), \n",
    "        '%(n_bid)d bidirectional\\npromoters' % locals(),\n",
    "])[-1][-1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "anaconda-cloud": {},
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
