All files are based on the WS260 production release, obtained from the [WormBase FTP](ftp://ftp.wormbase.org/pub/wormbase/releases/WS260/), with genomic coordinates backlifted to [`WBcel215/ce10`](http://hgdownload.soe.ucsc.edu/goldenPath/ce10/bigZips/).

![WS260_ce10_ztf11.png](WS260_ce10_ztf11.png)

- [`WS260_ce10.canonical_geneset.gtf.gz`](WS260_ce10.canonical_geneset.gtf.gz) has the entire WS260 [canonical geneset](http://blog.wormbase.org/2016/06/20/gene-transfer-format-gtf-files-now-available-2/) in WBcel215(ce10) genome coordinates, converted using Kenneth's liftover tool. The conversion from WBcel235 to WBcel215 changed 96% (699702 / 726282) of all records; however, in practice, features seem to stay within 100bp between the two genome releases.
- In addition to the transcript structure as shown on the default RefSeq annotation in IGV, the canonical geneset also includes "summary rafts that show gene boundaries" for every gene. [`WS260_ce10.transcripts.gtf.gz`](WS260_ce10.transcripts.gtf.gz) only includes the transcript structure (so "looks like the default IGV Gene track") whereas [`WS260_ce10.genes.gtf.gz`](WS260_ce10.genes.gtf.gz) only includes gene boundaries.
- [`WS260_ce10.genes_by_CV.gtf`](WS260_ce10.genes_by_CV.gtf) contains gene rafts, annotated by whether they are developmentally stable or regulated, following the approach taken in [(Evans et al. 2016)](https://doi.org/10.1073/pnas.1608162113). Genes were ranked by [(Gerstein et al., 2014)](https://doi.org/10.1038/nature13424) maximum expression values in six stages covering wild-type development (LE, L1, L2, L3, L4, YA). Top 15,000 most expressed genes were then ranked based on the CV values, with the top third annotated as stable (orange), middle third as intermediate (blue), and bottom third as regulated (black). The remaining genes (e.g. non-coding, lowly expressed) were coloured light blue.
- [`WS260_ce10.exon_first.bed`](WS260_ce10.exon_first.bed) and [`WS260_ce10.exon_other.bed`](WS260_ce10.exon_other.bed) contain first (other) exons of protein-coding transcripts. These are used to annotate the closest downstream gene (of a hypersensitive site), and identify sites that overlap known exons, and cannot thus be annotated for transcription elongation.
- [`WS260_ce10.transcripts.annot.gtf.gz`](WS260_ce10.transcripts.annot.gtf.gz) has WS260 transcripts, explicitly limited to what is used in [the annotation](/annot/S2_regulatory_annotation): `protein_coding`, `pseudogene`, `tRNA`, `snoRNA`, `miRNA`, `snRNA`, `rRNA`.
